// @bitrix24/b24style/dist/plugins/fonts.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FontLocal_path, _FontLocal_typeList, _FontLocal_unicodeRange, _FontLocal_family, _FontLocal_localTitle, _FontLocal_weight, _FontLocal_style, _FontLocal_display, _FontsCollection_list;
var FontStyle;
(function (FontStyle) {
    FontStyle["normal"] = "normal";
    FontStyle["italic"] = "italic";
})(FontStyle || (FontStyle = {}));
var FontDisplay;
(function (FontDisplay) {
    FontDisplay["swap"] = "swap";
})(FontDisplay || (FontDisplay = {}));
var FontType;
(function (FontType) {
    FontType["local"] = "local";
    FontType["woff"] = "woff";
    FontType["woff2"] = "woff2";
    FontType["ttf"] = "ttf";
    FontType["eot"] = "eot";
})(FontType || (FontType = {}));
class FontLocal {
    constructor(family, weight, style, display, localTitle, path, typeList = [], unicodeRange = null) {
        _FontLocal_path.set(this, void 0);
        _FontLocal_typeList.set(this, void 0);
        _FontLocal_unicodeRange.set(this, void 0);
        _FontLocal_family.set(this, void 0);
        _FontLocal_localTitle.set(this, void 0);
        _FontLocal_weight.set(this, void 0);
        _FontLocal_style.set(this, void 0);
        _FontLocal_display.set(this, void 0);
        __classPrivateFieldSet(this, _FontLocal_family, family, "f");
        __classPrivateFieldSet(this, _FontLocal_weight, weight, "f");
        __classPrivateFieldSet(this, _FontLocal_style, style, "f");
        __classPrivateFieldSet(this, _FontLocal_display, display, "f");
        __classPrivateFieldSet(this, _FontLocal_localTitle, localTitle, "f");
        __classPrivateFieldSet(this, _FontLocal_typeList, typeList, "f");
        __classPrivateFieldSet(this, _FontLocal_path, path, "f");
        __classPrivateFieldSet(this, _FontLocal_unicodeRange, unicodeRange, "f");
    }
    getPath() {
        return __classPrivateFieldGet(this, _FontLocal_path, "f");
    }
    getLocalTitle() {
        return __classPrivateFieldGet(this, _FontLocal_localTitle, "f");
    }
    getFamily() {
        return __classPrivateFieldGet(this, _FontLocal_family, "f");
    }
    getWeight() {
        return `${__classPrivateFieldGet(this, _FontLocal_weight, "f")}`;
    }
    getStyle() {
        return __classPrivateFieldGet(this, _FontLocal_style, "f");
    }
    getDisplay() {
        return __classPrivateFieldGet(this, _FontLocal_display, "f");
    }
    getTypeList() {
        return __classPrivateFieldGet(this, _FontLocal_typeList, "f");
    }
    getUnicodeRange() {
        return __classPrivateFieldGet(this, _FontLocal_unicodeRange, "f") || '';
    }
    getSrc() {
        if (this.getTypeList().length < 1) {
            return '';
        }
        return this.getTypeList().map((type) => {
            switch (type) {
                case FontType.local:
                    return `local('${this.getLocalTitle()}')`;
                case FontType.woff2:
                    return `url('${this.getPath()}.woff2') format('woff2')`;
                case FontType.woff:
                    return `url('${this.getPath()}.woff') format('woff')`;
                case FontType.ttf:
                    return `url('${this.getPath()}.ttf') format('truetype')`;
                case FontType.eot:
                    return `url('${this.getPath()}.eot') format('eot')`;
            }
        }).join(',');
    }
    getConfig() {
        return {
            '@font-face': Object.assign({}, {
                fontFamily: this.getFamily(),
                fontWeight: this.getWeight(),
                fontStyle: this.getStyle(),
                fontDisplay: this.getDisplay(),
                src: this.getSrc(),
            }, (!!__classPrivateFieldGet(this, _FontLocal_unicodeRange, "f")
                ? {
                    unicodeRange: this.getUnicodeRange()
                }
                : {}))
        };
    }
    getTitle() {
        return `${this.getFamily()} ${this.getWeight()} ${this.getStyle()}`;
    }
}
_FontLocal_path = new WeakMap(), _FontLocal_typeList = new WeakMap(), _FontLocal_unicodeRange = new WeakMap(), _FontLocal_family = new WeakMap(), _FontLocal_localTitle = new WeakMap(), _FontLocal_weight = new WeakMap(), _FontLocal_style = new WeakMap(), _FontLocal_display = new WeakMap();
class FontsCollection {
    constructor() {
        _FontsCollection_list.set(this, []);
    }
    addItem(item) {
        __classPrivateFieldGet(this, _FontsCollection_list, "f").push(item);
        return this;
    }
    getList() {
        return __classPrivateFieldGet(this, _FontsCollection_list, "f");
    }
}
_FontsCollection_list = new WeakMap();
// endregion ////
// region Init ////
const groupsFonts = new FontsCollection();
// region Roboto ////
groupsFonts.addItem(new FontLocal('Roboto', 300, FontStyle.normal, FontDisplay.swap, 'Roboto Light', 'fonts/roboto/roboto-light-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto', 300, FontStyle.italic, FontDisplay.swap, 'Roboto Light Italic', 'fonts/roboto/roboto-lightitalic-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto', 400, FontStyle.normal, FontDisplay.swap, 'Roboto', 'fonts/roboto/roboto-regular-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto', 400, FontStyle.italic, FontDisplay.swap, 'Roboto Italic', 'fonts/roboto/roboto-italic-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Medium', 500, FontStyle.normal, FontDisplay.swap, 'Roboto', 'fonts/roboto/roboto-medium-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Medium Italic', 500, FontStyle.italic, FontDisplay.swap, 'Roboto Italic', 'fonts/roboto/roboto-mediumitalic-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Bold', 700, FontStyle.normal, FontDisplay.swap, 'Roboto', 'fonts/roboto/roboto-bold-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Bold Italic', 700, FontStyle.italic, FontDisplay.swap, 'Roboto Italic', 'fonts/roboto/roboto-bolditalic-webfont', [FontType.local, FontType.woff2, FontType.woff]))
    // endregion ////
    // region Roboto.Mono ////
    .addItem(new FontLocal('Roboto Mono', 300, FontStyle.normal, FontDisplay.swap, 'Roboto Mono Light', 'fonts/robotomono/robotomono-light', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Mono', 300, FontStyle.italic, FontDisplay.swap, 'Roboto Mono Light Italic', 'fonts/robotomono/robotomono-lightitalic', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Mono', 400, FontStyle.normal, FontDisplay.swap, 'Roboto Mono', 'fonts/robotomono/robotomono-regular', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Mono', 400, FontStyle.italic, FontDisplay.swap, 'Roboto Mono Italic', 'fonts/robotomono/robotomono-regularitalic', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Mono', 500, FontStyle.normal, FontDisplay.swap, 'Roboto Mono Medium', 'fonts/robotomono/robotomono-medium', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Mono', 500, FontStyle.italic, FontDisplay.swap, 'Roboto Mono Medium Italic', 'fonts/robotomono/robotomono-mediumitalic', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Mono', 700, FontStyle.normal, FontDisplay.swap, 'Roboto Mono Bold', 'fonts/robotomono/robotomono-bold', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Roboto Mono', 700, FontStyle.italic, FontDisplay.swap, 'Roboto Mono Bold Italic', 'fonts/robotomono/robotomono-bolditalic', [FontType.local, FontType.woff2, FontType.woff]))
    // endregion ////
    // region Montserrat ////
    .addItem(new FontLocal('Montserrat', 400, FontStyle.normal, FontDisplay.swap, 'Montserrat', 'fonts/montserrat/montserrat-regular', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Montserrat', 500, FontStyle.normal, FontDisplay.swap, 'Montserrat Medium', 'fonts/montserrat/montserrat-medium', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Montserrat', 600, FontStyle.normal, FontDisplay.swap, 'Montserrat Semibold', 'fonts/montserrat/montserrat-semibold', [FontType.local, FontType.woff2, FontType.woff]))
    .addItem(new FontLocal('Montserrat', 700, FontStyle.normal, FontDisplay.swap, 'Montserrat Bold', 'fonts/montserrat/montserrat-bold', [FontType.local, FontType.woff2, FontType.woff]))
    // endregion ////
    // region Open Sans ////
    .addItem(new FontLocal('Open Sans', 300, FontStyle.normal, FontDisplay.swap, 'Open Sans Light', 'fonts/opensans/opensans-light', [FontType.local, FontType.woff, FontType.ttf]))
    .addItem(new FontLocal('Open Sans', 400, FontStyle.normal, FontDisplay.swap, 'Open Sans', 'fonts/opensans/opensans-regular', [FontType.local, FontType.woff, FontType.ttf]))
    .addItem(new FontLocal('Open Sans', 600, FontStyle.normal, FontDisplay.swap, 'Open Sans Semibold', 'fonts/opensans/opensans-semibold', [FontType.local, FontType.woff, FontType.ttf]))
    .addItem(new FontLocal('Open Sans', 700, FontStyle.normal, FontDisplay.swap, 'Open Sans Bold', 'fonts/opensans/opensans-bold', [FontType.local, FontType.woff, FontType.ttf]))
    // endregion ////
    // region Comforter Brush ////
    .addItem(new FontLocal('Comforter Brush', 400, FontStyle.normal, FontDisplay.swap, '', 'fonts/comforter-brush/comforter-brush-regular', [FontType.woff2], 'U+0400-045F, U+0490-0491, U+04B0-04B1, U+2116'))
    .addItem(new FontLocal('Comforter Brush', 400, FontStyle.normal, FontDisplay.swap, '', 'fonts/comforter-brush/comforter-brush-regular', [FontType.woff2], 'U+0100-024F, U+0259, U+1E00-1EFF, U+2020, U+20A0-20AB, U+20AD-20CF, U+2113, U+2C60-2C7F, U+A720-A7FF'))
    .addItem(new FontLocal('Comforter Brush', 400, FontStyle.normal, FontDisplay.swap, '', 'fonts/comforter-brush/comforter-brush-regular', [FontType.woff2], 'U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD'));
// endregion ////
// endregion ////
export default groupsFonts;
