// @bitrix24/b24ui-nuxt/dist/shared/b24ui-nuxt.COkgCO-x.mjs
import 'node:url';
import { kebabCase } from 'scule';
import { addTypeTemplate, addTemplate } from '@nuxt/kit';
import { defuFn } from 'defu';
import { genExport } from 'knitwork';

const name = "@bitrix24/b24ui-nuxt";
const version = "1.0.4";

const getDefaultUiConfig = () => ({
  version
});
const defaultOptions = {
  colorMode: true,
  version
};

const accordion = {
  slots: {
    root: "w-full",
    item: "text-(--b24ui-typography-label-color) border-b border-(--ui-color-divider-vibrant-accent-more) last:border-b-0",
    header: "flex",
    trigger: [
      "min-w-0 group flex-1 flex items-center gap-1.5",
      "py-3",
      "font-(--ui-font-weight-medium) text-(length:--ui-font-size-sm)",
      "focus-visible:outline-(--ui-color-accent-soft-element-blue)",
      "cursor-pointer"
    ].join(" "),
    content: "motion-safe:data-[state=open]:animate-[accordion-down_200ms_ease-out] motion-safe:data-[state=closed]:animate-[accordion-up_200ms_ease-out] overflow-hidden focus:outline-none",
    body: "text-(length:--ui-font-size-sm) pb-3",
    leadingIcon: "shrink-0 size-5",
    trailingIcon: "shrink-0 size-5 ms-auto group-data-[state=open]:rotate-180 transition-transform duration-200",
    label: "text-start break-words"
  },
  variants: {
    disabled: {
      true: {
        trigger: "cursor-not-allowed opacity-75"
      }
    }
  }
};

const advice = {
  slots: {
    root: [
      "light",
      "style-outline-accent-1",
      "flex items-end"
    ].join(" "),
    descriptionWrapper: "relative",
    descriptionBorder: "fill-(--b24ui-border-color)",
    descriptionBg: [
      "fill-(--b24ui-background)",
      "dark:fill-(--ui-color-base-6)"
    ].join(" "),
    descriptionAngle: [
      "absolute",
      "w-[14px] h-[12px]"
    ].join(" "),
    description: [
      "grow",
      "w-11/12 py-3 px-md2 ms-2",
      "rounded-[23px]",
      "font-[family-name:var(--ui-font-family-secondary)] text-(length:--ui-font-size-md)/(--ui-font-line-height-md) font-(--ui-font-weight-normal)",
      "border-1",
      "border-(--b24ui-border-color) bg-(--b24ui-background) text-(--b24ui-color)",
      "dark:bg-(--ui-color-base-6)"
    ].join(" "),
    leading: "me-1.5 ms-2 font-(--ui-font-weight-medium) text-(--ui-color-design-plain-content-icon-secondary)",
    leadingIcon: "shrink-0 size-[42px]",
    leadingAvatar: "shrink-0",
    leadingAvatarIcon: "text-(--b24ui-typography-label-color) bg-(--ui-color-base-8)",
    leadingAvatarSize: "lg"
  },
  variants: {
    angle: {
      top: {
        root: "items-start",
        leading: "mt-0.5",
        descriptionAngle: [
          "start-[0.8px] top-[9px]",
          "scale-x-100 -scale-y-100",
          "rtl:-scale-x-100"
        ].join(" ")
      },
      bottom: {
        root: "items-end",
        descriptionAngle: [
          "start-[0.8px] bottom-[9px]",
          "rtl:-scale-x-100"
        ].join(" ")
      }
    }
  },
  defaultVariants: {
    angle: "bottom"
  }
};

const alert = {
  slots: {
    root: [
      "relative overflow-hidden w-full flex",
      "text-(--b24ui-color)",
      "bg-(--b24ui-background)",
      "border-(--b24ui-border-color) border-(length:--b24ui-border-width)",
      "rounded-(--ui-border-radius-md)"
    ].join(" "),
    wrapper: [
      "min-w-0 flex-1 flex flex-col",
      "font-[family-name:var(--ui-font-family-primary)]"
    ].join(" "),
    title: "font-bold",
    description: "",
    icon: "shrink-0 size-6",
    avatar: "shrink-0",
    avatarSize: "",
    actions: "flex flex-wrap gap-1.5 shrink-0",
    close: "p-0"
  },
  variants: {
    color: {
      "air-primary": { root: "style-filled" },
      "air-primary-success": { root: "style-filled-success" },
      "air-primary-alert": { root: "style-filled-alert" },
      "air-primary-copilot": { root: "style-filled-copilot" },
      "air-primary-warning": { root: "style-filled-warning" },
      "air-secondary": { root: "style-tinted" },
      "air-secondary-alert": { root: "style-tinted-alert" },
      "air-secondary-accent": { root: "style-outline" },
      "air-secondary-accent-1": { root: "style-outline-accent-1" },
      "air-secondary-accent-2": { root: "style-outline-accent-2" },
      "air-tertiary": { root: "style-outline-no-accent" },
      // @deprecate ////
      "default": { root: "style-old-default" },
      "danger": { root: "style-old-danger" },
      "success": { root: "style-old-success" },
      "warning": { root: "style-old-warning" },
      "primary": { root: "style-old-primary" },
      "secondary": { root: "style-old-secondary" },
      "collab": { root: "style-old-collab" },
      "ai": { root: "style-old-ai" }
    },
    size: {
      sm: {
        root: "py-xs ps-sm pe-xs gap-2",
        title: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-lg)",
        description: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-lg)",
        avatarSize: "md"
      },
      md: {
        root: "py-md ps-md pe-xs gap-2.5",
        title: "text-(length:--ui-font-size-md)/(--ui-font-line-height-reset)",
        description: "text-(length:--ui-font-size-md)/(--ui-font-line-height-3xs)",
        avatarSize: "xl"
      }
    },
    orientation: {
      horizontal: {
        root: "items-center",
        actions: "items-center"
      },
      vertical: {
        root: "items-start",
        actions: "items-start mt-2"
      }
    },
    title: {
      true: {
        description: "mt-1"
      }
    },
    inverted: {
      true: "",
      false: ""
    }
  },
  compoundVariants: [
    // region color ////
    {
      inverted: true,
      color: "air-primary",
      class: {
        root: "style-filled-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-success",
      class: {
        root: "style-filled-success-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-alert",
      class: {
        root: "style-filled-alert-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-copilot",
      class: {
        root: "style-filled-copilot-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-warning",
      class: {
        root: "style-filled-warning-inverted"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-secondary-accent",
    size: "md",
    inverted: false
  }
};

const avatar = {
  slots: {
    root: [
      "air-secondary-accent",
      "inline-flex items-center justify-center shrink-0",
      "select-none",
      "rounded-full",
      "align-middle",
      "bg-(--ui-color-base-8)",
      "ring ring-(--ui-color-base-7)"
      // 'overflow-hidden'
    ].join(" "),
    image: "h-full w-full rounded-[inherit] object-cover",
    fallback: "font-(--ui-font-weight-medium) text-(--ui-color-design-plain-content-icon-secondary) truncate",
    icon: "text-(--ui-color-design-plain-content-icon-secondary) shrink-0"
  },
  variants: {
    size: {
      "3xs": {
        root: "size-[10px] text-4xs font-(--ui-font-weight-regular)",
        icon: "size-[10px]"
      },
      "2xs": {
        root: "size-[20px] text-(length:--ui-font-size-4xs)/(--ui-font-line-height-reset) font-(--ui-font-weight-regular)",
        icon: "size-[20px]"
      },
      "xs": {
        root: "size-[24px] text-(length:--ui-font-size-3xs)/(--ui-font-line-height-reset) font-(--ui-font-weight-regular)",
        icon: "size-[24px]"
      },
      "sm": {
        root: "size-[28px] text-(length:--ui-font-size-xs)/(--ui-font-line-height-reset)",
        icon: "size-[28px]"
      },
      "md": {
        root: "size-[32px] text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset)",
        icon: "size-[32px]"
      },
      "lg": {
        root: "size-[42px] text-(length:--ui-font-size-2xl)/(--ui-font-line-height-reset)",
        icon: "size-[42px]"
      },
      "xl": {
        root: "ring-2 size-[48px] text-(length:--ui-font-size-2xl)/(--ui-font-line-height-reset)",
        icon: "size-[48px]"
      },
      "2xl": {
        root: "ring-2 size-[60px] text-(length:--ui-font-size-5xl)/(--ui-font-line-height-reset)",
        icon: "size-[60px]"
      },
      "3xl": {
        root: "ring-2 size-[94px] text-[34px]/(--ui-font-line-height-reset)",
        icon: "size-[94px]"
      }
    }
  },
  defaultVariants: {
    size: "md"
  }
};

const avatarGroup = {
  slots: {
    root: "inline-flex justify-end",
    base: "relative rounded-full last:me-0"
  },
  variants: {
    size: {
      "3xs": { base: "-me-0.5" },
      "2xs": { base: "-me-0.5" },
      "xs": { base: "-me-0.5" },
      "sm": { base: "-me-1.5" },
      "md": { base: "-me-1.5" },
      "lg": { base: "-me-1.5" },
      "xl": { base: "-me-2" },
      "2xl": { base: "-me-2" },
      "3xl": { base: "-me-2" }
    }
  },
  defaultVariants: {
    size: "md"
  }
};

const buttonGroupVariant = {
  buttonGroup: {
    horizontal: [
      "focus-visible:outline-none",
      "ring ring-inset ring-0 focus-visible:ring-2",
      "group-[.is-button-group]/items:not-only:first:rounded-e-none group-[.is-button-group]/items:not-only:last:rounded-s-none group-[.is-button-group]/items:not-last:not-first:rounded-none",
      "group-[.is-button-group]/items:not-only:first:border-e-0 group-[.is-button-group]/items:not-only:not-first:border-s-0",
      "focus-visible:z-[1]"
    ].join(" "),
    vertical: [
      "focus-visible:outline-none",
      "ring ring-inset ring-0 focus-visible:ring-2",
      "not-only:first:rounded-b-none not-only:last:rounded-t-none not-last:not-first:rounded-none",
      "focus-visible:z-[1]"
    ].join(" ")
  },
  noSplit: {
    false: [
      "group-[.is-button-group]/items:not-only:not-first:after:content-[''] group-[.is-button-group]/items:not-only:not-first:after:absolute",
      "group-[.is-button-group]/items:not-only:not-first:after:top-[7px] group-[.is-button-group]/items:not-only:not-first:after:bottom-[6px] group-[.is-button-group]/items:not-only:not-first:after:left-0 group-[.is-button-group]/items:not-only:not-first:after:w-px",
      "group-[.is-button-group]/items:not-only:not-first:after:bg-current/30"
    ].join(" ")
  }
};
const buttonGroupVariantWithRoot = {
  buttonGroup: {
    horizontal: {
      root: "group leading-none has-focus-visible:z-[1]",
      base: [
        "focus-visible:outline-none",
        "ring ring-inset ring-1 focus-visible:ring-2",
        "group-not-only:group-first:rounded-e-3xl group-not-only:group-last:rounded-s-none group-not-last:group-not-first:rounded-none",
        "group-not-only:group-first:rounded-e-none group-not-only:group-last:rounded-s-none group-not-last:group-not-first:rounded-none",
        "group-not-only:group-first:border-e-0 group-not-only:group-not-first:border-s-0"
      ].join(" ")
    },
    vertical: {
      root: "group has-focus-visible:z-[1]",
      base: [
        "focus-visible:outline-none",
        "ring ring-inset ring-1 focus-visible:ring-2",
        "group-not-only:group-first:rounded-b-none group-not-only:group-last:rounded-t-none group-not-last:group-not-first:rounded-none"
      ].join(" ")
    }
  },
  noSplit: {
    false: [
      "group-not-only:not-first:after:content-[''] group-not-only:not-first:after:absolute",
      "group-not-only:not-first:after:top-[7px] group-not-only:not-first:after:bottom-[6px] group-not-only:not-first:after:left-0 group-not-only:not-first:after:w-px",
      "group-not-only:not-first:after:bg-current/30"
    ].join(" ")
  }
};
const buttonGroup = {
  base: "relative",
  variants: {
    size: {
      xss: "",
      xs: "",
      sm: "",
      md: "",
      lg: "",
      xl: ""
    },
    orientation: {
      horizontal: "flex flex-row -space-x-px",
      vertical: "flex flex-col -space-y-px"
    }
  }
};

const badge = {
  slots: {
    base: [
      "ui-label__scope --air",
      "select-none",
      "font-[family-name:var(--ui-font-family-secondary)]",
      "font-(--ui-label-font-weight)",
      "text-(length:--ui-label-font-size)/normal",
      "inline-flex items-center",
      "transition-all duration-200 ease-linear",
      "px-(--ui-label-inline-space)",
      "text-(--b24ui-color)",
      "bg-(--b24ui-background)",
      "border-(--b24ui-border-color) border-(length:--b24ui-border-width)"
    ].join(" "),
    wrapper: "h-(--ui-label-height) inline-flex items-center",
    label: "max-w-full whitespace-nowrap text-ellipsis decoration-from-font",
    leadingIcon: "shrink-0",
    leadingAvatar: "shrink-0",
    leadingAvatarSize: "",
    trailingIcon: "shrink-0 cursor-pointer hover:rounded-(--ui-border-radius-circle) hover:bg-current/20"
  },
  variants: {
    ...buttonGroupVariant,
    useLink: {
      true: {
        base: "cursor-pointer",
        wrapper: "group",
        label: "group-hover:underline group-hover:decoration-dashed"
      }
    },
    useClose: {
      true: "pe-2xs",
      false: ""
    },
    leading: {
      true: "ps-1",
      false: ""
    },
    color: {
      "air-primary": { base: "style-filled" },
      "air-primary-success": { base: "style-filled-success" },
      "air-primary-alert": { base: "style-filled-alert" },
      "air-primary-copilot": { base: "style-filled-copilot" },
      "air-primary-warning": { base: "style-filled-warning" },
      "air-secondary": { base: "style-tinted" },
      "air-secondary-alert": { base: "style-tinted-alert" },
      "air-secondary-accent": { base: "style-outline" },
      "air-secondary-accent-1": { base: "style-outline-accent-1" },
      "air-secondary-accent-2": { base: "style-outline-accent-2" },
      "air-tertiary": { base: "style-outline-no-accent" },
      "air-selection": { base: "style-selection" },
      // @deprecate ////
      "default": { base: "style-old-default" },
      "danger": { base: "style-old-danger" },
      "success": { base: "style-old-success" },
      "warning": { base: "style-old-warning" },
      "primary": { base: "style-old-primary" },
      "secondary": { base: "style-old-secondary" },
      "collab": { base: "style-old-collab" },
      "ai": { base: "style-old-ai" }
    },
    size: {
      xss: {
        base: "ui-label-xss gap-0.5",
        wrapper: "gap-0.5",
        label: "underline-offset-1",
        leadingIcon: "size-[12px]",
        leadingAvatarSize: "3xs",
        trailingIcon: "size-[12px]"
      },
      xs: {
        base: "ui-label-xs gap-0.5",
        wrapper: "gap-0.5",
        label: "underline-offset-1",
        leadingIcon: "size-[12px]",
        leadingAvatarSize: "3xs",
        trailingIcon: "size-[12px]"
      },
      sm: {
        base: "ui-label-sm gap-1",
        wrapper: "gap-1",
        label: "underline-offset-1",
        leadingIcon: "size-[14px]",
        leadingAvatarSize: "3xs",
        trailingIcon: "size-[14px]"
      },
      md: {
        base: "ui-label-md gap-1",
        wrapper: "gap-1",
        label: "underline-offset-1",
        leadingIcon: "size-[15px]",
        leadingAvatarSize: "3xs",
        trailingIcon: "size-[15px]"
      },
      lg: {
        base: "ui-label-lg gap-1",
        wrapper: "gap-1",
        label: "",
        leadingIcon: "size-[22px]",
        leadingAvatarSize: "2xs",
        trailingIcon: "size-[22px]"
      },
      xl: {
        base: "ui-label-xl gap-1",
        wrapper: "gap-1",
        label: "",
        leadingIcon: "size-[22px]",
        leadingAvatarSize: "2xs",
        trailingIcon: "size-[22px]"
      }
    },
    square: {
      true: {
        base: "rounded-[calc(var(--ui-label-height)_/_4)] ",
        wrapper: "w-(--ui-label-height)",
        label: "overflow-hidden"
      },
      false: {
        base: "rounded-[calc(var(--ui-label-height)_/_2)]"
      }
    },
    inverted: {
      true: {
        base: "border-(--b24ui-color)"
      },
      false: ""
    }
  },
  compoundVariants: [
    // region color ////
    {
      inverted: true,
      color: "air-primary",
      class: {
        base: "style-filled-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-success",
      class: {
        base: "style-filled-success-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-alert",
      class: {
        base: "style-filled-alert-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-copilot",
      class: {
        base: "style-filled-copilot-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-warning",
      class: {
        base: "style-filled-warning-inverted"
      }
    },
    // endregion ////
    // region square ////
    {
      size: "xss",
      square: true,
      class: {
        base: "p-0 ps-0 pe-0",
        wrapper: "px-[2px] gap-0",
        leadingIcon: "size-[6px]"
      }
    },
    {
      size: "xs",
      square: true,
      class: {
        base: "p-0 ps-0 pe-0",
        wrapper: "px-[2px] gap-0",
        leadingIcon: "size-[10px]"
      }
    },
    {
      size: "sm",
      square: true,
      class: {
        base: "p-0 ps-0 pe-0",
        wrapper: "p-[1px] gap-0",
        leadingIcon: "size-[16px]"
      }
    },
    {
      size: "md",
      square: true,
      class: {
        base: "p-0 ps-0 pe-0",
        wrapper: "p-[1px] gap-0",
        leadingIcon: "size-[18px]"
      }
    },
    {
      size: "lg",
      square: true,
      class: {
        base: "p-0 ps-0 pe-0",
        wrapper: "p-[1px] gap-0",
        leadingIcon: "size-[23px]"
      }
    },
    {
      size: "xl",
      square: true,
      class: {
        base: "p-0 ps-0 pe-0",
        wrapper: "p-[1px] gap-0",
        leadingIcon: "size-[28px]"
      }
    },
    // endregion ////
    // region buttonGroup ////
    {
      buttonGroup: ["horizontal", "vertical"],
      class: "rounded-(--ui-border-radius-md)"
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-primary",
    size: "md",
    square: false,
    inverted: false
  }
};

const safeList$1 = [
  "invisible"
].join(" ");
const button = {
  slots: {
    base: [
      "ui-btn",
      "font-[family-name:var(--ui-font-family-primary)]",
      "select-none cursor-pointer inline-flex items-center",
      "relative",
      "outline-transparent focus-visible:outline-2 focus-visible:outline-offset-2",
      "disabled:cursor-not-allowed aria-disabled:cursor-not-allowed disabled:opacity-30 aria-disabled:opacity-30",
      "transition duration-0 ease-linear",
      // transition-colors
      "border-(length:--ui-btn-border-width)",
      "text-(--ui-btn-color) bg-(--ui-btn-background) border-(--ui-btn-border-color)",
      "hover:text-(--ui-btn-color-hover) hover:bg-(--ui-btn-background-hover) hover:border-(--ui-btn-border-color-hover)",
      "focus:text-(--ui-btn-color-hover) focus:bg-(--ui-btn-background-hover) focus:border-(--ui-btn-border-color-hover)",
      "active:text-(--ui-btn-color-active) active:bg-(--ui-btn-background-active) active:border-(--ui-btn-border-color-active)",
      "disabled:bg-(--ui-btn-background) disabled:border-(--ui-btn-border-color)",
      "aria-disabled:bg-(--ui-btn-background) aria-disabled:border-(--ui-btn-border-color)",
      "focus-visible:outline-(--ui-btn-background)",
      "ring-(--ui-btn-background-hover) focus:outline-none focus-visible:ring-(--ui-btn-background-hover)",
      "h-(--ui-btn-height)",
      "text-(length:--ui-btn-font-size) leading-(--ui-btn-height) font-(--ui-btn-font-weight)",
      ""
    ],
    baseLoading: "h-full w-full absolute inset-0 flex flex-row flex-nowrap items-center justify-center",
    baseLoadingWaitIcon: "text-(--ui-btn-color) size-[calc(var(--ui-btn-wait-icon-size)_+_7px)]",
    baseLoadingClockIcon: "text-(--ui-btn-color) size-[calc(var(--ui-btn-wait-icon-size)_+_7px)]",
    baseLoadingSpinnerIcon: "text-(--ui-btn-color) size-(--ui-btn-wait-icon-size) animate-spin stroke-2",
    baseLine: [
      "w-full inline-flex items-center justify-center gap-(--ui-btn-icon-space)",
      "ps-(--ui-btn-padding-left) pe-(--ui-btn-padding-right)"
    ].join(" "),
    label: [
      "h-(--ui-btn-height) max-w-full",
      "inline-flex items-center",
      "tracking-(--ui-btn-letter-spacing)",
      "overflow-visible",
      "text-clip",
      "mt-(--ui-btn-title-compensation)"
    ].join(" "),
    labelInner: [
      "truncate",
      "inline-block",
      "max-w-full",
      "mt-(--ui-btn-title-compensation)"
    ].join(" "),
    leadingIcon: "text-(--ui-btn-color) shrink-0 size-(--ui-btn-icon-size)",
    leadingAvatar: "shrink-0 me-[4px]",
    leadingAvatarSize: "",
    trailingIcon: "text-(--ui-btn-color) shrink-0 size-(--ui-btn-icon-size) mt-(--ui-btn-title-compensation)",
    safeList: safeList$1
  },
  variants: {
    ...buttonGroupVariant,
    color: {
      "air-primary": "--style-filled",
      "air-primary-success": "--style-filled-success",
      "air-primary-alert": "--style-filled-alert",
      "air-primary-copilot": "--style-filled-copilot",
      "air-secondary": "--style-tinted",
      "air-secondary-alert": "--style-tinted-alert",
      "air-secondary-accent": "--style-outline",
      "air-secondary-accent-1": "--style-outline-accent-1",
      "air-secondary-accent-2": "--style-outline-accent-2",
      "air-secondary-no-accent": "--style-outline-no-accent",
      "air-tertiary": "--style-plain",
      "air-tertiary-accent": "--style-plain-accent",
      "air-tertiary-no-accent": "--style-plain-no-accent",
      "air-selection": "--style-selection",
      "air-boost": {
        base: [
          "--style-filled-boost",
          "bg-transparent hover:bg-transparent focus:bg-transparent active:bg-transparent disabled:bg-transparent aria-disabled:bg-transparent",
          "from-0% via-58.65% to-100%",
          "bg-radial-[110.42%_110.42%_at_-10.42%_31.25%] from-(--ui-color-design-filled-boost-bg-gradient-1) via-(--ui-color-design-filled-boost-bg-gradient-2) to-(--ui-color-design-filled-boost-bg-gradient-3)",
          "hover:bg-radial-[110.42%_110.42%_at_-10.42%_31.25%] hover:from-(--hover-color-1) via-(--hover-color-2) hover:via-58.65% hover:to-(--hover-color-3)",
          "focus:bg-radial-[110.42%_110.42%_at_-10.42%_31.25%] focus:from-(--hover-color-1) via-(--hover-color-2) focus:via-58.65% to-(--hover-color-3)",
          "active:bg-radial-[110.42%_110.42%_at_-10.42%_31.25%] active:from-(--active-color-1) via-(--active-color-2) active:via-58.65% to-(--active-color-3)",
          "disabled:bg-radial-[110.42%_110.42%_at_-10.42%_31.25%] disabled:from-(--ui-color-design-filled-boost-bg-gradient-1) via-(--ui-color-design-filled-boost-bg-gradient-2) to-(--ui-color-design-filled-boost-bg-gradient-3)",
          "aria-disabled:bg-radial-[110.42%_110.42%_at_-10.42%_31.25%] aria-disabled:from-(--ui-color-design-filled-boost-bg-gradient-1) aria-disabled:via-(--ui-color-design-filled-boost-bg-gradient-2) aria-disabled:to-(--ui-color-design-filled-boost-bg-gradient-3)"
          // 'bg-(--ui-btn-background-gradient)',
          // 'hover:bg-(--ui-btn-background-gradient-hover)',
          // 'focus:bg-(--ui-btn-background-gradient-hover)',
          // 'active:bg-(--ui-btn-background-gradient-active)',
          // 'disabled:bg-(--ui-btn-background-gradient)',
          // 'aria-disabled:bg-(--ui-btn-background-gradient)'
        ].join(" ")
      },
      // @deprecate ////
      "default": "",
      "danger": "",
      "success": "",
      "warning": "",
      "primary": "",
      "secondary": "",
      "collab": "",
      "ai": "",
      "link": ""
    },
    // @deprecate ////
    depth: {
      light: "",
      normal: "",
      dark: ""
    },
    size: {
      xss: {
        base: "ui-btn-xss",
        leadingAvatarSize: "2xs"
      },
      xs: {
        base: "ui-btn-xs",
        leadingAvatarSize: "2xs"
      },
      sm: {
        base: "ui-btn-sm",
        leadingAvatarSize: "xs"
      },
      md: {
        base: "ui-btn-md",
        leadingAvatarSize: "xs"
      },
      lg: {
        base: "ui-btn-lg",
        leadingAvatarSize: "md"
      },
      xl: {
        base: "ui-btn-xl",
        leadingAvatarSize: "md"
      }
    },
    block: {
      true: {
        base: "w-full"
      }
    },
    rounded: {
      true: "rounded-(--ui-border-radius-lg)",
      false: "rounded-(--ui-btn-radius)"
    },
    leading: {
      true: ""
    },
    active: {
      true: {
        base: ""
      },
      false: {
        base: ""
      }
    },
    useLabel: {
      true: "",
      false: {
        baseLine: "ps-[4px] pe-[4px]",
        leadingAvatar: "me-0"
      }
    },
    useDropdown: {
      true: ""
    },
    useWait: {
      true: ""
    },
    useClock: {
      true: ""
    },
    loading: {
      true: ""
    },
    normalCase: {
      true: "normal-case",
      false: "uppercase"
    },
    isAir: {
      true: "--air",
      // old theme
      false: ""
    }
  },
  compoundVariants: [
    // region default ////
    {
      color: "default",
      depth: "light",
      class: "--style-default-light"
    },
    {
      color: "default",
      depth: "normal",
      class: "--style-default"
    },
    {
      color: "default",
      depth: "dark",
      class: "--style-default-dark"
    },
    // endregion ////
    // region danger ////
    {
      color: "danger",
      depth: "light",
      class: "--style-danger-light"
    },
    {
      color: "danger",
      depth: "normal",
      class: "--style-danger"
    },
    {
      color: "danger",
      depth: "dark",
      class: "--style-danger-dark"
    },
    // endregion ////
    // region success ////
    {
      color: "success",
      depth: "light",
      class: "--style-success-light"
    },
    {
      color: "success",
      depth: "normal",
      class: "--style-success"
    },
    {
      color: "success",
      depth: "dark",
      class: "--style-success-dark"
    },
    // endregion ////
    // region warning ////
    {
      color: "warning",
      depth: "light",
      class: "--style-warning-light"
    },
    {
      color: "warning",
      depth: "normal",
      class: "--style-warning"
    },
    {
      color: "warning",
      depth: "dark",
      class: "--style-warning-dark"
    },
    // endregion ////
    // region primary ////
    {
      color: "primary",
      depth: "light",
      class: "--style-primary-light"
    },
    {
      color: "primary",
      depth: "normal",
      class: "--style-primary"
    },
    {
      color: "primary",
      depth: "dark",
      class: "--style-primary-dark"
    },
    // endregion ////
    // region secondary ////
    {
      color: "secondary",
      depth: "light",
      class: "--style-secondary-light"
    },
    {
      color: "secondary",
      depth: "normal",
      class: "--style-secondary"
    },
    {
      color: "secondary",
      depth: "dark",
      class: "--style-secondary-dark"
    },
    // endregion ////
    // region collab ////
    {
      color: "collab",
      depth: "light",
      class: "--style-collab-light"
    },
    {
      color: "collab",
      depth: "normal",
      class: "--style-collab"
    },
    {
      color: "collab",
      depth: "dark",
      class: "--style-collab-dark"
    },
    // endregion ////
    // region ai ////
    {
      color: "ai",
      depth: "light",
      class: "--style-ai-light"
    },
    {
      color: "ai",
      depth: "normal",
      class: "--style-ai"
    },
    {
      color: "ai",
      depth: "dark",
      class: "--style-ai-dark"
    },
    // endregion ////
    // region link ////
    {
      color: "link",
      depth: "light",
      class: "--style-link-light"
    },
    {
      color: "link",
      depth: "normal",
      class: "--style-link"
    },
    {
      color: "link",
      depth: "dark",
      class: "--style-link-dark"
    },
    // endregion ////
    {
      leading: true,
      useLabel: true,
      useDropdown: false,
      class: {
        baseLine: "ps-[calc(var(--ui-btn-padding-left)_-_var(--ui-btn-icon-compensation))]"
      }
    },
    // endregion ////
    // region size && useDropdown ////
    // @todo ////
    {
      leading: false,
      useLabel: true,
      useDropdown: true,
      class: {
        baseLine: "pe-[calc(var(--ui-btn-padding-right)_-_var(--ui-btn-icon-compensation))]"
      }
    },
    // endregion ////
    // region size && leading && useDropdown ////
    {
      leading: true,
      useLabel: true,
      useDropdown: true,
      class: {
        baseLine: [
          "ps-[calc(var(--ui-btn-padding-left)_-_var(--ui-btn-icon-compensation))]",
          "pe-[calc(var(--ui-btn-padding-right)_-_var(--ui-btn-icon-compensation))]"
        ].join(" ")
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    size: "md",
    color: "air-secondary-no-accent",
    depth: "normal",
    normalCase: true,
    isAir: true
  }
};

const calendar = {
  slots: {
    root: [
      "font-[family-name:var(--ui-font-family-primary)]",
      "w-full"
    ].join(" "),
    header: "flex items-center justify-between",
    body: "flex flex-col space-y-4 pt-4 sm:flex-row sm:space-x-4 sm:space-y-0",
    heading: [
      "mx-auto",
      "text-center",
      "font-(--ui-font-weight-semibold)",
      "truncate"
    ].join(" "),
    grid: "w-full border-collapse select-none space-y-1 focus:outline-none",
    gridRow: "grid grid-cols-7 place-items-center",
    gridWeekDaysRow: "mb-1 grid w-full grid-cols-7",
    gridBody: "grid",
    headCell: [
      "font-(--ui-font-weight-normal)",
      "text-(--ui-color-design-plain-na-content)"
      // text-base-500
    ].join(" "),
    cell: "relative text-center cursor-pointer aria-disabled:cursor-not-allowed",
    cellTrigger: [
      "m-0.5 relative",
      "flex items-center justify-center",
      "rounded-full whitespace-nowrap",
      "focus-visible:ring-2 focus:outline-none",
      "text-(--b24ui-typography-label-color)",
      "data-disabled:text-(--b24ui-typography-legend-color)",
      "data-unavailable:text-(--b24ui-typography-legend-color)",
      "data-outside-view:text-(--ui-color-design-plain-na-content-secondary)",
      "data-[selected]:text-(--b24ui-color)",
      // data-[selected]:text-white
      "focus-visible:ring-(--b24ui-background-hover)",
      "data-[selected]:bg-(--b24ui-background)",
      "data-today:not-data-[selected]:text-(--b24ui-background)",
      "data-[highlighted]:bg-(--b24ui-background)",
      "data-[highlighted]:text-(--b24ui-color)",
      "hover:not-data-[disabled]:not-data-[selected]:bg-(--b24ui-background)",
      "hover:not-data-[disabled]:not-data-[selected]:text-(--b24ui-color)",
      "data-unavailable:line-through",
      "data-unavailable:pointer-events-none",
      "data-today:font-(--ui-font-weight-semibold)",
      "transition"
    ].join(" ")
  },
  variants: {
    color: {
      "air-primary": { root: "style-filled" },
      "air-primary-success": { root: "style-filled-success" },
      "air-primary-alert": { root: "style-filled-alert" },
      "air-primary-copilot": { root: "style-filled-copilot" },
      "air-primary-warning": { root: "style-filled-warning" },
      // @deprecate ////
      "default": { root: "style-old-default" },
      "danger": { root: "style-old-danger" },
      "success": { root: "style-old-success" },
      "warning": { root: "style-old-warning" },
      "primary": { root: "style-old-primary" },
      "secondary": { root: "style-old-secondary" },
      "collab": { root: "style-old-collab" },
      "ai": { root: "style-old-ai" }
    },
    size: {
      xs: {
        heading: "text-(length:--ui-font-size-md)",
        cell: "text-(length:--ui-font-size-sm)",
        headCell: "text-[10px]",
        cellTrigger: "size-[28px]",
        body: "space-y-2 pt-2"
      },
      sm: {
        heading: "text-(length:--ui-font-size-md)",
        headCell: "text-(length:--ui-font-size-sm)",
        cell: "text-(length:--ui-font-size-sm)",
        cellTrigger: "size-[28px]"
      },
      md: {
        heading: "text-(length:--ui-font-size-lg)",
        headCell: "text-(length:--ui-font-size-md)",
        cell: "text-(length:--ui-font-size-md)",
        cellTrigger: "size-[32px]"
      },
      lg: {
        heading: "text-(length:--ui-font-size-2xl)",
        headCell: "text-(length:--ui-font-size-lg)",
        cell: "text-(length:--ui-font-size-lg)",
        cellTrigger: "size-[36px] text-(length:--ui-font-size-lg)"
      }
    }
  },
  defaultVariants: {
    size: "md",
    color: "air-primary"
  }
};

const checkbox = {
  slots: {
    root: "relative flex items-start",
    base: [
      "cursor-pointer",
      "flex items-center justify-center",
      "shrink-0",
      "rounded-(--ui-border-radius-2xs)",
      "text-(--b24ui-typography-label-color)",
      "ring ring-inset ring-(--ui-color-base-5)",
      "focus-visible:outline-(--b24ui-border-color)",
      "outline-transparent focus-visible:outline-2 focus-visible:outline-offset-2"
    ].join(" "),
    indicator: [
      "flex items-center justify-center",
      "size-full",
      "text-(--b24ui-color)"
    ].join(" "),
    container: "flex items-center",
    icon: "shrink-0 size-full",
    wrapper: [
      "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)"
    ].join(" "),
    label: [
      "cursor-pointer",
      "block",
      "text-(--b24ui-typography-label-color)"
    ].join(" "),
    description: "text-(--b24ui-typography-description-color)"
  },
  variants: {
    color: {
      "air-primary": { root: "style-filled" },
      "air-primary-success": { root: "style-filled-success" },
      "air-primary-alert": { root: "style-filled-alert" },
      "air-primary-copilot": { root: "style-filled-copilot" },
      "air-primary-warning": { root: "style-filled-warning" },
      // @deprecate ////
      "default": { root: "style-old-default" },
      "danger": { root: "style-old-danger" },
      "success": { root: "style-old-success" },
      "warning": { root: "style-old-warning" },
      "primary": { root: "style-old-primary" },
      "secondary": { root: "style-old-secondary" },
      "collab": { root: "style-old-collab" },
      "ai": { root: "style-old-ai" }
    },
    variant: {
      list: {
        root: ""
      },
      card: {
        // @memo style-outline-no-accent
        root: "border border-(--ui-color-design-outline-na-stroke) bg-(--ui-color-design-outline-na-bg)"
      }
    },
    indicator: {
      start: {
        root: "flex-row",
        wrapper: "ms-2"
      },
      end: {
        root: "flex-row-reverse",
        wrapper: "me-2"
      },
      hidden: {
        base: "sr-only",
        wrapper: "text-center"
      }
    },
    size: {
      xs: {
        base: "size-[12px]",
        container: "h-[12px]",
        wrapper: "text-(length:--ui-font-size-xs)",
        label: "leading-[11px]"
      },
      sm: {
        base: "size-[14px]",
        container: "h-[14px]",
        wrapper: "text-(length:--ui-font-size-sm)",
        label: "leading-[14px]"
      },
      md: {
        base: "size-[16px]",
        container: "h-[16px]",
        wrapper: "text-(length:--ui-font-size-lg)",
        label: "leading-[15px]"
      },
      lg: {
        base: "size-[20px]",
        container: "h-[20px]",
        wrapper: "text-(length:--ui-font-size-xl)",
        label: "leading-[18px]"
      }
    },
    required: {
      true: {
        label: "after:content-['*'] after:ms-0.5 after:text-(--ui-color-accent-main-alert)"
      }
    },
    disabled: {
      true: {
        base: "cursor-not-allowed opacity-30",
        label: "cursor-not-allowed opacity-30",
        description: "cursor-not-allowed opacity-30"
      }
    },
    checked: {
      true: {
        base: "ring-1 ring-(--b24ui-background) bg-(--b24ui-background)"
      }
    }
  },
  compoundVariants: [
    // region card.size ////
    { size: "xs", variant: "card", class: { root: "px-[13px] py-[7px] rounded-(--ui-border-radius-xs)" } },
    // (28 - (1 + 1) - 12) / 2 = 7
    { size: "sm", variant: "card", class: { root: "px-[13px] py-[9px] rounded-(--ui-border-radius-sm)" } },
    // (34 - (1 + 1) - 14) / 2 = 9
    { size: "md", variant: "card", class: { root: "px-[17px] py-[10px] rounded-(--ui-border-radius-md)" } },
    // (38 - (1 + 1) - 16) / 2 = 10
    { size: "lg", variant: "card", class: { root: "px-[23px] py-[12px] rounded-(--ui-border-radius-md)" } },
    // (46 - (1 + 1) - 20) / 2 = 12
    // endregion ////
    // region card.border ////
    {
      variant: "card",
      checked: true,
      class: {
        root: "border-(--b24ui-border-color) cursor-pointer"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-primary",
    size: "md",
    variant: "list",
    indicator: "start"
  }
};

const chip = {
  slots: {
    root: [
      "relative shrink-0 isolate",
      "inline-flex items-center justify-center"
    ].join(" "),
    base: [
      "ui-counter__scope --air",
      "font-[family-name:var(--ui-font-family-primary)]",
      "font-(--ui-font-weight-medium)",
      "select-none",
      "relative",
      "min-w-(--ui-counter-size) h-(--ui-counter-size)",
      "py-0 px-(--ui-counter-inline-space)",
      "inline-flex items-center justify-center",
      "bg-(--b24ui-background)",
      "rounded-(--ui-counter-current-size)",
      "ring-(length:--b24ui-border-width) ring-(--b24ui-border-color)",
      "text-center align-middle",
      "text-(length:--ui-counter-font-size) text-(--b24ui-color)",
      "leading-(--ui-counter-current-size)",
      "overflow-hidden",
      "z-1",
      "text-nowrap"
      // truncate
    ].join(" "),
    trailingIcon: [
      "size-(--ui-counter-size)",
      "text-inherit",
      "text-(length:--ui-counter-symbol-font-size)",
      "opacity-96",
      "tracking-(--ui-letter-spacing-xl)",
      "me-(--ui-counter-symbol-compensation) empty:me-[0]"
    ].join(" ")
  },
  variants: {
    color: {
      "air-primary": { base: "style-filled" },
      "air-primary-success": { base: "style-filled-success" },
      "air-primary-alert": { base: "style-filled-alert" },
      "air-primary-copilot": { base: "style-filled-copilot" },
      "air-primary-warning": { base: "style-filled-warning" },
      "air-secondary": { base: "style-tinted-no-accent-1" },
      "air-secondary-accent": { base: "style-filled-no-accent" },
      "air-secondary-accent-1": { base: "style-filled-no-accent-inverted edge-dark:text-(--ui-color-g-content-grey-2)" },
      "air-tertiary": { base: "style-outline-no-accent" },
      // @deprecate ////
      "default": { base: "style-old-default" },
      "danger": { base: "style-old-danger" },
      "success": { base: "style-old-success" },
      "warning": { base: "style-old-warning" },
      "primary": { base: "style-old-primary" },
      "secondary": { base: "style-old-secondary" },
      "collab": { base: "style-old-collab" },
      "ai": { base: "style-old-ai" }
    },
    size: {
      sm: "ui-counter-sm font-(--ui-font-weight-regular)",
      md: "ui-counter-md",
      lg: "ui-counter-lg"
    },
    position: {
      "top-right": "top-0 right-0",
      "bottom-right": "bottom-0 right-0",
      "top-left": "top-0 left-0",
      "bottom-left": "bottom-0 left-0"
    },
    inverted: {
      true: "",
      false: ""
    },
    inset: {
      false: ""
    },
    standalone: {
      true: "",
      false: "absolute"
    },
    hideZero: {
      true: {
        base: "data-[value=0]:hidden"
      }
    },
    oneDigit: {
      true: {
        base: "px-0"
      }
    }
  },
  compoundVariants: [
    // region position ////
    // not inset ////
    {
      position: "top-right",
      inset: false,
      standalone: false,
      class: "-translate-y-1/2 translate-x-1/2 transform"
    },
    {
      position: "bottom-right",
      inset: false,
      standalone: false,
      class: "translate-y-1/2 translate-x-1/2 transform"
    },
    {
      position: "top-left",
      inset: false,
      standalone: false,
      class: "-translate-y-1/2 -translate-x-1/2 transform"
    },
    {
      position: "bottom-left",
      inset: false,
      standalone: false,
      class: "translate-y-1/2 -translate-x-1/2 transform"
    },
    // inset ////
    {
      position: "top-right",
      size: ["2xs", "xs", "sm", "md", "lg", "xl", "2xl"],
      inset: true,
      standalone: false,
      class: "-translate-y-1/3 translate-x-1/3 transform"
    },
    {
      position: "top-right",
      size: ["2xs", "3xl"],
      inset: true,
      standalone: false,
      class: "translate-y-0 translate-x-0 transform"
    },
    {
      position: "bottom-right",
      size: ["2xs", "xs", "sm", "md", "lg", "xl", "2xl"],
      inset: true,
      standalone: false,
      class: "translate-y-1/3 translate-x-1/3 transform"
    },
    {
      position: "bottom-right",
      size: ["2xs", "3xl"],
      inset: true,
      standalone: false,
      class: "translate-y-0 translate-x-0 transform"
    },
    {
      position: "top-left",
      size: ["2xs", "xs", "sm", "md", "lg", "xl", "2xl"],
      inset: true,
      standalone: false,
      class: "-translate-y-1/3 -translate-x-1/3 transform"
    },
    {
      position: "top-left",
      size: ["2xs", "3xl"],
      inset: true,
      standalone: false,
      class: "-translate-y-0 -translate-x-0 transform"
    },
    {
      position: "bottom-left",
      size: ["2xs", "xs", "sm", "md", "lg", "xl", "2xl"],
      inset: true,
      standalone: false,
      class: "translate-y-1/3 -translate-x-1/3 transform"
    },
    {
      position: "bottom-left",
      size: ["2xs", "3xl"],
      inset: true,
      standalone: false,
      class: "translate-y-0 -translate-x-0 transform"
    },
    // endregion ////
    // region color ////
    {
      inverted: true,
      color: "air-primary",
      class: {
        base: "style-filled-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-success",
      class: {
        base: "style-filled-success-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-alert",
      class: {
        base: "style-filled-alert-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-copilot",
      class: {
        base: "style-filled-copilot-inverted"
      }
    },
    {
      inverted: true,
      color: "air-primary-warning",
      class: {
        base: "style-filled-warning-inverted"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    size: "sm",
    color: "air-primary-alert",
    position: "top-right",
    inverted: false
  }
};

const collapsible = {
  slots: {
    root: "",
    content: [
      "motion-safe:data-[state=open]:animate-[collapsible-down_200ms_ease-out] motion-safe:data-[state=closed]:animate-[collapsible-up_200ms_ease-out]",
      "overflow-hidden"
    ].join(" ")
  }
};

const container = {
  base: "w-full max-w-[1280px] mx-auto px-[22px]"
  // max-w-7xl w-full
};

const countdown = {
  slots: {
    base: [
      "relative flex flex-row flex-nowrap items-center justify-between",
      "text-(--b24ui-typography-legend-color)"
    ].join(" "),
    label: "",
    leadingIcon: "shrink-0",
    leadingAvatar: "shrink-0",
    leadingAvatarSize: "",
    circleBase: "-scale-x-100 absolute inset-x-0 inset-y-0",
    circleGroup: "fill-none stroke-none",
    circleElement: "stroke-transparent stroke-1",
    circlePath: [
      "stroke-[7px] rotate-90 origin-center stroke-current transition-all duration-1000 ease-linear"
    ].join(" ")
  },
  variants: {
    size: {
      xs: {
        base: "gap-0.5 text-(length:--ui-font-size-5xs)/(--ui-font-line-height-reset)",
        leadingIcon: "size-sm",
        leadingAvatarSize: "3xs"
      },
      sm: {
        base: "gap-1 text-(length:--ui-font-size-4xs)/(--ui-font-line-height-reset)",
        leadingIcon: "size-sm2",
        leadingAvatarSize: "3xs"
      },
      md: {
        base: "gap-1 text-(length:--ui-font-size-md)/(--ui-font-line-height-reset)",
        leadingIcon: "size-[16px]",
        leadingAvatarSize: "3xs"
      },
      lg: {
        base: "gap-1 text-(length:--ui-font-size-lg)/(--ui-font-line-height-reset)",
        leadingIcon: "size-[22px]",
        leadingAvatarSize: "2xs"
      },
      xl: {
        base: "gap-1 text-(length:--ui-font-size-xl)/(--ui-font-line-height-reset)",
        leadingIcon: "size-[26px]",
        leadingAvatarSize: "xs"
      }
    },
    leading: {
      true: ""
    },
    useCircle: {
      true: {
        base: "justify-center",
        circleBase: "size-full"
      }
    }
  },
  compoundVariants: [
    {
      size: "xs",
      useCircle: true,
      class: "text-(length:--ui-font-size-7xs)/[normal] p-0.5"
    },
    {
      size: "sm",
      useCircle: true,
      class: "text-(length:--ui-font-size-6xs)/[normal] p-1.5"
    },
    {
      size: "md",
      useCircle: true,
      class: "text-(length:--ui-font-size-3xs)/[normal] p-1.5"
    },
    {
      size: "lg",
      useCircle: true,
      class: "text-(length:--ui-font-size-xs)/[normal] p-1.5 pb-2"
    },
    {
      size: "xl",
      useCircle: true,
      class: "text-(length:--ui-font-size-sm)/[normal] p-2 pb-2.5"
    }
  ],
  defaultVariants: {
    size: "md"
  }
};

const descriptionList = {
  slots: {
    root: "w-full shrink-0",
    legend: [
      "font-(--ui-font-weight-semibold)",
      "text-(--b24ui-typography-label-color)"
    ].join(" "),
    text: "text-(--b24ui-typography-description-color)",
    container: "grid grid-cols-1 sm:grid-cols-[min(50%,theme(spacing.80))_auto]",
    labelWrapper: [
      "col-start-1 border-t first:border-none sm:border-t",
      "flex flex-nowrap flex-row items-center justify-start gap-1.5",
      "border-(--ui-color-divider-vibrant-default) sm:border-(--ui-color-divider-vibrant-default)",
      "text-(--b24ui-typography-description-color)"
    ].join(" "),
    icon: [
      "shrink-0",
      "size-6",
      "text-(--b24ui-typography-description-color)"
    ].join(" "),
    avatar: "shrink-0",
    avatarSize: "",
    label: "",
    descriptionWrapper: [
      "sm:border-t sm:[&:nth-child(2)]:border-none",
      "sm:border-(--ui-color-divider-vibrant-default)",
      "text-(--b24ui-typography-label-color)"
    ].join(" "),
    description: "",
    actions: "flex flex-wrap gap-1.5 shrink-0",
    footer: [
      "border-t",
      "border-(--ui-color-divider-vibrant-default)"
    ].join(" ")
  },
  variants: {
    size: {
      sm: {
        legend: "text-md",
        text: "mt-1 max-w-2/3 text-sm",
        container: "mt-2.5 text-md",
        labelWrapper: "pt-3 sm:py-3",
        avatarSize: "xs",
        label: "",
        descriptionWrapper: "pb-3 pt-1 sm:py-3",
        description: "",
        footer: "mt-2 p-2"
      },
      md: {
        legend: "text-xl",
        text: "mt-2 max-w-2/3 text-lg leading-5",
        container: "mt-3 text-lg",
        labelWrapper: "pt-3 sm:py-3",
        avatarSize: "xs",
        label: "",
        descriptionWrapper: "pb-3 pt-1 sm:py-3",
        description: "",
        footer: "mt-4 p-4"
      }
    },
    orientation: {
      horizontal: {
        descriptionWrapper: "w-full flex flex-row items-center justify-between gap-4",
        actions: "items-center"
      },
      vertical: {
        descriptionWrapper: "",
        actions: "items-start mt-2.5"
      }
    },
    title: {
      true: {
        description: "mt-1"
      }
    }
  },
  compoundVariants: [],
  defaultVariants: {
    size: "md"
  }
};

const dropdownMenu = {
  slots: {
    content: [
      "light",
      "bg-(--popup-window-background-color)",
      "shadow-(--popup-window-box-shadow)",
      "rounded-(--popup-window-border-radius) will-change-[opacity]",
      "motion-safe:data-[state=open]:animate-[scale-in_100ms_ease-out] motion-safe:data-[state=closed]:animate-[scale-out_100ms_ease-in]",
      "origin-(--reka-dropdown-menu-content-transform-origin)",
      "font-[family-name:var(--ui-font-family-primary)]",
      "relative",
      "isolate",
      "px-0 py-(--menu-popup-padding)",
      "pointer-events-auto"
    ].join(" "),
    viewport: [
      "relative",
      "w-[240px] max-h-[40vh]",
      "overflow-x-hidden overflow-y-auto scrollbar-thin"
      // scrollbar-transparent
    ].join(" "),
    arrow: "fill-(--popup-window-background-color)",
    // for content bottom|top::start -> ml-[12px]
    group: "grid",
    // p-1 isolate
    label: [
      "w-full min-w-[195px] h-(--popup-window-delimiter-section-height)",
      "px-[18px] mt-(--menu-item-block-stack-space)",
      "flex flex-row rtl:flex-row-reverse items-center",
      "select-none outline-none whitespace-nowrap",
      "text-start",
      "text-(length:--popup-window-delimiter-font-size)",
      "text-(--popup-window-delimiter-text-color)",
      "font-(--popup-window-delimiter-font-weight)",
      "after:ms-[10px] after:block after:flex-1 after:min-w-[15px] after:h-px after:bg-(--popup-window-delimiter-bg-color)"
    ].join(" "),
    separator: "my-[8px] mx-[18px] h-[1px] bg-(--popup-window-delimiter-bg-color)",
    item: [
      "group",
      "w-full min-w-[195px] h-[36px]",
      "px-[18px] mt-(--menu-item-block-stack-space)",
      "relative",
      "flex flex-row rtl:flex-row-reverse items-center",
      "select-none outline-none whitespace-nowrap",
      "cursor-pointer",
      "data-disabled:cursor-not-allowed data-disabled:opacity-30",
      "text-start",
      "text-(length:--menu-popup-item-font-size)",
      "text-(--menu-popup-item-color) hover:text-(--menu-popup-item-color-hover)",
      "data-highlighted:text-(--menu-popup-item-color-hover)",
      "data-[state=open]:text-(--menu-popup-item-color-hover)",
      "hover:bg-(--menu-popup-item-bg-color-hover)",
      "data-highlighted:bg-(--menu-popup-item-bg-color-hover)",
      "data-[state=open]:bg-(--menu-popup-item-bg-color-hover)",
      "transition-colors"
    ].join(" "),
    itemLeadingIcon: [
      "shrink-0",
      "size-[18px]",
      "text-(--ui-color-design-plain-content-icon-secondary)",
      "group-data-highlighted:text-(--ui-color-accent-main-primary)",
      "group-data-[state=open]:text-(--ui-color-accent-main-primary)",
      "group-data-[state=checked]:text-(--ui-color-accent-main-primary)",
      "transition-colors"
    ].join(" "),
    itemLeadingAvatar: "shrink-0 size-[16px] mx-px",
    // @memo 18-2px
    itemLeadingAvatarSize: "2xs",
    // @memo this wrong
    itemTrailing: "ml-auto rtl:ml-0 rtl:mr-auto inline-flex gap-1.5 items-center",
    itemTrailingIcon: "shrink-0 size-[24px] text-(--ui-color-design-plain-content-icon-secondary)",
    itemTrailingKbds: "shrink-0 hidden lg:inline-flex items-center gap-0.5",
    itemTrailingKbdsSize: "md",
    itemLabel: [
      "truncate ms-[2px] -mt-px",
      "group-data-[state=checked]:text-(--ui-color-accent-main-primary)"
    ].join(" "),
    itemLabelExternalIcon: "inline-block size-[16px] text-(--ui-color-design-plain-content-icon-secondary)"
  },
  variants: {
    color: {
      "air-primary": { item: "style-filled" },
      "air-primary-success": { item: "style-filled-success" },
      "air-primary-alert": { item: "style-filled-alert" },
      "air-primary-copilot": { item: "style-filled-copilot" },
      "air-primary-warning": { item: "style-filled-warning" },
      // @deprecate ////
      "default": { item: "style-old-default" },
      "danger": { item: "style-old-danger" },
      "success": { item: "style-old-success" },
      "warning": { item: "style-old-warning" },
      "primary": { item: "style-old-primary" },
      "secondary": { item: "style-old-secondary" },
      "collab": { item: "style-old-collab" },
      "ai": { item: "style-old-ai" }
    },
    active: {
      true: {
        item: [
          "text-(--ui-color-accent-main-primary)",
          "hover:text-(--ui-color-accent-main-primary)"
        ].join(" "),
        itemLeadingIcon: [
          "text-(--ui-color-accent-main-primary)",
          "hover:text-(--ui-color-accent-main-primary)",
          "group-data-[state=open]:text-(--ui-color-accent-main-primary)"
        ].join(" ")
      },
      false: {}
    },
    loading: {
      true: {
        itemLeadingIcon: "animate-spin"
      }
    }
  },
  compoundVariants: [
    {
      color: ["air-primary", "air-primary-success", "air-primary-alert", "air-primary-copilot", "air-primary-warning"],
      active: false,
      class: {
        item: [
          "text-(--b24ui-background) data-highlighted:text-(--b24ui-background-hover) data-[state=open]:text-(--b24ui-background-hover)"
        ].join(" "),
        itemLeadingIcon: [
          "text-(--b24ui-background) group-data-highlighted:text-(--b24ui-background-hover) group-data-[state=open]:text-(--b24ui-background-hover)"
        ].join(" ")
      }
    },
    {
      color: ["air-primary", "air-primary-success", "air-primary-alert", "air-primary-copilot", "air-primary-warning"],
      active: true,
      class: {
        item: [
          "text-(--b24ui-background-active)"
        ].join(" "),
        itemLeadingIcon: [
          "text-(--b24ui-background-active) group-data-[state=open]:text-(--b24ui-background-active)"
        ].join(" ")
      }
    },
    {
      color: "default",
      active: false,
      class: ""
    },
    {
      color: "default",
      active: true,
      class: ""
    },
    {
      color: ["danger", "success", "warning", "primary", "secondary", "collab", "ai"],
      active: false,
      class: {
        item: [
          "text-(--b24ui-background-active) data-highlighted:text-(--b24ui-background-hover) data-[state=open]:text-(--b24ui-background-hover)"
        ].join(" "),
        itemLeadingIcon: [
          "text-(--b24ui-icon) group-data-highlighted:text-(--b24ui-icon) group-data-[state=open]:text-(--b24ui-icon)"
        ].join(" ")
      }
    },
    {
      color: ["danger", "success", "warning", "primary", "secondary", "collab", "ai"],
      active: true,
      class: {
        item: [
          "text-(--b24ui-background-active)"
        ].join(" "),
        itemLeadingIcon: [
          "text-(--b24ui-background-active) group-data-[state=open]:text-(--b24ui-background-active)"
        ].join(" ")
      }
    }
  ],
  defaultVariants: {}
};

const form = {
  base: ""
};

const formField = {
  slots: {
    root: "font-[family-name:var(--ui-font-family-system)] font-(--ui-font-weight-regular)",
    wrapper: "leading-(--ui-font-line-height-reset)",
    labelWrapper: "flex content-center items-center justify-between",
    label: [
      "block",
      "text-(--b24ui-typography-label-color)"
    ].join(" "),
    hint: "text-(--b24ui-typography-description-color)",
    container: "relative",
    description: "mt-[2px] leading-(--ui-font-line-height-2xs) text-(--b24ui-typography-description-color)",
    error: "mt-[4px] text-(--ui-color-accent-main-alert)",
    errorWrapper: "flex flex-row flex-nowrap items-center gap-0.5",
    errorIcon: "size-[18px] mt-[2px]",
    help: "mt-[6px] leading-(--ui-font-line-height-2xs) italic text-(--b24ui-typography-description-color)"
  },
  variants: {
    useDescription: {
      true: { wrapper: "mb-[6px]" },
      false: { wrapper: "mb-[10px]" }
    },
    size: {
      // xss: { root: 'text-(length:--ui-font-size-3xs)', errorIcon: 'size-[16px]' },
      xs: { root: "text-(length:--ui-font-size-xs)", errorIcon: "size-[16px]" },
      sm: { root: "text-(length:--ui-font-size-xs)", errorIcon: "size-[16px]" },
      md: { root: "text-(length:--ui-font-size-sm)", errorIcon: "size-[18px]" },
      lg: { root: "text-(length:--ui-font-size-md)" }
      // xl: { root: 'text-(length:--ui-font-size-md)' }
    },
    required: {
      true: {
        label: "after:content-['*'] after:ms-0.5 after:text-(--ui-color-accent-main-alert)"
      }
    }
  },
  compoundVariants: [],
  defaultVariants: {
    size: "md"
  }
};

const input = {
  slots: {
    root: "isolate relative inline-flex items-center w-full",
    // @todo remove ? w-full -> see select
    base: [
      "px-3",
      "w-full py-0 border-0 focus:outline-none",
      "disabled:cursor-not-allowed",
      "disabled:pointer-events-none",
      "disabled:opacity-30",
      "disabled:resize-none",
      "appearance-none transition duration-300 ease-linear",
      // transition-colors
      "ring ring-inset",
      "ring-(--ui-color-design-outline-stroke)",
      "text-(--ui-color-base-1)",
      "style-blurred-bg-input",
      "placeholder:text-(--ui-color-design-plain-na-content-secondary)",
      "hover:text-(--ui-color-base-1)",
      "focus:text-(--ui-color-base-1)",
      "active:text-(--ui-color-base-1)",
      "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
      "align-middle",
      "text-ellipsis whitespace-nowrap",
      "focus-visible:ring-1",
      "focus-visible:ring-inset",
      "focus-visible:ring-(--b24ui-border-color)"
    ].join(" "),
    leading: "absolute inset-y-0 start-0 flex items-center",
    leadingIcon: [
      "shrink-0",
      "text-(--b24ui-icon-color)"
    ].join(" "),
    leadingAvatar: "shrink-0",
    leadingAvatarSize: "",
    trailing: "absolute inset-y-0 end-0 flex items-center",
    trailingIcon: [
      "shrink-0",
      "text-(--b24ui-icon-color)"
    ].join(" "),
    tag: [
      "pointer-events-none select-none",
      "absolute",
      "z-10",
      "-top-[7px] right-[14px]",
      "flex flex-col justify-center items-center"
    ].join(" ")
  },
  variants: {
    ...buttonGroupVariantWithRoot,
    size: {
      xss: {
        base: "h-[20px] gap-1 text-(length:--ui-font-size-4xs)/[normal]",
        leading: "px-1",
        trailing: "px-1",
        leadingIcon: "size-[12px]",
        leadingAvatarSize: "3xs",
        trailingIcon: "size-[12px]"
      },
      xs: {
        base: "h-[24px] gap-1 text-(length:--ui-font-size-xs)/[normal]",
        leading: "px-1",
        trailing: "px-1",
        leadingIcon: "size-[14px]",
        leadingAvatarSize: "3xs",
        trailingIcon: "size-[14px]"
      },
      sm: {
        base: "h-[28px] gap-1.5 text-(length:--ui-font-size-sm)/[normal]",
        leading: "px-1.5",
        trailing: "px-1.5",
        leadingIcon: "size-[16px]",
        leadingAvatar: "size-[16px]",
        leadingAvatarSize: "2xs",
        trailingIcon: "size-[16px]"
      },
      md: {
        base: "h-[34px] gap-1.5 text-(length:--ui-font-size-lg)/[normal]",
        leading: "px-2",
        trailing: "px-2",
        leadingIcon: "size-[18px]",
        leadingAvatarSize: "2xs",
        trailingIcon: "size-[18px]"
      },
      lg: {
        base: "h-[38px] gap-2 text-(length:--ui-font-size-lg)/[normal]",
        leading: "px-2",
        trailing: "px-2",
        leadingIcon: "size-[22px]",
        leadingAvatarSize: "2xs",
        trailingIcon: "size-[22px]"
      },
      xl: {
        base: "h-[46px] gap-2 text-(length:--ui-font-size-2xl)/[normal]",
        leading: "px-2",
        trailing: "px-2",
        leadingIcon: "size-[22px]",
        leadingAvatarSize: "2xs",
        trailingIcon: "size-[22px]"
      }
    },
    color: {
      "air-primary": { base: "style-filled" },
      "air-primary-success": { base: "style-filled-success" },
      "air-primary-alert": { base: "style-filled-alert" },
      "air-primary-copilot": { base: "style-filled-copilot" },
      "air-primary-warning": { base: "style-filled-warning" },
      // @deprecate ////
      "default": { base: "style-old-default" },
      "danger": { base: "style-old-danger" },
      "success": { base: "style-old-success" },
      "warning": { base: "style-old-warning" },
      "primary": { base: "style-old-primary" },
      "secondary": { base: "style-old-secondary" },
      "collab": { base: "style-old-collab" },
      "ai": { base: "style-old-ai" }
    },
    /**
     * @memo now get from Badge
     */
    // tagColor: {},
    rounded: {
      true: "rounded-(--ui-border-radius-3xl)",
      false: "rounded-(--ui-border-radius-sm)"
    },
    noPadding: { true: { base: "px-0" } },
    noBorder: { true: "ring-0 focus-visible:ring-0 style-transparent-bg" },
    underline: {
      true: [
        "ring-0 focus-visible:ring-0 style-transparent-bg",
        "border-b-1",
        "border-b-(--ui-color-design-outline-stroke)",
        "rounded-none"
      ].join(" ")
    },
    leading: { true: "" },
    trailing: { true: "" },
    loading: { true: "" },
    highlight: { true: "ring ring-inset ring-(--b24ui-border-color)" },
    type: {
      file: [
        "file:me-1.5",
        "file:text-(--ui-color-design-plain-na-content-secondary)",
        "file:outline-none"
      ]
    }
  },
  compoundVariants: [
    // region ring for focus and highlight ////
    // region color ////
    {
      noBorder: false,
      underline: false,
      class: ""
    },
    {
      highlight: true,
      noBorder: false,
      underline: false,
      class: "ring ring-inset ring-(--b24ui-border-color)"
    },
    {
      noBorder: false,
      underline: true,
      class: "focus-visible:border-(--b24ui-border-color)"
    },
    {
      highlight: true,
      noBorder: false,
      underline: true,
      class: "border-b-(--b24ui-border-color)"
    },
    // endregion ////
    // endregion ////
    // region type file ////
    {
      type: "file",
      size: "xss",
      class: "py-[2px]"
    },
    {
      type: "file",
      size: "xs",
      class: "py-[4px]"
    },
    {
      type: "file",
      size: "sm",
      class: "py-[5px]"
    },
    {
      type: "file",
      size: "md",
      class: "py-[7px]"
    },
    {
      type: "file",
      size: "lg",
      class: "py-[9px]"
    },
    {
      type: "file",
      size: "xl",
      class: "py-[11px]"
    },
    // endregion ////
    // region leading ////
    {
      leading: true,
      noPadding: false,
      size: "xss",
      class: "ps-[20px]"
    },
    {
      leading: true,
      noPadding: false,
      size: "xs",
      class: "ps-[22px]"
    },
    {
      leading: true,
      noPadding: false,
      size: "sm",
      class: "ps-[28px]"
    },
    {
      leading: true,
      noPadding: false,
      size: "md",
      class: "ps-[34px]"
    },
    {
      leading: true,
      noPadding: false,
      size: "lg",
      class: "ps-[38px]"
    },
    {
      leading: true,
      noPadding: false,
      size: "xl",
      class: "ps-[38px]"
    },
    // endregion ////
    // region trailing ////
    {
      trailing: true,
      noPadding: false,
      size: "xss",
      class: "pe-[20px]"
    },
    {
      trailing: true,
      noPadding: false,
      size: "xs",
      class: "pe-[22px]"
    },
    {
      trailing: true,
      noPadding: false,
      size: "sm",
      class: "pe-[28px]"
    },
    {
      trailing: true,
      noPadding: false,
      size: "md",
      class: "pe-[34px]"
    },
    {
      trailing: true,
      noPadding: false,
      size: "lg",
      class: "pe-[38px]"
    },
    {
      trailing: true,
      noPadding: false,
      size: "xl",
      class: "pe-[38px]"
    },
    // endregion ////
    // region loading ////
    {
      loading: true,
      leading: true,
      class: {
        leadingIcon: "size-[21px]"
      }
    },
    {
      loading: true,
      leading: false,
      trailing: true,
      class: {
        trailingIcon: "size-[21px]"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-primary",
    size: "md"
  }
};

const inputMenu = () => {
  return defuFn({
    slots: {
      base: [
        "w-full py-0 border-0 focus:outline-none",
        "disabled:cursor-not-allowed",
        "disabled:pointer-events-none",
        "disabled:opacity-30",
        "disabled:resize-none",
        "appearance-none transition duration-300 ease-linear transition-colors",
        // transition-colors
        "ring ring-inset",
        "ring-(--ui-color-design-outline-stroke)",
        "text-(--ui-color-base-1)",
        "style-blurred-bg-input",
        "hover:text-(--ui-color-base-1)",
        "focus:text-(--ui-color-base-1)",
        "active:text-(--ui-color-base-1)",
        "placeholder:text-(--ui-color-design-plain-na-content-secondary)",
        "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
        "align-middle",
        "focus-visible:ring-1",
        "focus-visible:ring-inset",
        "focus-visible:ring-(--b24ui-border-color)"
      ].join(" "),
      trailing: [
        "group",
        "absolute inset-y-0 end-0 flex items-center",
        "disabled:cursor-not-allowed",
        "disabled:opacity-30"
      ].join(" "),
      content: [
        "light",
        // w-(--reka-combobox-trigger-width)
        "bg-(--popup-window-background-color)",
        "shadow-(--popup-window-box-shadow)",
        "rounded-(--popup-window-border-radius) will-change-[opacity]",
        "motion-safe:data-[state=open]:animate-[scale-in_100ms_ease-out] motion-safe:data-[state=closed]:animate-[scale-out_100ms_ease-in]",
        "origin-(--reka-dropdown-menu-content-transform-origin)",
        "font-[family-name:var(--ui-font-family-primary)]",
        "relative",
        "isolate",
        "px-0 py-(--menu-popup-padding)",
        "pointer-events-auto"
      ].join(" "),
      viewport: [
        "relative",
        "w-[240px] max-h-[40vh]",
        "overflow-x-hidden overflow-y-auto scrollbar-thin"
        // scrollbar-transparent
      ].join(" "),
      arrow: "fill-(--popup-window-background-color)",
      // for content bottom|top::start -> ml-[12px]
      group: "grid",
      empty: [
        "h-(--popup-window-delimiter-section-height)",
        // 'mt-(--menu-item-block-stack-space)',
        "py-[8px]",
        "select-none outline-none whitespace-nowrap",
        "text-center",
        "text-(length:--popup-window-delimiter-font-size)/(--ui-font-line-height-lg)",
        "text-(--popup-window-delimiter-text-color)",
        "font-(--popup-window-delimiter-font-weight)"
      ].join(" "),
      label: [
        "w-full min-w-[195px] h-(--popup-window-delimiter-section-height)",
        "px-[18px] mt-(--menu-item-block-stack-space)",
        "flex flex-row rtl:flex-row-reverse items-center",
        "select-none outline-none whitespace-nowrap",
        "text-start",
        "text-(length:--popup-window-delimiter-font-size)",
        "text-(--popup-window-delimiter-text-color)",
        "font-(--popup-window-delimiter-font-weight)",
        "after:ms-[10px] after:block after:flex-1 after:min-w-[15px] after:h-px after:bg-(--popup-window-delimiter-bg-color)"
      ].join(" "),
      separator: "my-[8px] mx-[18px] h-[1px] bg-(--popup-window-delimiter-bg-color)",
      item: [
        "group",
        "w-full min-w-[195px] h-[36px]",
        "px-[18px] mt-(--menu-item-block-stack-space)",
        "relative",
        "flex flex-row rtl:flex-row-reverse items-center",
        "select-none outline-none whitespace-nowrap",
        "cursor-pointer",
        "data-disabled:cursor-not-allowed data-disabled:opacity-30",
        "text-start",
        "text-(length:--menu-popup-item-font-size)",
        "text-(--menu-popup-item-color) hover:text-(--menu-popup-item-color-hover)",
        "data-highlighted:not-data-disabled:text-(--menu-popup-item-color-hover)",
        "data-[state=open]:text-(--menu-popup-item-color-hover)",
        "data-[state=checked]:text-(--menu-popup-item-color-hover)",
        "hover:bg-(--menu-popup-item-bg-color-hover)",
        "data-highlighted:bg-(--menu-popup-item-bg-color-hover)",
        "data-[state=open]:bg-(--menu-popup-item-bg-color-hover)",
        "transition-colors"
      ].join(" "),
      itemLeadingIcon: [
        "shrink-0",
        "size-[18px]",
        "text-(--ui-color-design-plain-content-icon-secondary)",
        "group-data-highlighted:not-data-disabled:text-(--ui-color-accent-main-primary)",
        "group-data-[state=open]:text-(--ui-color-accent-main-primary)",
        "group-data-[state=checked]:text-(--ui-color-accent-main-primary)",
        "transition-colors"
      ].join(" "),
      itemLeadingAvatar: "shrink-0 size-[16px] mx-px",
      // @memo 18-2px
      itemLeadingAvatarSize: "2xs",
      // @memo this wrong
      itemLeadingChip: "shrink-0 size-[16px] mx-px",
      itemLeadingChipSize: "sm",
      itemTrailing: "ml-auto rtl:ml-0 rtl:mr-auto inline-flex gap-1.5 items-center",
      itemTrailingIcon: "shrink-0 size-[24px] text-(--ui-color-design-plain-content-icon-secondary)",
      itemLabel: [
        "truncate ms-[2px] -mt-px",
        "group-data-[state=checked]:text-(--ui-color-accent-main-primary)"
      ].join(" "),
      tagsItem: [
        "ps-[13px] pe-[6px]",
        "rounded-(--ui-border-radius-xs)",
        "h-(--main-ui-square-item-height)",
        "leading-(--main-ui-square-item-height)",
        "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
        "inline-flex items-center gap-1",
        "data-disabled:cursor-not-allowed data-disabled:opacity-30",
        "text-(--ui-color-design-tinted-content)",
        "bg-(--ui-color-design-tinted-bg-alt)"
      ].join(" "),
      tagsItemText: "truncate max-w-[180px]",
      tagsItemDelete: [
        "cursor-pointer",
        "inline-flex items-center",
        "disabled:pointer-events-none",
        "text-(--b24ui-icon-color-secondary)",
        "hover:text-(--b24ui-icon-color-secondary-hover)",
        "transition-none"
      ].join(" "),
      tagsItemDeleteIcon: "",
      tagsInput: ""
    },
    variants: {
      addNew: {
        true: {
          group: "",
          // p-0 isolate -m-px
          item: [
            "text-(--menu-popup-item-color)",
            "data-highlighted:not-data-disabled:text-(--menu-popup-item-color-hover)",
            "data-[state=checked]:text-(--menu-popup-item-color-hover)"
          ].join(" "),
          itemLabel: "flex flex-row flex-nowrap items-center justify-start gap-2",
          itemLeadingIcon: [
            "size-[20px] rounded-full",
            "text-(--ui-color-base-white-fixed)",
            "bg-(--ui-color-design-selection-content-icon-secondary)",
            "group-data-highlighted:not-data-disabled:text-(--ui-color-base-white-fixed)",
            "group-data-highlighted:not-data-disabled:bg-(--ui-color-design-selection-content-icon)",
            "group-data-[state=checked]:text-(--ui-color-base-white-fixed)",
            "group-data-[state=checked]:bg-(--ui-color-design-selection-content-icon)"
          ].join(" ")
        }
      },
      multiple: {
        true: {
          root: "flex-wrap",
          base: [
            "py-[6px] ps-[6px] pe-[39px]"
          ].join(" "),
          tagsInput: [
            "flex-1",
            "border-0",
            "bg-transparent",
            "ps-[6px] pe-[0px] py-0",
            "placeholder:text-(--ui-color-design-plain-na-content-secondary)",
            "focus:outline-none",
            "disabled:cursor-not-allowed",
            "disabled:opacity-30",
            "focus:ring-0 focus-visible:ring-0"
          ].join(" ")
        },
        false: {
          base: [
            "px-3",
            "placeholder:text-(--ui-color-design-plain-na-content-secondary)",
            "focus:outline-none",
            "disabled:cursor-not-allowed",
            "disabled:opacity-30"
          ].join(" ")
        }
      },
      size: {
        xss: {
          base: "[--main-ui-square-item-height:12px] h-[20px] gap-1 text-(length:--ui-font-size-4xs)/[normal]",
          leading: "px-1",
          trailing: "px-1",
          leadingIcon: "size-[12px]",
          leadingAvatarSize: "3xs",
          trailingIcon: "size-[12px]",
          tagsInput: "text-(length:--ui-font-size-4xs)/[normal]",
          tagsItem: [
            "text-(length:--ui-font-size-5xs)/(--main-ui-square-item-height)",
            "gap-0.5"
          ].join(" "),
          tagsItemDeleteIcon: "size-[10px]"
        },
        xs: {
          base: "[--main-ui-square-item-height:16px] h-[24px] gap-1 text-(length:--ui-font-size-xs)/[normal]",
          leading: "px-1",
          trailing: "px-1",
          leadingIcon: "size-[14px]",
          leadingAvatarSize: "3xs",
          trailingIcon: "size-[14px]",
          tagsInput: "text-(length:--ui-font-size-xs)/[normal]",
          tagsItem: [
            "text-(length:--ui-font-size-5xs)/(--main-ui-square-item-height)",
            "gap-0.5"
          ].join(" "),
          tagsItemDeleteIcon: "size-[10px]"
        },
        sm: {
          base: "[--main-ui-square-item-height:20px] h-[28px] gap-1.5 text-(length:--ui-font-size-sm)/[normal]",
          leading: "px-1.5",
          trailing: "px-1.5",
          leadingIcon: "size-[16px]",
          leadingAvatar: "size-[16px]",
          leadingAvatarSize: "2xs",
          trailingIcon: "size-[16px]",
          tagsInput: "text-(length:--ui-font-size-sm)/[normal]",
          tagsItem: [
            "text-(length:--ui-font-size-5xs)/(--main-ui-square-item-height)",
            "gap-0.5"
          ].join(" "),
          tagsItemDeleteIcon: "size-[12px]"
        },
        md: {
          base: "[--main-ui-square-item-height:24px] h-[34px] gap-1.5 text-(length:--ui-font-size-lg)/[normal]",
          leading: "px-2",
          trailing: "px-2",
          leadingIcon: "size-[18px]",
          leadingAvatarSize: "2xs",
          trailingIcon: "size-[18px]",
          tagsInput: "text-(length:--ui-font-size-lg)/[normal]",
          tagsItem: [
            "text-(length:--ui-font-size-sm)/(--main-ui-square-item-height)",
            "gap-[4px]"
          ].join(" "),
          tagsItemDeleteIcon: "size-[18px]"
        },
        lg: {
          base: "[--main-ui-square-item-height:28px] h-[38px] gap-2 text-(length:--ui-font-size-lg)/[normal]",
          leading: "px-2",
          trailing: "px-2",
          leadingIcon: "size-[22px]",
          leadingAvatarSize: "2xs",
          trailingIcon: "size-[22px]",
          tagsInput: "text-(length:--ui-font-size-lg)/[normal]",
          tagsItem: [
            "text-(length:--ui-font-size-md)/(--main-ui-square-item-height)",
            "gap-1"
          ].join(" "),
          tagsItemDeleteIcon: "size-[22px]"
        },
        xl: {
          base: "[--main-ui-square-item-height:32px] h-[46px] gap-2 text-(length:--ui-font-size-2xl)/[normal]",
          leading: "px-2",
          trailing: "px-2",
          leadingIcon: "size-[22px]",
          leadingAvatarSize: "2xs",
          trailingIcon: "size-[22px]",
          tagsInput: "text-(length:--ui-font-size-2xl)/[normal]",
          tagsItem: [
            "text-(length:--ui-font-size-xl)/(--main-ui-square-item-height)",
            "gap-1"
          ].join(" "),
          tagsItemDeleteIcon: "size-[22px]"
        }
      },
      colorItem: {
        "air-primary": { item: "style-filled" },
        "air-primary-success": { item: "style-filled-success" },
        "air-primary-alert": { item: "style-filled-alert" },
        "air-primary-copilot": { item: "style-filled-copilot" },
        "air-primary-warning": { item: "style-filled-warning" },
        "default": { item: "style-old-default" },
        "danger": { item: "style-old-danger" },
        "success": { item: "style-old-success" },
        "warning": { item: "style-old-warning" },
        "primary": { item: "style-old-primary" },
        "secondary": { item: "style-old-secondary" },
        "collab": { item: "style-old-collab" },
        "ai": { item: "style-old-ai" }
      }
    },
    compoundVariants: [
      // 'has-focus-visible:ring-1 has-focus-visible:ring-red-500'
      // from dropdown-menu
      {
        colorItem: ["air-primary", "air-primary-success", "air-primary-alert", "air-primary-copilot", "air-primary-warning"],
        active: false,
        class: {
          item: [
            "text-(--b24ui-background) data-highlighted:text-(--b24ui-background-hover) data-[state=open]:text-(--b24ui-background-hover)"
          ].join(" "),
          itemLeadingIcon: [
            "text-(--b24ui-background) group-data-highlighted:text-(--b24ui-background-hover) group-data-[state=open]:text-(--b24ui-background-hover)"
          ].join(" ")
        }
      },
      {
        colorItem: ["air-primary", "air-primary-success", "air-primary-alert", "air-primary-copilot", "air-primary-warning"],
        active: true,
        class: {
          item: [
            "text-(--b24ui-background-active)"
          ].join(" "),
          itemLeadingIcon: [
            "text-(--b24ui-background-active) group-data-[state=open]:text-(--b24ui-background-active)"
          ].join(" ")
        }
      },
      {
        colorItem: "default",
        active: false,
        class: ""
      },
      {
        colorItem: "default",
        active: true,
        class: ""
      },
      {
        colorItem: ["danger", "success", "warning", "primary", "secondary", "collab", "ai"],
        active: false,
        class: {
          item: [
            "text-(--b24ui-background-active) data-highlighted:text-(--b24ui-background-hover) data-[state=open]:text-(--b24ui-background-hover)"
          ].join(" "),
          itemLeadingIcon: [
            "text-(--b24ui-icon) group-data-highlighted:text-(--b24ui-icon) group-data-[state=open]:text-(--b24ui-icon)"
          ].join(" ")
        }
      },
      {
        colorItem: ["danger", "success", "warning", "primary", "secondary", "collab", "ai"],
        active: true,
        class: {
          item: [
            "text-(--b24ui-background-active)"
          ].join(" "),
          itemLeadingIcon: [
            "text-(--b24ui-background-active) group-data-[state=open]:text-(--b24ui-background-active)"
          ].join(" ")
        }
      },
      // region multiple.size -> h-auto ////
      {
        size: "xss",
        multiple: true,
        class: { base: "min-h-[20px] h-auto py-[2px] ps-[4px]" }
      },
      {
        size: "xs",
        multiple: true,
        class: { base: "min-h-[24px] h-auto py-[2px] ps-[4px]" }
      },
      {
        size: "sm",
        multiple: true,
        class: { base: "min-h-[28px] h-auto py-[4px] ps-[4px]" }
      },
      {
        size: "md",
        multiple: true,
        class: { base: "min-h-[34px] h-auto py-[4px] ps-[4px]" }
      },
      {
        size: "lg",
        multiple: true,
        class: { base: "min-h-[38px] h-auto py-[4px] ps-[4px]" }
      },
      {
        size: "xl",
        multiple: true,
        class: { base: "min-h-[46px] h-auto" }
      }
      // endregion ////
    ]
  }, input);
};

const inputNumber = () => {
  const input$1 = input;
  return {
    slots: {
      root: "isolate relative inline-flex items-center",
      base: [
        "w-full py-0 border-0 focus:outline-none",
        "disabled:cursor-not-allowed",
        "disabled:pointer-events-none",
        "disabled:opacity-30",
        "disabled:resize-none",
        "appearance-none transition duration-300 ease-linear",
        // transition-colors
        "ring ring-inset",
        "ring-(--ui-color-design-outline-stroke)",
        "text-(--ui-color-base-1)",
        "style-blurred-bg-input",
        "placeholder:text-(--ui-color-design-plain-na-content-secondary)",
        "hover:text-(--ui-color-base-1)",
        "focus:text-(--ui-color-base-1)",
        "active:text-(--ui-color-base-1)",
        "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
        "align-middle",
        "text-ellipsis whitespace-nowrap",
        "focus-visible:ring-1",
        "focus-visible:ring-inset",
        "focus-visible:ring-(--b24ui-border-color)"
      ].join(" "),
      increment: "absolute flex items-center",
      decrement: "absolute flex items-center",
      tag: input$1.slots.tag
    },
    variants: {
      ...buttonGroupVariantWithRoot,
      size: {
        xss: "px-2 h-[20px] gap-1 text-(length:--ui-font-size-4xs)/[normal]",
        xs: "px-2 h-[24px] gap-1 text-(length:--ui-font-size-xs)/[normal]",
        sm: "px-2.5 h-[28px] gap-1.5 text-(length:--ui-font-size-sm)/[normal]",
        md: "px-2.5 h-[34px] gap-1.5 text-(length:--ui-font-size-lg)/[normal]",
        lg: "px-3 h-[38px] gap-2 text-(length:--ui-font-size-lg)/[normal]",
        xl: "px-3 h-[46px] gap-2 text-(length:--ui-font-size-2xl)/[normal]"
      },
      color: {
        "air-primary": { base: "style-filled" },
        "air-primary-success": { base: "style-filled-success" },
        "air-primary-alert": { base: "style-filled-alert" },
        "air-primary-copilot": { base: "style-filled-copilot" },
        "air-primary-warning": { base: "style-filled-warning" },
        // @deprecate ////
        "default": { base: "style-old-default" },
        "danger": { base: "style-old-danger" },
        "success": { base: "style-old-success" },
        "warning": { base: "style-old-warning" },
        "primary": { base: "style-old-primary" },
        "secondary": { base: "style-old-secondary" },
        "collab": { base: "style-old-collab" },
        "ai": { base: "style-old-ai" }
      },
      /**
       * @memo now get from Badge
       */
      // tagColor: {},
      rounded: {
        ...input$1.variants.rounded
      },
      noPadding: {
        ...input$1.variants.noPadding
      },
      noBorder: {
        ...input$1.variants.noBorder
      },
      underline: {
        ...input$1.variants.underline
      },
      disabled: {
        true: {
          increment: "cursor-not-allowed",
          decrement: "cursor-not-allowed"
        }
      },
      orientation: {
        horizontal: {
          base: "text-center",
          increment: "inset-y-0 end-0 py-0 pe-0 [&>button]:p-1",
          decrement: "inset-y-0 start-0 py-0 ps-0 [&>button]:p-1"
        },
        vertical: {
          increment: "top-1 end-0 pe-1 [&>button]:p-0",
          decrement: "bottom-1 end-0 pe-1 [&>button]:p-0"
        }
      },
      highlight: {
        true: ""
      }
    },
    compoundVariants: [
      // region ring for focus and highlight ////
      // region color ////
      {
        noBorder: false,
        underline: false,
        class: ""
      },
      {
        highlight: true,
        noBorder: false,
        underline: false,
        class: "ring ring-inset ring-(--b24ui-border-color)"
      },
      {
        noBorder: false,
        underline: true,
        class: "focus-visible:border-(--b24ui-border-color)"
      },
      {
        highlight: true,
        noBorder: false,
        underline: true,
        class: "border-b-(--b24ui-border-color)"
      },
      // endregion ////
      // endregion ////
      // region orientation && size ////
      // region horizontal.rounded ////
      {
        orientation: "horizontal",
        rounded: true,
        class: {
          increment: "[&>button]:rounded-3xl",
          decrement: "[&>button]:rounded-3xl"
        }
      },
      // endregion ////
      // region horizontal ////
      {
        orientation: "horizontal",
        size: "xss",
        class: {
          base: "px-[24px]",
          increment: "[&>button]:p-[2px] [&>button]:h-[24px] scale-70",
          decrement: "[&>button]:p-[2px] [&>button]:h-[24px] scale-70"
        }
      },
      {
        orientation: "horizontal",
        size: "xs",
        class: {
          base: "px-[24px]",
          increment: "[&>button]:p-[2px] [&>button]:h-[24px] scale-70",
          decrement: "[&>button]:p-[2px] [&>button]:h-[24px] scale-70"
        }
      },
      {
        orientation: "horizontal",
        size: "sm",
        class: {
          base: "px-[28px]",
          increment: "[&>button]:p-[2px] scale-70",
          decrement: "[&>button]:p-[2px] scale-70"
        }
      },
      {
        orientation: "horizontal",
        size: "md",
        class: {
          base: "px-[30px]",
          increment: "[&>button]:p-[2px] scale-60",
          decrement: "[&>button]:p-[2px] scale-60"
        }
      },
      {
        orientation: "horizontal",
        size: "lg",
        class: {
          base: "px-[30px]",
          increment: "[&>button]:p-[2px] scale-60",
          decrement: "[&>button]:p-[2px] scale-60"
        }
      },
      {
        orientation: "horizontal",
        size: "xl",
        class: {
          base: "px-[40px]",
          increment: "[&>button]:p-[4px] scale-60",
          decrement: "[&>button]:p-[4px] scale-60"
        }
      },
      // endregion ////
      // region vertical ////
      {
        orientation: "vertical",
        size: "xss",
        class: {
          base: "pe-[24px]",
          increment: "top-0 pe-0 [&>button]:h-[13px] scale-56",
          decrement: "bottom-0 pe-0 [&>button]:h-[13px] scale-56"
        }
      },
      {
        orientation: "vertical",
        size: "xs",
        class: {
          base: "pe-[24px]",
          increment: "top-0 pe-0 [&>button]:h-[13px] scale-80",
          decrement: "bottom-0 pe-0 [&>button]:h-[13px] scale-80"
        }
      },
      {
        orientation: "vertical",
        size: "sm",
        class: {
          base: "pe-[30px]",
          increment: "top-0 pe-0 [&>button]:h-[15px] scale-80",
          decrement: "bottom-0 pe-0 [&>button]:h-[15px] scale-80"
        }
      },
      {
        orientation: "vertical",
        size: "md",
        class: {
          base: "pe-[34px]",
          increment: "top-0 pe-0 [&>button]:h-[18px] scale-80",
          decrement: "bottom-0 pe-0 [&>button]:h-[18px] scale-80"
        }
      },
      {
        orientation: "vertical",
        size: "lg",
        class: {
          base: "pe-[38px]",
          increment: "[&>button]:h-[16px] scale-80",
          decrement: "[&>button]:h-[16px] scale-80"
        }
      },
      {
        orientation: "vertical",
        size: "xl",
        class: {
          base: "pe-[38px]",
          increment: "[&>button]:h-[20px] scale-80",
          decrement: "[&>button]:h-[20px] scale-80"
        }
      }
      // endregion ////
      // endregion ////
    ],
    defaultVariants: {
      color: "air-primary",
      size: "md"
    }
  };
};

const kbd = {
  slots: {
    base: [
      "inline-flex items-center justify-center",
      "px-1 rounded-(--ui-border-radius-2xs)",
      "font-normal font-[family-name:var(--ui-font-family-system-mono)]",
      "border border-(length:--b24ui-border-width) border-(--b24ui-border-color)",
      "text-(--b24ui-color) bg-(--b24ui-background)"
    ].join(" ")
  },
  variants: {
    accent: {
      default: "style-outline",
      accent: "style-outline-accent-1",
      less: "style-outline-no-accent"
    },
    size: {
      sm: "h-[20px] min-w-[20px] text-(length:--ui-font-size-4xs)/(--ui-font-line-height-reset)",
      md: "h-[24px] min-w-[24px] text-(length:--ui-font-size-md)/(--ui-font-line-height-reset)",
      lg: "h-[28px] min-w-[28px] text-(length:--ui-font-size-2xl)/(--ui-font-line-height-reset)"
    }
  },
  defaultVariants: {
    accent: "less",
    size: "md"
  }
};

const link = (options) => ({
  base: [
    "cursor-pointer",
    // 'focus-visible:outline-info-text' // fix
    "focus-visible:outline-(--ui-color-accent-main-primary)",
    "focus-visible:outline-1 focus-visible:rounded-[4px]",
    "text-start"
  ].join(" "),
  variants: {
    active: {
      true: [
        "text-(--ui-color-accent-main-primary) outline-(--ui-color-accent-main-primary)",
        "hover:not-disabled:not-aria-disabled:underline underline-offset-2"
      ].join(" "),
      false: [
        "text-(--ui-color-accent-main-link)",
        "underline-offset-2"
      ].join(" ")
    },
    disabled: {
      true: "cursor-not-allowed opacity-75"
    },
    isAction: {
      true: [
        "text-nowrap",
        "text-sm h-auto py-0 font-normal rounded-none",
        "border border-x-0 border-t-0 border-dashed",
        "text-(--ui-color-design-outline-a1-content) border-b-(--ui-color-design-outline-a1-content)",
        "hover:not-disabled:not-aria-disabled:no-underline",
        "hover:text(--ui-color-accent-soft-element-red) hover:not-disabled:not-aria-disabled:text-(--ui-color-accent-soft-element-red) hover:border-b-(--ui-color-accent-soft-element-red)",
        "focus-visible:outline-(--ui-color-accent-soft-element-red)"
      ].join(" ")
    }
  },
  compoundVariants: [
    {
      active: false,
      disabled: false,
      class: [
        "hover:text-(--ui-color-accent-main-primary-alt)",
        "hover:underline"
      ].join(" ")
    }
  ]
});

const modal = {
  slots: {
    overlay: "fixed inset-0 bg-[#003366]/20",
    content: [
      "light",
      "bg-(--popup-window-background-color)",
      "fixed",
      "flex flex-col gap-[20px]",
      "focus:outline-none",
      "p-[24px] pt-[20px]"
    ].join(" "),
    contentWrapper: [
      "flex flex-col gap-[15px] pt-[4px]"
    ].join(" "),
    header: "flex items-start justify-between gap-[6px]",
    wrapper: "",
    title: [
      "font-[family-name:var(--ui-font-family-primary)]",
      "text-(--b24ui-typography-label-color)",
      "font-(--ui-font-weight-medium)",
      "mb-0",
      "text-[calc(var(--ui-font-size-2xl)+2px)]/(--ui-font-size-2xl)"
    ].join(" "),
    description: [
      "mt-1",
      "text-(--b24ui-typography-description-color)",
      "text-(length:--ui-font-size-sm)"
    ].join(" "),
    close: "-mt-[4px]",
    body: "flex-1 overflow-y-auto text-(length:--ui-font-size-md) leading-normal",
    footer: [
      "flex items-center justify-between gap-[10px]",
      "border-t border-t-1 border-t-(--ui-color-divider-default)",
      "pt-[18px]"
    ].join(" ")
  },
  variants: {
    overlayBlur: {
      auto: { overlay: "motion-safe:backdrop-blur-[2px]" },
      on: { overlay: "backdrop-blur-[2px]" },
      off: { overlay: "" }
    },
    transition: {
      true: {
        overlay: "motion-safe:data-[state=open]:animate-[fade-in_200ms_ease-out] motion-safe:data-[state=closed]:animate-[fade-out_200ms_ease-in]",
        content: "motion-safe:data-[state=open]:animate-[scale-in_200ms_ease-out] motion-safe:data-[state=closed]:animate-[scale-out_200ms_ease-in]"
      }
    },
    fullscreen: {
      true: {
        content: "inset-0"
      },
      false: {
        content: [
          "top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2",
          "w-[calc(100vw-2rem)] max-w-[32rem] max-h-[calc(100dvh-2rem)] sm:max-h-[calc(100dvh-4rem)]",
          "rounded-[calc(var(--popup-window-border-radius)-2px)] shadow-lg"
          // @memo see components/popup.css
          // 'ring ring-(--popup-window-border)'
        ].join(" "),
        contentWrapper: "overflow-hidden"
      }
    },
    scrollbarThin: {
      true: {
        body: "scrollbar-thin"
      }
    }
  },
  defaultVariants: {
    scrollbarThin: true,
    overlayBlur: "auto"
  }
};

const navigationMenu = {
  slots: {
    root: "relative flex [&>div]:min-w-0 font-[family-name:var(--ui-font-family-secondary)]",
    list: "isolate min-w-0",
    label: [
      "w-full h-[22px] overflow-hidden",
      "opacity-70 text-(length:--ui-font-size-sm)"
    ].join(" "),
    item: "min-w-0",
    link: [
      "min-w-[38px] max-w-full",
      "p-0",
      "m-0",
      "group relative",
      "cursor-pointer",
      "w-full",
      "flex items-center gap-[2px]",
      "font-normal text-lg",
      "focus:outline-none focus-visible:rounded-(--menu-item-border-radius) focus-visible:outline-none focus-visible:ring-inset focus-visible:ring-1 focus-visible:ring-(--ui-color-base-4)",
      "rounded-(--menu-item-border-radius)",
      "text-(--menu-item-color)",
      "bg-(--menu-item-background)",
      "hover:bg-(--menu-item-background-hover) data-[state=open]:bg-(--menu-item-background-hover)",
      "border-0"
    ].join(" "),
    linkLeadingIcon: "shrink-0 size-[26px]",
    linkLeadingAvatar: "shrink-0",
    linkLeadingAvatarSize: "xs",
    linkLeadingHint: [
      "inline-flex m-0 absolute -top-[5px] left-[8px]",
      "text-(length:--ui-font-size-4xs)",
      "leading-[8px]",
      "font-semibold",
      "text-(--b24ui-typography-description-color)",
      "uppercase",
      "ml-px"
    ].join(" "),
    linkLeadingBadge: "inline-flex m-0 absolute",
    linkLeadingBadgeSize: "xs",
    linkTrailing: "group inline-flex mt-[2px] items-center   ",
    linkTrailingIcon: [
      "text-(--ui-color-design-plain-na-content-icon)",
      "shrink-0"
    ].join(" "),
    linkLabel: "truncate -mt-px",
    linkLabelWrapper: "flex items-center justify-between rtl:flex-row-reverse",
    linkLabelExternalIcon: "inline-block size-[16px] text-(--ui-color-design-plain-content-icon-secondary)",
    childList: "isolate",
    childLabel: "",
    childItem: "h-[36px] mt-(--menu-item-block-stack-space)",
    childLink: [
      "group relative",
      "size-full",
      "flex flex-row rtl:flex-row-reverse items-center transition-colors",
      "text-start"
    ].join(" "),
    childLinkWrapper: "min-w-0 flex-1 flex flex-row items-center justify-start rtl:justify-end gap-0.5",
    childLinkIcon: "size-[18px] shrink-0",
    childLinkHint: [
      "inline-flex m-0 absolute -top-[2px] left-[24px]",
      "text-(length:--ui-font-size-4xs)",
      "leading-[8px]",
      "font-semibold",
      "text-(--b24ui-typography-description-color)",
      "uppercase",
      "ml-px"
    ].join(" "),
    childLinkBadge: "inline-flex m-0",
    childLinkBadgeSize: "xs",
    childLinkLabel: "truncate ms-[2px] -mt-px",
    childLinkLabelExternalIcon: "inline-block size-4 text-(--ui-color-design-plain-content-icon-secondary)",
    separator: "h-px bg-(--leftmenu-bg-divider) my-[16px]",
    popoverWrapper: "px-0 py-(--menu-popup-padding)",
    viewportWrapper: "absolute top-[53px] left-0 flex w-full",
    viewport: [
      "light",
      "relative overflow-hidden",
      "w-full",
      "bg-(--popup-window-background-color)",
      // 'bg-(--ui-color-base-white-fixed)/94',
      "shadow-(--popup-window-box-shadow)",
      "h-(--reka-navigation-menu-viewport-height)",
      "w-(--reka-navigation-menu-viewport-width) left-(--reka-navigation-menu-viewport-left)",
      "rounded-(--popup-window-border-radius) will-change-[opacity]",
      "[&:has(>[data-viewport=rtl])]:left-auto [&:has(>[data-viewport=rtl])]:-right-[calc(100%-var(--reka-navigation-menu-viewport-width))]",
      "transition-[width,height] duration-200 origin-[top_center]",
      // left
      // @memo see components/popup.css
      // 'border border-(--popup-window-border)'
      "z-[1]"
    ].join(" "),
    content: ""
  },
  variants: {
    orientation: {
      horizontal: {
        root: "relative h-full items-center justify-between",
        list: "flex items-center gap-x-1 h-full",
        item: "empty:hidden",
        link: [
          "menu-item-horizontal",
          "h-[32px] min-h-[32px]",
          "px-[10px]",
          "border border-(--menu-item-background) hover:border-(--ui-color-design-plain-na-focused-stroke) data-[state=open]:bg-(--ui-color-design-plain-na-focused-stroke)"
        ].join(" "),
        linkTrailingIcon: "size-[16px]",
        linkLeadingBadge: "-top-[6px] -right-[14px] -translate-x-1/2",
        linkLabelWrapper: "gap-[4px]",
        childList: "grid px-0 py-(--menu-popup-padding)",
        childLink: [
          "px-[18px]",
          // @memo 10 + 15 = 25 != 18px
          "min-w-[195px]",
          "whitespace-nowrap",
          "font-[family-name:var(--ui-font-family-primary)]",
          "text-(length:--menu-popup-item-font-size)",
          "text-(--menu-popup-item-color) hover:text-(--menu-popup-item-color-hover)",
          "hover:bg-(--menu-popup-item-bg-color-hover)"
        ].join(" "),
        childLinkLabel: "",
        content: "absolute top-0 left-0 w-full max-h-[70vh] overflow-y-auto scrollbar-thin scrollbar-transparent"
      },
      vertical: {
        root: "flex-col w-full ps-(--menu-items-block-padding-x) rtl:pe-(--menu-items-block-padding-x)",
        list: "flex flex-col",
        item: [
          "mt-(--menu-item-block-stack-space)",
          "data-[state=open]:rounded-(--menu-item-border-radius) data-[state=open]:border-(length:--leftmenu-group-stroke-weight) data-[state=open]:border-(--leftmenu-group-stroke)"
        ].join(" "),
        link: [
          "menu-item-vertical",
          "h-[38px] min-h-[38px]",
          "p-[6px]",
          "flex-row rtl:flex-row-reverse justify-between",
          "",
          "data-[state=open]:text-(length:--ui-font-size-sm) data-[state=open]:opacity-70"
        ].join(" "),
        linkLeadingIcon: "",
        // group-data-[state=open]:size-[22px]
        linkTrailingIcon: "size-[20px] group-data-[state=open]:rotate-180 transition-transform duration-200",
        linkLeadingBadge: "-top-[4px] left-[24px] -translate-x-1/2",
        linkLabelWrapper: "relative",
        childList: "",
        childLink: [
          "px-[18px]",
          // @memo 10 + 15 = 25 != 18
          "min-w-[195px]",
          "whitespace-nowrap",
          "font-[family-name:var(--ui-font-family-primary)]",
          "text-(length:--menu-popup-item-font-size)",
          "text-(--menu-popup-item-color) hover:text-(--menu-popup-item-color-hover)",
          "hover:bg-(--menu-popup-item-bg-color-hover)"
        ].join(" "),
        // @memo 10 + 15 = 25 != 18
        childLabel: [
          "w-full min-w-[195px] h-(--popup-window-delimiter-section-height)",
          "px-[18px] mt-(--menu-item-block-stack-space)",
          "flex flex-row rtl:flex-row-reverse items-center",
          "select-none outline-none whitespace-nowrap",
          "text-start",
          "text-(length:--popup-window-delimiter-font-size)",
          "text-(--popup-window-delimiter-text-color)",
          "font-(--popup-window-delimiter-font-weight)",
          "after:ms-[10px] after:block after:flex-1 after:min-w-[15px] after:h-px after:bg-(--popup-window-delimiter-bg-color)"
        ].join(" ")
      }
    },
    active: {
      true: {
        childLink: [
          "text-(--ui-color-accent-main-primary) hover:text-(--ui-color-accent-main-primary)"
          // [#0075ff]
        ].join(" "),
        childLinkIcon: "text-(--ui-color-accent-main-primary)"
      },
      false: {
        linkLeadingIcon: "text-(--ui-color-design-plain-content-icon-secondary)",
        childLinkIcon: [
          "text-(--ui-color-design-plain-content-icon-secondary)",
          "group-hover:text-(--ui-color-accent-main-primary)"
        ].join(" ")
      }
    },
    disabled: {
      true: {
        link: "cursor-not-allowed opacity-75"
      }
    },
    level: {
      true: ""
    },
    collapsed: {
      true: ""
    }
  },
  compoundVariants: [
    // region horizontal ////
    {
      orientation: "horizontal",
      class: {
        childList: "",
        content: "w-[240px]"
      }
    },
    {
      orientation: "vertical",
      collapsed: false,
      class: {
        item: "data-[state=open]:bg-(--leftmenu-group-bg)",
        childList: "",
        childItem: "",
        content: "motion-safe:data-[state=open]:animate-[collapsible-down_200ms_ease-out] motion-safe:data-[state=closed]:animate-[collapsible-up_200ms_ease-out] overflow-hidden",
        linkLabel: "ms-[9px]"
      }
    },
    {
      orientation: "vertical",
      collapsed: true,
      class: {
        childList: "grid px-0 py-(--menu-popup-padding)"
      }
    },
    {
      orientation: "vertical",
      collapsed: false,
      class: {
        link: "collapsed data-[state=open]:-mt-(--leftmenu-group-stroke-weight) data-[state=open]:-mx-(--leftmenu-group-stroke-weight)"
      }
    },
    // {
    //   orientation: 'horizontal',
    //   class: {
    //     link: ['after:absolute after:-bottom-2 after:inset-x-2.5 after:block after:h-px after:rounded-full', 'after:transition-colors']
    //   }
    // },
    // endregion ////
    // region vertical ////
    // {
    //   orientation: 'vertical',
    //   level: true,
    //   class: {
    //     link: ['after:absolute after:-start-1.5 after:inset-y-0.5 after:block after:w-[7px] after:rounded-full', 'after:transition-colors']
    //   }
    // },
    // endregion ////
    // region pill ////
    {
      disabled: false,
      active: false,
      class: {
        link: "transition-colors",
        linkLeadingIcon: "group-hover:text-(--ui-color-design-selection-content-icon)"
      }
    },
    {
      active: true,
      class: {
        link: [
          "leading-9",
          "text-(--menu-item-color-active)",
          "bg-(--menu-item-background-active)"
        ].join(" ")
      }
    },
    {
      active: true,
      orientation: "horizontal",
      class: {
        link: "menu-item-horizontal-active border-(--menu-item-border)"
      }
    },
    {
      active: true,
      orientation: "vertical",
      class: {
        link: "menu-item-vertical-active"
      }
    }
    // endregion ////
  ],
  defaultVariants: {}
};

const popover = {
  slots: {
    content: [
      "light",
      "bg-(--popup-window-background-color)",
      "shadow-(--popup-window-box-shadow)",
      "rounded-(--popup-window-border-radius) will-change-[opacity]",
      "motion-safe:data-[state=open]:animate-[scale-in_100ms_ease-out] motion-safe:data-[state=closed]:animate-[scale-out_100ms_ease-in]",
      "origin-(--reka-popover-content-transform-origin)",
      "focus:outline-none pointer-events-auto"
    ].join(" "),
    arrow: "fill-(--popup-window-background-color)"
  }
};

const progress = {
  slots: {
    root: "gap-2",
    base: [
      "relative overflow-hidden",
      "rounded-(--ui-border-radius-pill)",
      "bg-(--ui-color-base-5)"
    ],
    indicator: [
      "rounded-(--ui-border-radius-pill)",
      "size-full",
      "transition-transform duration-200 ease-out",
      "bg-(--b24ui-background)"
    ].join(" "),
    status: [
      "flex justify-end",
      "text-(--b24ui-typography-legend-color)",
      // text-base-500
      "transition-[width] duration-200"
    ].join(" "),
    steps: [
      "grid items-end",
      "text-(--b24ui-typography-legend-color)"
    ].join(" "),
    step: "truncate text-end row-start-1 col-start-1 transition-opacity"
  },
  variants: {
    animation: {
      "loading": "",
      "carousel": "",
      "carousel-inverse": "",
      "swing": "",
      "elastic": ""
    },
    color: {
      "air-primary": { root: "style-filled" },
      "air-primary-success": { root: "style-filled-success" },
      "air-primary-alert": { root: "style-filled-alert" },
      "air-primary-copilot": { root: "style-filled-copilot" },
      "air-primary-warning": { root: "style-filled-warning" },
      "air-secondary": { root: "style-tinted" },
      // @deprecate ////
      "default": { root: "style-old-default" },
      "danger": { root: "style-old-danger" },
      "success": { root: "style-old-success" },
      "warning": { root: "style-old-warning" },
      "primary": { root: "style-old-primary" },
      "secondary": { root: "style-old-secondary" },
      "collab": { root: "style-old-collab" },
      "ai": { root: "style-old-ai" }
    },
    size: {
      xs: {
        status: "text-(length:--ui-font-size-xs)/(--ui-font-line-height-sm)",
        steps: "text-(length:--ui-font-size-xs)/(--ui-font-line-height-sm)"
      },
      sm: {
        status: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm)",
        steps: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm)"
      },
      md: {
        status: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm)",
        steps: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm)"
      },
      lg: {
        status: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm)",
        steps: "text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm)"
      }
    },
    step: {
      active: {
        step: "opacity-100"
      },
      first: {
        step: "opacity-100 text-(--b24ui-typography-legend-color)"
      },
      other: {
        step: "opacity-0"
      },
      last: {
        step: ""
      }
    },
    orientation: {
      horizontal: {
        root: "w-full flex flex-col",
        base: "w-full",
        status: "flex-row"
      },
      vertical: {
        root: "h-full flex flex-row-reverse",
        base: "h-full",
        status: "flex-col min-w-[32px]"
      }
    },
    inverted: {
      true: {
        status: "self-end"
      }
    }
  },
  compoundVariants: [
    // region base ////
    {
      inverted: true,
      orientation: "horizontal",
      class: {
        step: "text-start",
        status: "flex-row-reverse"
      }
    },
    {
      inverted: true,
      orientation: "vertical",
      class: {
        steps: "items-start",
        status: "flex-col-reverse"
      }
    },
    // endregion ////
    // region horizontal ////
    {
      orientation: "horizontal",
      size: "xs",
      class: "h-[2px]"
    },
    {
      orientation: "horizontal",
      size: "sm",
      class: "h-[4px]"
    },
    {
      orientation: "horizontal",
      size: "md",
      class: "h-[8px]"
    },
    {
      orientation: "horizontal",
      size: "lg",
      class: "h-[12px]"
    },
    // endregion ////
    // region vertical ////
    {
      orientation: "vertical",
      size: "xs",
      class: "w-[2px]"
    },
    {
      orientation: "vertical",
      size: "sm",
      class: "w-[4px]"
    },
    {
      orientation: "vertical",
      size: "md",
      class: "w-[8px]"
    },
    {
      orientation: "vertical",
      size: "lg",
      class: "w-[12px]"
    },
    // endregion ////
    // region animation ////
    {
      orientation: "horizontal",
      animation: "carousel",
      class: {
        indicator: "data-[state=indeterminate]:animate-[carousel_2s_ease-in-out_infinite] data-[state=indeterminate]:rtl:animate-[carousel-rtl_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "vertical",
      animation: "carousel",
      class: {
        indicator: "data-[state=indeterminate]:animate-[carousel-vertical_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "horizontal",
      animation: "carousel-inverse",
      class: {
        indicator: "data-[state=indeterminate]:animate-[carousel-inverse_2s_ease-in-out_infinite] data-[state=indeterminate]:rtl:animate-[carousel-inverse-rtl_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "vertical",
      animation: "carousel-inverse",
      class: {
        indicator: "data-[state=indeterminate]:animate-[carousel-inverse-vertical_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "horizontal",
      animation: "swing",
      class: {
        indicator: "data-[state=indeterminate]:animate-[swing_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "vertical",
      animation: "swing",
      class: {
        indicator: "data-[state=indeterminate]:animate-[swing-vertical_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "horizontal",
      animation: "elastic",
      class: {
        indicator: "data-[state=indeterminate]:animate-[elastic_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "vertical",
      animation: "elastic",
      class: {
        indicator: "data-[state=indeterminate]:animate-[elastic-vertical_2s_ease-in-out_infinite]"
      }
    },
    {
      orientation: "horizontal",
      animation: "loading",
      class: {
        indicator: "data-[state=indeterminate]:animate-[progressbar-loading_2s_infinite]"
      }
    },
    {
      orientation: "vertical",
      animation: "loading",
      class: {
        indicator: "data-[state=indeterminate]:animate-[progressbar-loading-vertical_2s_infinite]"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-primary",
    animation: "loading",
    size: "md"
  }
};

const radioGroup = {
  slots: {
    root: "relative",
    fieldset: "flex",
    legend: [
      "mb-1.5",
      "block",
      "text-(--b24ui-typography-label-color)"
    ].join(" "),
    item: "flex items-start",
    base: [
      "cursor-pointer",
      "rounded-(--ui-border-radius-pill)",
      "ring ring-inset ring-(--ui-color-base-5)",
      "outline-(--ui-color-background-transparent) focus-visible:outline-2 focus-visible:outline-offset-2",
      "focus-visible:outline-(--b24ui-background)"
    ].join(" "),
    indicator: [
      "flex items-center justify-center",
      "size-full",
      "rounded-(--ui-border-radius-pill)",
      "after:bg-(--b24ui-color)",
      "after:rounded-(--ui-border-radius-pill)",
      "bg-(--b24ui-background)"
    ].join(" "),
    container: "flex items-center",
    wrapper: [
      "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
      "w-full"
    ].join(" "),
    label: [
      "cursor-pointer",
      "block",
      "text-(--b24ui-typography-label-color)"
    ].join(" "),
    description: "text-(--b24ui-typography-description-color)"
  },
  variants: {
    color: {
      "air-primary": { root: "style-filled" },
      "air-primary-success": { root: "style-filled-success" },
      "air-primary-alert": { root: "style-filled-alert" },
      "air-primary-copilot": { root: "style-filled-copilot" },
      "air-primary-warning": { root: "style-filled-warning" },
      // @deprecate ////
      "default": { root: "style-old-default" },
      "danger": { root: "style-old-danger" },
      "success": { root: "style-old-success" },
      "warning": { root: "style-old-warning" },
      "primary": { root: "style-old-primary" },
      "secondary": { root: "style-old-secondary" },
      "collab": { root: "style-old-collab" },
      "ai": { root: "style-old-ai" }
    },
    variant: {
      list: {},
      card: {
        // @memo style-outline-no-accent
        item: [
          "cursor-pointer",
          "items-start",
          "border border-(--ui-color-design-outline-na-stroke)",
          "bg-(--ui-color-design-outline-na-bg)",
          "has-data-[state=checked]:border-(--b24ui-border-color)"
        ].join(" ")
      },
      table: {
        item: [
          "cursor-pointer",
          "border border-(--ui-color-design-outline-na-stroke)",
          "bg-(--ui-color-design-outline-na-bg)",
          "has-data-[state=checked]:bg-(--b24ui-background)/24",
          "has-data-[state=checked]:border-(--b24ui-border-color)",
          "has-data-[state=checked]:text-(--b24ui-color)",
          "has-data-[state=checked]:z-[1]"
        ].join(" ")
      }
    },
    orientation: {
      horizontal: {
        fieldset: "flex-row",
        root: "me-2"
      },
      vertical: {
        fieldset: "flex-col"
      }
    },
    indicator: {
      start: {
        item: "flex-row",
        base: "me-2"
      },
      end: {
        item: "flex-row-reverse",
        base: "ms-2"
      },
      hidden: {
        base: "sr-only",
        wrapper: "text-center"
      }
    },
    size: {
      xs: {
        fieldset: "gap-x-[12px] gap-y-[4px]",
        legend: "text-(length:--ui-font-size-xs)",
        base: "size-[12px]",
        item: "text-(length:--ui-font-size-xs)",
        label: "leading-[11px]",
        container: "h-[12px]",
        indicator: "after:size-[4px]"
      },
      sm: {
        fieldset: "gap-x-[14px] gap-y-[6px]",
        legend: "text-(length:--ui-font-size-xs)",
        base: "size-[14px]",
        item: "text-(length:--ui-font-size-sm)",
        label: "leading-[14px]",
        container: "h-[14px]",
        indicator: "after:size-[6px]"
      },
      md: {
        fieldset: "gap-x-[16px] gap-y-[8px]",
        legend: "text-(length:--ui-font-size-sm)",
        base: "size-[16px]",
        item: "text-(length:--ui-font-size-lg)",
        label: "leading-[15px]",
        container: "h-[16px]",
        indicator: "after:size-[6px]"
      },
      lg: {
        fieldset: "gap-x-[16px] gap-y-[8px]",
        legend: "text-(length:--ui-font-size-sm)",
        base: "size-[20px]",
        item: "text-(length:--ui-font-size-xl)",
        label: "leading-[18px]",
        container: "h-[20px]",
        indicator: "after:size-[8px]"
      }
    },
    disabled: {
      true: {
        base: "cursor-not-allowed opacity-30",
        label: "cursor-not-allowed opacity-30"
      }
    },
    required: {
      true: {
        label: "after:content-['*'] after:ms-0.5 after:text-(--ui-color-accent-main-alert)"
      }
    }
  },
  compoundVariants: [
    // region card.size ////
    { size: "xs", variant: "card", class: { item: "px-[13px] py-[7px] rounded-(--ui-border-radius-xs)" } },
    // (28 - (1 + 1) - 12) / 2 = 7
    { size: "sm", variant: "card", class: { item: "px-[13px] py-[9px] rounded-(--ui-border-radius-sm)" } },
    // (34 - (1 + 1) - 14) / 2 = 9
    { size: "md", variant: "card", class: { item: "px-[17px] py-[10px] rounded-(--ui-border-radius-md)" } },
    // (38 - (1 + 1) - 16) / 2 = 10
    { size: "lg", variant: "card", class: { item: "px-[23px] py-[12px] rounded-(--ui-border-radius-md)" } },
    // (46 - (1 + 1) - 20) / 2 = 12
    // endregion ////
    // region table.size ////
    { size: "xs", variant: "table", class: { item: "px-[13px] py-[7px]" } },
    { size: "sm", variant: "table", class: { item: "px-[13px] py-[9px]" } },
    { size: "md", variant: "table", class: { item: "px-[17px] py-[10px]" } },
    { size: "lg", variant: "table", class: { item: "px-[23px] py-[12px]" } },
    // endregion ////
    // region orientation/table ////
    {
      size: "xs",
      variant: "table",
      orientation: "horizontal",
      class: {
        item: "first-of-type:rounded-s-(--ui-border-radius-xs) last-of-type:rounded-e-(--ui-border-radius-xs)",
        fieldset: "gap-0 -space-x-px"
      }
    },
    {
      size: "xs",
      variant: "table",
      orientation: "vertical",
      class: {
        item: "first-of-type:rounded-t-(--ui-border-radius-xs) last-of-type:rounded-b-(--ui-border-radius-xs)",
        fieldset: "gap-0 -space-y-px"
      }
    },
    {
      size: "sm",
      variant: "table",
      orientation: "horizontal",
      class: {
        item: "first-of-type:rounded-s-(--ui-border-radius-sm) last-of-type:rounded-e-(--ui-border-radius-sm)",
        fieldset: "gap-0 -space-x-px"
      }
    },
    {
      size: "sm",
      variant: "table",
      orientation: "vertical",
      class: {
        item: "first-of-type:rounded-t-(--ui-border-radius-sm) last-of-type:rounded-b-(--ui-border-radius-sm)",
        fieldset: "gap-0 -space-y-px"
      }
    },
    {
      size: ["lg", "md"],
      variant: "table",
      orientation: "horizontal",
      class: {
        item: "first-of-type:rounded-s-(--ui-border-radius-md) last-of-type:rounded-e-(--ui-border-radius-md)",
        fieldset: "gap-0 -space-x-px"
      }
    },
    {
      size: ["lg", "md"],
      variant: "table",
      orientation: "vertical",
      class: {
        item: "first-of-type:rounded-t-(--ui-border-radius-md) last-of-type:rounded-b-(--ui-border-radius-md)",
        fieldset: "gap-0 -space-y-px"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-primary",
    size: "md",
    variant: "list",
    orientation: "vertical",
    indicator: "start"
  }
};

const range = {
  slots: {
    root: "relative flex items-center select-none touch-none",
    track: [
      "relative",
      "rounded-(--ui-border-radius-pill)",
      "bg-(--ui-color-base-5)",
      "grow",
      "overflow-hidden"
    ].join(" "),
    range: [
      "absolute",
      "rounded-(--ui-border-radius-pill)",
      "bg-(--b24ui-background)"
    ].join(" "),
    thumb: [
      "rounded-(--ui-border-radius-pill)",
      "bg-(--ui-color-base-white-fixed)",
      "ring-2",
      "outline-transparent focus-visible:outline-2 focus-visible:outline-offset-2",
      "ring-(--b24ui-background)",
      "focus-visible:outline-(--b24ui-background-hover)"
    ].join(" ")
  },
  variants: {
    color: {
      "air-primary": { root: "style-filled" },
      "air-primary-success": { root: "style-filled-success" },
      "air-primary-alert": { root: "style-filled-alert" },
      "air-primary-copilot": { root: "style-filled-copilot" },
      "air-primary-warning": { root: "style-filled-warning" },
      // @deprecate ////
      "default": { root: "style-old-default" },
      "danger": { root: "style-old-danger" },
      "success": { root: "style-old-success" },
      "warning": { root: "style-old-warning" },
      "primary": { root: "style-old-primary" },
      "secondary": { root: "style-old-secondary" },
      "collab": { root: "style-old-collab" },
      "ai": { root: "style-old-ai" }
    },
    size: {
      xs: { thumb: "size-[12px]" },
      sm: { thumb: "size-[14px]" },
      md: { thumb: "size-[16px]" },
      lg: { thumb: "size-[20px]" }
    },
    orientation: {
      horizontal: {
        root: "w-full",
        range: "h-full"
      },
      vertical: {
        root: "flex-col h-full",
        range: "w-full"
      }
    },
    disabled: {
      true: { root: "opacity-30 cursor-not-allowed" }
    }
  },
  compoundVariants: [
    // region horizontal ////
    {
      orientation: "horizontal",
      size: "xs",
      class: {
        track: "h-[6px]"
      }
    },
    {
      orientation: "horizontal",
      size: "sm",
      class: {
        track: "h-[7px]"
      }
    },
    {
      orientation: "horizontal",
      size: "md",
      class: {
        track: "h-[8px]"
      }
    },
    {
      orientation: "horizontal",
      size: "lg",
      class: {
        track: "h-[9px]"
      }
    },
    {
      orientation: "horizontal",
      size: "xl",
      class: {
        track: "h-[10px]"
      }
    },
    // endregion ////
    // region vertical ////
    {
      orientation: "vertical",
      size: "xs",
      class: {
        track: "w-[6px]"
      }
    },
    {
      orientation: "vertical",
      size: "sm",
      class: {
        track: "w-[7px]"
      }
    },
    {
      orientation: "vertical",
      size: "md",
      class: {
        track: "w-[8px]"
      }
    },
    {
      orientation: "vertical",
      size: "lg",
      class: {
        track: "w-[9px]"
      }
    },
    {
      orientation: "vertical",
      size: "xl",
      class: {
        track: "w-[10px]"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-primary",
    size: "md"
  }
};

const select = () => {
  return defuFn(
    {
      slots: {
        root: () => "isolate relative inline-flex items-center w-full",
        // @memo w-full need
        base: () => [
          "relative inline-flex items-center group",
          "px-3",
          "w-full py-0 border-0 focus:outline-none",
          "cursor-pointer",
          "disabled:cursor-not-allowed",
          "disabled:pointer-events-none",
          "disabled:opacity-30",
          "disabled:resize-none",
          "appearance-none transition duration-300 ease-linear",
          // transition-colors
          "ring ring-inset",
          "ring-(--ui-color-design-outline-stroke)",
          "text-(--ui-color-base-1)",
          "style-blurred-bg-input",
          "hover:text-(--ui-color-base-1)",
          "focus:text-(--ui-color-base-1)",
          "active:text-(--ui-color-base-1)",
          "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
          "align-middle",
          "text-ellipsis whitespace-nowrap",
          "focus-visible:ring-1",
          "focus-visible:ring-inset",
          "focus-visible:ring-(--b24ui-border-color)"
        ].join(" "),
        value: "truncate pointer-events-none",
        placeholder: "truncate text-(--ui-color-design-plain-na-content-secondary)",
        content: [
          "light",
          // w-(--reka-combobox-trigger-width)
          "bg-(--popup-window-background-color)",
          "shadow-(--popup-window-box-shadow)",
          "rounded-(--popup-window-border-radius) will-change-[opacity]",
          "motion-safe:data-[state=open]:animate-[scale-in_100ms_ease-out] motion-safe:data-[state=closed]:animate-[scale-out_100ms_ease-in]",
          "origin-(--reka-dropdown-menu-content-transform-origin)",
          "font-[family-name:var(--ui-font-family-primary)]",
          "relative",
          "isolate",
          "px-0 py-(--menu-popup-padding)",
          "pointer-events-auto"
        ].join(" "),
        viewport: [
          "relative",
          "w-[240px] max-h-[40vh]",
          "overflow-x-hidden overflow-y-auto scrollbar-thin"
          // scrollbar-transparent
        ].join(" "),
        arrow: "fill-(--popup-window-background-color)",
        // for content bottom|top::start -> ml-[12px]
        group: "grid",
        empty: [
          "h-(--popup-window-delimiter-section-height)",
          "mt-(--menu-item-block-stack-space)",
          "py-[8px]",
          "select-none outline-none whitespace-nowrap",
          "text-center",
          "text-(length:--popup-window-delimiter-font-size)/(--ui-font-line-height-lg)",
          "text-(--popup-window-delimiter-text-color)",
          "font-(--popup-window-delimiter-font-weight)"
        ].join(" "),
        label: [
          "w-full min-w-[195px] h-(--popup-window-delimiter-section-height)",
          "px-[18px] mt-(--menu-item-block-stack-space)",
          "flex flex-row rtl:flex-row-reverse items-center",
          "select-none outline-none whitespace-nowrap",
          "text-start",
          "text-(length:--popup-window-delimiter-font-size)",
          "text-(--popup-window-delimiter-text-color)",
          "font-(--popup-window-delimiter-font-weight)",
          "after:ms-[10px] after:block after:flex-1 after:min-w-[15px] after:h-px after:bg-(--popup-window-delimiter-bg-color)"
        ].join(" "),
        separator: "my-[8px] mx-[18px] h-[1px] bg-(--popup-window-delimiter-bg-color)",
        item: [
          "group",
          "w-full min-w-[195px] h-[36px]",
          "px-[18px] mt-(--menu-item-block-stack-space)",
          "relative",
          "flex flex-row rtl:flex-row-reverse items-center",
          "select-none outline-none whitespace-nowrap",
          "cursor-pointer",
          "data-disabled:cursor-not-allowed data-disabled:opacity-30",
          "text-start",
          "text-(length:--menu-popup-item-font-size)",
          "text-(--menu-popup-item-color) hover:text-(--menu-popup-item-color-hover)",
          "data-highlighted:not-data-disabled:text-(--menu-popup-item-color-hover)",
          "data-[state=open]:text-(--menu-popup-item-color-hover)",
          "data-[state=checked]:text-(--menu-popup-item-color-hover)",
          "hover:bg-(--menu-popup-item-bg-color-hover)",
          "data-highlighted:bg-(--menu-popup-item-bg-color-hover)",
          "data-[state=open]:bg-(--menu-popup-item-bg-color-hover)",
          "transition-colors"
        ].join(" "),
        itemLeadingIcon: [
          "shrink-0",
          "size-[18px]",
          "text-(--ui-color-design-plain-content-icon-secondary)",
          "group-data-highlighted:not-data-disabled:text-(--ui-color-accent-main-primary)",
          "group-data-[state=open]:text-(--ui-color-accent-main-primary)",
          "group-data-[state=checked]:text-(--ui-color-accent-main-primary)",
          "transition-colors"
        ].join(" "),
        itemLeadingAvatar: "shrink-0 size-[16px] mx-px",
        // @memo 18-2px
        itemLeadingAvatarSize: "2xs",
        // @memo this wrong
        itemLeadingChip: "shrink-0 size-[16px] mx-px",
        itemLeadingChipSize: "sm",
        itemTrailing: "ml-auto rtl:ml-0 rtl:mr-auto inline-flex gap-1.5 items-center",
        itemTrailingIcon: "shrink-0 size-[24px] text-(--ui-color-design-plain-content-icon-secondary)",
        itemLabel: [
          "truncate ms-[2px] -mt-px",
          "group-data-[state=checked]:text-(--ui-color-accent-main-primary)"
        ].join(" ")
      },
      variants: {
        ...buttonGroupVariantWithRoot,
        size: {
          xss: {
            base: "h-[20px] gap-1 text-(length:--ui-font-size-4xs)/[normal]",
            leading: "px-1",
            trailing: "px-1",
            leadingIcon: "size-[12px]",
            leadingAvatarSize: "3xs",
            trailingIcon: "size-[12px]"
          },
          xs: {
            base: "h-[24px] gap-1 text-(length:--ui-font-size-xs)/[normal]",
            leading: "px-1",
            trailing: "px-1",
            leadingIcon: "size-[14px]",
            leadingAvatarSize: "3xs",
            trailingIcon: "size-[14px]"
          },
          sm: {
            base: "h-[28px] gap-1.5 text-(length:--ui-font-size-sm)/[normal]",
            leading: "px-1.5",
            trailing: "px-1.5",
            leadingIcon: "size-[16px]",
            leadingAvatar: "size-[16px]",
            leadingAvatarSize: "2xs",
            trailingIcon: "size-[16px]"
          },
          md: {
            base: "h-[34px] gap-1.5 text-(length:--ui-font-size-lg)/[normal]",
            leading: "px-2",
            trailing: "px-2",
            leadingIcon: "size-[18px]",
            leadingAvatarSize: "2xs",
            trailingIcon: "size-[18px]"
          },
          lg: {
            base: "h-[38px] gap-2 text-(length:--ui-font-size-lg)/[normal]",
            leading: "px-2",
            trailing: "px-2",
            leadingIcon: "size-[22px]",
            leadingAvatarSize: "2xs",
            trailingIcon: "size-[22px]"
          },
          xl: {
            base: "h-[46px] gap-2 text-(length:--ui-font-size-2xl)/[normal]",
            leading: "px-2",
            trailing: "px-2",
            leadingIcon: "size-[22px]",
            leadingAvatarSize: "2xs",
            trailingIcon: "size-[22px]"
          }
        },
        colorItem: {
          "air-primary": { item: "style-filled" },
          "air-primary-success": { item: "style-filled-success" },
          "air-primary-alert": { item: "style-filled-alert" },
          "air-primary-copilot": { item: "style-filled-copilot" },
          "air-primary-warning": { item: "style-filled-warning" },
          "default": { item: "style-old-default" },
          "danger": { item: "style-old-danger" },
          "success": { item: "style-old-success" },
          "warning": { item: "style-old-warning" },
          "primary": { item: "style-old-primary" },
          "secondary": { item: "style-old-secondary" },
          "collab": { item: "style-old-collab" },
          "ai": { item: "style-old-ai" }
        }
      },
      compoundVariants: (prev) => {
        return [
          // from dropdown-menu
          {
            colorItem: ["air-primary", "air-primary-success", "air-primary-alert", "air-primary-copilot", "air-primary-warning"],
            active: false,
            class: {
              item: [
                "text-(--b24ui-background) data-highlighted:text-(--b24ui-background-hover) data-[state=open]:text-(--b24ui-background-hover)"
              ].join(" "),
              itemLeadingIcon: [
                "text-(--b24ui-background) group-data-highlighted:text-(--b24ui-background-hover) group-data-[state=open]:text-(--b24ui-background-hover)"
              ].join(" ")
            }
          },
          {
            colorItem: ["air-primary", "air-primary-success", "air-primary-alert", "air-primary-copilot", "air-primary-warning"],
            active: true,
            class: {
              item: [
                "text-(--b24ui-background-active)"
              ].join(" "),
              itemLeadingIcon: [
                "text-(--b24ui-background-active) group-data-[state=open]:text-(--b24ui-background-active)"
              ].join(" ")
            }
          },
          {
            colorItem: "default",
            active: false,
            class: ""
          },
          {
            colorItem: "default",
            active: true,
            class: ""
          },
          {
            colorItem: ["danger", "success", "warning", "primary", "secondary", "collab", "ai"],
            active: false,
            class: {
              item: [
                "text-(--b24ui-background-active) data-highlighted:text-(--b24ui-background-hover) data-[state=open]:text-(--b24ui-background-hover)"
              ].join(" "),
              itemLeadingIcon: [
                "text-(--b24ui-icon) group-data-highlighted:text-(--b24ui-icon) group-data-[state=open]:text-(--b24ui-icon)"
              ].join(" ")
            }
          },
          {
            colorItem: ["danger", "success", "warning", "primary", "secondary", "collab", "ai"],
            active: true,
            class: {
              item: [
                "text-(--b24ui-background-active)"
              ].join(" "),
              itemLeadingIcon: [
                "text-(--b24ui-background-active) group-data-[state=open]:text-(--b24ui-background-active)"
              ].join(" ")
            }
          },
          ...prev.map((item) => ({
            ...item,
            class: typeof item.class === "string" ? replaceFocus$1(item.class) : item.class
          }))
        ];
      }
    },
    input
  );
};
function replaceFocus$1(str) {
  return str.replace(/focus-visible:/g, "focus:");
}

const selectMenu = () => {
  return defuFn({
    slots: {
      input: "border-b border-(--popup-window-delimiter-bg-color)",
      focusScope: "flex flex-col min-h-0",
      content: (content) => [
        content,
        "origin-(--reka-combobox-content-transform-origin) "
        // w-(--reka-combobox-trigger-width)
      ]
    },
    variants: {
      addNew: {
        true: {
          group: "",
          // p-0 isolate -m-px
          item: [
            "text-(--menu-popup-item-color)",
            "data-highlighted:not-data-disabled:text-(--menu-popup-item-color-hover)",
            "data-[state=checked]:text-(--menu-popup-item-color-hover)"
          ].join(" "),
          itemLabel: "flex flex-row flex-nowrap items-center justify-start gap-2",
          itemLeadingIcon: [
            "size-[20px] rounded-full",
            "text-(--ui-color-base-white-fixed)",
            "bg-(--ui-color-design-selection-content-icon-secondary)",
            "group-data-highlighted:not-data-disabled:text-(--ui-color-base-white-fixed)",
            "group-data-highlighted:not-data-disabled:bg-(--ui-color-design-selection-content-icon)",
            "group-data-[state=checked]:text-(--ui-color-base-white-fixed)",
            "group-data-[state=checked]:bg-(--ui-color-design-selection-content-icon)"
          ].join(" ")
        }
      }
    },
    compoundVariants: (prev) => prev.map((item) => ({
      ...item,
      class: typeof item.class === "string" ? replaceFocus(item.class) : item.class
    }))
  }, select());
};
function replaceFocus(str) {
  return str.replace(/focus:/g, "focus-visible:");
}

const separator = {
  slots: {
    root: "flex items-center align-center text-center",
    border: "",
    container: "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-normal) flex",
    icon: "shrink-0 size-7",
    avatar: "shrink-0",
    avatarSize: "sm",
    label: ""
  },
  variants: {
    accent: {
      default: {
        container: "text-(--ui-color-design-plain-na-content-secondary) text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset)",
        border: "border-(--ui-color-divider-vibrant-default)"
      },
      accent: {
        container: "text-(--b24ui-typography-description-color) text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset)",
        border: "border-(--ui-color-divider-vibrant-accent-more)"
      },
      less: {
        container: "text-(--ui-color-base-6) text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset)",
        border: "border-(--ui-color-divider-vibrant-less)"
      }
    },
    orientation: {
      horizontal: {
        root: "w-full flex-row",
        border: "w-full",
        container: "mx-3 whitespace-nowrap"
      },
      vertical: {
        root: "h-full flex-col",
        border: "h-full",
        container: "my-2"
      }
    },
    size: {
      thin: "",
      thick: ""
    },
    type: {
      solid: {
        border: "border-solid"
      },
      dashed: {
        border: "border-dashed"
      },
      dotted: {
        border: "border-dotted"
      },
      double: {
        border: "border-double"
      }
    }
  },
  compoundVariants: [
    // region horizontal ////
    {
      orientation: "horizontal",
      size: "thin",
      class: { border: "border-t-(length:--ui-border-width-thin)" }
    },
    {
      orientation: "horizontal",
      size: "thick",
      class: { border: "border-t-(length:--ui-border-width-thick)" }
    },
    // endregion ////
    // region vertical ////
    {
      orientation: "vertical",
      size: "thin",
      class: { border: "border-s-(length:--ui-border-width-thin)" }
    },
    {
      orientation: "vertical",
      size: "thick",
      class: { border: "border-s-(length:--ui-border-width-thick)" }
    },
    // endregion ////
    // region custom for double.thick ////
    {
      type: "double",
      size: "thick",
      orientation: "horizontal",
      class: { border: "border-t-[length:calc(var(--ui-border-width-thick)_+_1px)]" }
    },
    {
      type: "double",
      size: "thick",
      orientation: "vertical",
      class: { border: "border-s-[length:calc(var(--ui-border-width-thick)_+_1px)]" }
    }
    // endregion ////
  ],
  defaultVariants: {
    accent: "default",
    size: "thin",
    type: "solid"
  }
};

const skeleton = {
  slots: {
    base: "animate-pulse rounded-(--ui-border-radius-md) "
  },
  variants: {
    accent: {
      default: "bg-(--ui-color-g-glass-grey-bg-1)",
      accent: "bg-(--ui-color-g-glass-grey-bg-2)",
      less: "bg-(--ui-color-g-glass-grey-bg-3)"
    }
  },
  defaultVariants: {
    accent: "default"
  }
};

const safeList = [
  "group-hover:rounded-full group-hover:border-1 group-hover:border-current"
].join(" ");
const slideover = {
  slots: {
    overlay: "fixed inset-0 bg-linear-to-b from-[#00204e]/52 to-[#00204e]",
    //
    content: [
      "fixed",
      "sm:shadow-lg",
      "flex flex-col focus:outline-none",
      "h-full"
    ].join(" "),
    sidebarLayoutRoot: "",
    sidebarLayoutHeaderWrapper: "",
    sidebarLayoutPageBottomWrapper: "",
    sidebarLayoutLoadingWrapper: "",
    sidebarLayoutLoadingIcon: "",
    header: [
      "pt-[24px]",
      "flex-1 flex items-center gap-x-[12px] gap-y-1.5"
    ].join(" "),
    wrapper: "min-h-[30px]",
    title: [
      "font-[family-name:var(--ui-font-family-primary)]",
      "text-(--b24ui-typography-label-color)",
      "font-(--ui-font-weight-semi-bold)",
      "mb-0",
      "text-(length:--ui-font-size-4xl)/[calc(var(--ui-font-size-4xl)+2px)]"
    ].join(" "),
    description: [
      "mt-1",
      "text-(--b24ui-typography-description-color)",
      "text-(length:--ui-font-size-sm)"
    ].join(" "),
    close: "absolute",
    body: [
      "size-full",
      "flex-1"
    ].join(" "),
    footer: [
      "light",
      "bg-(--popup-window-background-color)",
      // --ui-color-bg-content-primary
      "flex items-center justify-center gap-3",
      "border-t border-t-1 border-t-(--ui-color-divider-less)",
      "shadow-top-md",
      "py-[9px] px-2 pr-(--scrollbar-width)"
    ].join(" "),
    safeList
  },
  variants: {
    overlayBlur: {
      auto: {
        overlay: "motion-safe:backdrop-blur-sm"
      },
      on: {
        overlay: "backdrop-blur-sm"
      },
      off: {
        overlay: ""
      }
    },
    side: {
      // @todo fix if parent SidebarLayout.isUseSideBar > w-[calc(100%-135x)] > w-[calc(100%-150px)]
      right: {
        content: "right-0 inset-y-0 w-[calc(100%-60px)] sm:w-[calc(100%-150px)]",
        sidebarLayoutRoot: "sm:rounded-t-none"
      },
      left: {
        content: "left-0 inset-y-0 w-[calc(100%-60px)] sm:w-[calc(100%-150px)]",
        sidebarLayoutRoot: "sm:rounded-t-none"
      },
      top: {
        content: "inset-x-0 top-0 max-h-full",
        sidebarLayoutRoot: "sm:rounded-t-none"
      },
      bottom: {
        content: [
          "right-[5px] sm:right-[70px] top-0 sm:top-[18px] bottom-0",
          "w-[calc(100%-60px-5px)] sm:w-[calc(100%-150px-70px)]",
          "sm:max-h-[calc(100%-18px)]"
        ].join(" "),
        // sidebarLayoutRoot: 'sm:rounded-t-none'
        sidebarLayoutRoot: "sm:rounded-t-[18px]"
        // fix sidebarLayoutHeaderWrapper: '' // sm:rounded-t-[18px]
      }
    },
    transition: {
      true: {
        overlay: "motion-safe:data-[state=open]:animate-[fade-in_200ms_ease-out] motion-safe:data-[state=closed]:animate-[fade-out_200ms_ease-in]"
      }
    }
  },
  compoundVariants: [
    // region btn.close ////
    {
      side: ["right", "bottom"],
      class: {
        close: [
          "pl-1.5 pr-[4px]",
          "top-[17px] -translate-x-full left-[1px]",
          "rounded-l-full"
        ].join(" ")
      }
    },
    {
      side: "left",
      class: {
        close: [
          "pr-1.5 pl-[4px]",
          "top-[17px] translate-x-full right-[1px]",
          "rounded-r-full",
          "[&>div]:flex-row-reverse"
        ].join(" ")
      }
    },
    {
      side: "top",
      class: {
        close: [
          "top-4 end-4"
        ].join(" ")
      }
    },
    // endregion ////
    // region transition ////
    {
      transition: true,
      side: "top",
      class: {
        content: "motion-safe:data-[state=open]:animate-[slide-in-from-top_200ms_ease-in-out] motion-safe:data-[state=closed]:animate-[slide-out-to-top_200ms_ease-in-out]"
      }
    },
    {
      transition: true,
      side: "right",
      class: {
        content: "motion-safe:data-[state=open]:animate-[slide-in-from-right_200ms_ease-in-out] motion-safe:data-[state=closed]:animate-[slide-out-to-right_200ms_ease-in-out]"
      }
    },
    {
      transition: true,
      side: "bottom",
      class: {
        content: "motion-safe:data-[state=open]:animate-[slide-in-from-bottom_200ms_ease-in-out] motion-safe:data-[state=closed]:animate-[slide-out-to-bottom_200ms_ease-in-out]"
      }
    },
    {
      transition: true,
      side: "left",
      class: {
        content: "motion-safe:data-[state=open]:animate-[slide-in-from-left_200ms_ease-in-out] motion-safe:data-[state=closed]:animate-[slide-out-to-left_200ms_ease-in-out]"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    side: "bottom",
    overlayBlur: "off"
  }
};

const _switch = {
  slots: {
    root: "relative flex items-start",
    base: [
      "cursor-pointer",
      "inline-flex items-center",
      "shrink-0",
      "rounded-(--ui-border-radius-pill)",
      "outline-(--ui-color-background-transparent)",
      "focus-visible:outline-2 focus-visible:outline-offset-2",
      "transition-[background] duration-200",
      "border-2 border-(--ui-color-background-transparent)",
      "data-[state=unchecked]:bg-(--ui-color-base-5)",
      "data-[state=checked]:bg-(--b24ui-background) focus-visible:outline-(--b24ui-border-color)"
    ].join(" "),
    container: "flex items-center",
    thumb: [
      "group",
      "pointer-events-none",
      "flex items-center justify-center",
      "rounded-(--ui-border-radius-pill)",
      "bg-(--ui-color-base-white-fixed)",
      "shadow-lg",
      "ring-0",
      "transition-transform duration-200",
      "data-[state=unchecked]:translate-x-0 data-[state=unchecked]:rtl:-translate-x-0"
    ].join(" "),
    icon: [
      "absolute shrink-0",
      "group-data-[state=unchecked]:text-(--ui-color-base-5) edge-dark:group-data-[state=unchecked]:text-(--ui-color-gray-60)",
      "opacity-0",
      "p-0.5",
      "size-full",
      "transition-[color,opacity] duration-200",
      "group-data-[state=checked]:text-(--b24ui-background)"
    ].join(" "),
    wrapper: [
      "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
      "ms-2"
    ].join(" "),
    label: [
      "cursor-pointer",
      "block",
      "text-(--b24ui-typography-label-color)"
    ].join(" "),
    description: "text-(--b24ui-typography-description-color)"
  },
  variants: {
    color: {
      "air-primary": { base: "style-filled" },
      "air-primary-success": { base: "style-filled-success" },
      "air-primary-alert": { base: "style-filled-alert" },
      "air-primary-copilot": { base: "style-filled-copilot" },
      "air-primary-warning": { base: "style-filled-warning" },
      // @deprecate ////
      "default": { base: "style-old-default" },
      "danger": { base: "style-old-danger" },
      "success": { base: "style-old-success" },
      "warning": { base: "style-old-warning" },
      "primary": { base: "style-old-primary" },
      "secondary": { base: "style-old-secondary" },
      "collab": { base: "style-old-collab" },
      "ai": { base: "style-old-ai" }
    },
    size: {
      xs: {
        base: "w-[28px]",
        container: "h-[16px]",
        thumb: "size-[12px] data-[state=checked]:translate-x-3 data-[state=checked]:rtl:-translate-x-3",
        wrapper: "text-(length:--ui-font-size-xs)",
        label: "leading-[16px]"
      },
      sm: {
        base: "w-[32px]",
        container: "h-[16px]",
        thumb: "size-[14px] data-[state=checked]:translate-x-3.5 data-[state=checked]:rtl:-translate-x-3.5",
        wrapper: "text-(length:--ui-font-size-sm)",
        label: "leading-[16px]"
      },
      md: {
        base: "w-[36px]",
        container: "h-[20px]",
        thumb: "size-[16px] data-[state=checked]:translate-x-4 data-[state=checked]:rtl:-translate-x-4",
        wrapper: "text-(length:--ui-font-size-md)",
        label: "leading-[20px]"
      },
      lg: {
        base: "w-[40px]",
        container: "h-[20px]",
        thumb: "size-[20px] data-[state=checked]:translate-x-4 data-[state=checked]:rtl:-translate-x-4",
        wrapper: "text-(length:--ui-font-size-xl)",
        label: "leading-[20px]"
      }
    },
    checked: {
      true: {
        icon: "group-data-[state=checked]:opacity-100"
      }
    },
    unchecked: {
      true: {
        icon: "group-data-[state=unchecked]:opacity-100"
      }
    },
    loading: {
      true: {
        icon: "animate-spin"
      }
    },
    required: {
      true: {
        label: "after:content-['*'] after:ms-0.5 after:text-(--ui-color-accent-main-alert)"
      }
    },
    disabled: {
      true: {
        base: "cursor-not-allowed opacity-30",
        label: "cursor-not-allowed opacity-30",
        description: "cursor-not-allowed opacity-30"
      }
    }
  },
  defaultVariants: {
    color: "air-primary",
    size: "md"
  }
};

const tabs = {
  slots: {
    root: [
      "style-outline-accent-2",
      // air-secondary-accent-2
      "flex items-center gap-2"
    ],
    list: "relative flex p-1 group",
    indicator: "absolute transition-[translate,width] duration-200",
    trigger: [
      "group relative inline-flex items-center min-w-0",
      "data-[state=inactive]:text-(--ui-color-design-plain-na-content)",
      "hover:data-[state=inactive]:not-disabled:text-(--ui-color-design-selection-content)",
      "font-(--ui-font-weight-medium)",
      "cursor-pointer",
      "disabled:cursor-not-allowed",
      "disabled:opacity-30",
      "transition-colors",
      "rounded-(--ui-border-radius-md)"
      // @memo for ring
    ].join(" "),
    leadingIcon: "shrink-0",
    leadingAvatar: "shrink-0",
    leadingAvatarSize: "",
    label: "",
    trailingBadge: "shrink-0",
    trailingBadgeSize: "sm",
    content: "focus:outline-none w-full"
  },
  variants: {
    variant: {
      link: {
        list: "border-(--ui-color-divider-vibrant-accent-more)",
        indicator: "rounded-(--ui-border-radius-pill)",
        trigger: "focus:outline-none"
      }
    },
    orientation: {
      horizontal: {
        root: "flex-col",
        list: "w-full",
        indicator: "left-0 w-(--reka-tabs-indicator-size) translate-x-(--reka-tabs-indicator-position)",
        trigger: "justify-center"
      },
      vertical: {
        list: "flex-col",
        indicator: "top-0 h-(--reka-tabs-indicator-size) translate-y-(--reka-tabs-indicator-position)"
      }
    },
    size: {
      xss: {
        trigger: "px-2 py-1 text-(length:--ui-font-size-4xs)/[normal] gap-1",
        leadingIcon: "size-4",
        leadingAvatarSize: "3xs"
      },
      xs: {
        trigger: "px-2 py-1 text-(length:--ui-font-size-xs)/[normal] gap-1",
        leadingIcon: "size-4",
        leadingAvatarSize: "3xs"
      },
      sm: {
        trigger: "px-2.5 py-1.5 text-(length:--ui-font-size-sm)/[normal] gap-1.5",
        leadingIcon: "size-4",
        leadingAvatarSize: "3xs"
      },
      md: {
        trigger: "px-3 py-1.5 text-(length:--ui-font-size-md)/[normal] gap-1.5",
        leadingIcon: "size-5",
        leadingAvatarSize: "2xs"
      },
      lg: {
        trigger: "px-3 py-2 text-(length:--ui-font-size-lg)/[normal] gap-2",
        leadingIcon: "size-5",
        leadingAvatarSize: "2xs"
      },
      xl: {
        trigger: "px-3 py-2 text-(length:--ui-font-size-xl)/[normal] gap-2",
        leadingIcon: "size-6",
        leadingAvatarSize: "xs"
      }
    }
  },
  compoundVariants: [
    // region orientation ////
    // region orientation horizontal.link ////
    {
      orientation: "horizontal",
      variant: "link",
      class: {
        list: "border-b -mb-px",
        indicator: "-bottom-px h-px"
      }
    },
    // endregion ////
    // region orientation vertical.link ////
    {
      orientation: "vertical",
      variant: "link",
      class: {
        list: "border-s -ms-px",
        indicator: "-start-px w-px"
      }
    },
    // endregion ////
    // endregion ////
    // region color ////
    {
      variant: "link",
      class: {
        indicator: "bg-(--ui-color-design-selection-content)",
        trigger: [
          "focus-visible:ring-1 focus-visible:ring-inset",
          "data-[state=active]:text-(--b24ui-color)",
          "focus-visible:ring-(--ui-color-design-selection-content)"
        ]
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    variant: "link",
    size: "md"
  }
};

const textarea = {
  slots: {
    root: "isolate relative inline-flex items-center w-full",
    // @todo remove ? w-full -> see select
    base: [
      "px-3",
      "w-full pt-[7px] pb-[8px] border-0 focus:outline-none",
      "disabled:cursor-not-allowed",
      "disabled:pointer-events-none",
      "disabled:opacity-30",
      "disabled:resize-none",
      "appearance-none transition duration-300 ease-linear",
      // transition-colors
      "ring ring-inset",
      "ring-(--ui-color-design-outline-stroke)",
      "text-(--ui-color-base-1)",
      "style-blurred-bg-input",
      "placeholder:text-(--ui-color-design-plain-na-content-secondary)",
      "hover:text-(--ui-color-base-1)",
      "focus:text-(--ui-color-base-1)",
      "active:text-(--ui-color-base-1)",
      "font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-regular)",
      // @memo need set lineHeight
      "text-(length:--ui-font-size-lg)/(--ui-font-line-height-2xs)",
      "align-middle",
      "focus-visible:ring-1",
      "focus-visible:ring-inset",
      "focus-visible:ring-(--b24ui-border-color)"
    ].join(" "),
    leading: "absolute inset-y-[7px] start-0 flex items-start px-[8px]",
    leadingIcon: [
      "shrink-0",
      "size-[18px]",
      "text-(--b24ui-icon-color)"
    ].join(" "),
    leadingAvatar: "shrink-0 size-[20px]",
    leadingAvatarSize: "2xs",
    trailing: "absolute inset-y-[8px] end-0 flex items-start px-[6px]",
    trailingIcon: [
      "shrink-0",
      "size-[18px]",
      "text-(--b24ui-icon-color)"
    ].join(" "),
    tag: [
      "pointer-events-none select-none",
      "absolute",
      "z-10",
      "-top-[6px]",
      "right-[12px]"
    ].join(" ")
  },
  variants: {
    ...buttonGroupVariantWithRoot,
    autoresize: { true: { base: "resize-none" } },
    color: {
      "air-primary": { base: "style-filled" },
      "air-primary-success": { base: "style-filled-success" },
      "air-primary-alert": { base: "style-filled-alert" },
      "air-primary-copilot": { base: "style-filled-copilot" },
      "air-primary-warning": { base: "style-filled-warning" },
      // @deprecate ////
      "default": { base: "style-old-default" },
      "danger": { base: "style-old-danger" },
      "success": { base: "style-old-success" },
      "warning": { base: "style-old-warning" },
      "primary": { base: "style-old-primary" },
      "secondary": { base: "style-old-secondary" },
      "collab": { base: "style-old-collab" },
      "ai": { base: "style-old-ai" }
    },
    /**
     * @memo now get from Badge
     */
    // tagColor: {},
    rounded: {
      true: "rounded-(--ui-border-radius-3xl)",
      false: "rounded-(--ui-border-radius-sm)"
    },
    noPadding: { true: { base: "px-0" } },
    noBorder: { true: "ring-0 focus-visible:ring-0 style-transparent-bg" },
    underline: {
      true: [
        "ring-0 focus-visible:ring-0 style-transparent-bg",
        "border-b-1",
        "border-b-(--ui-color-design-outline-stroke)",
        "rounded-none"
      ].join(" ")
    },
    leading: { true: "" },
    trailing: { true: "" },
    loading: { true: "" },
    highlight: { true: "ring ring-inset ring-(--b24ui-border-color)" }
  },
  compoundVariants: [
    // region ring for focus and highlight ////
    // region color ////
    {
      noBorder: false,
      underline: false,
      class: ""
    },
    {
      highlight: true,
      noBorder: false,
      underline: false,
      class: "ring ring-inset ring-(--b24ui-border-color)"
    },
    {
      noBorder: false,
      underline: true,
      class: "focus-visible:border-(--b24ui-border-color)"
    },
    {
      highlight: true,
      noBorder: false,
      underline: true,
      class: "border-b-(--b24ui-border-color)"
    },
    // endregion ////
    // endregion ////
    // region leading ////
    {
      leading: true,
      noPadding: false,
      class: "ps-[34px]"
    },
    // endregion ////
    // region trailing ////
    {
      trailing: true,
      noPadding: false,
      class: "pe-[34px]"
    },
    // endregion ////
    // region loading ////
    {
      loading: true,
      leading: true,
      class: {
        leadingIcon: "size-[21px]"
      }
    },
    {
      loading: true,
      leading: false,
      trailing: true,
      class: {
        trailingIcon: "size-[21px]"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    color: "air-primary"
  }
};

const toast = {
  slots: {
    root: [
      "dark",
      "relative group overflow-hidden",
      "rounded-[26px] py-3.5 ps-6 pe-4",
      "flex items-center gap-2.5",
      "focus-visible:outline-(length:--ui-design-outline-stroke-weight) focus-visible:outline-offset-2 focus-visible:outline-(--ui-color-design-outline-content-divider)",
      "font-[family-name:var(--ui-font-family-primary)]",
      "bg-(--ui-color-bg-content-primary)/80",
      "text-(--ui-color-design-plain-na-focused-content)",
      "text-(length:--ui-font-size-sm) font-(--ui-font-weight-normal)"
    ].join(" "),
    wrapper: "w-0 flex-1 flex flex-col",
    title: "font-(--ui-font-weight-medium)",
    description: "",
    icon: "shrink-0 size-6 text-(--b24ui-border-color)",
    avatar: "shrink-0",
    avatarSize: "xl",
    actions: "flex gap-1.5 shrink-0",
    progress: "absolute inset-x-0 bottom-0",
    //  h-1 z-[-1] bg-(--b24ui-border-color)
    close: "p-0"
  },
  variants: {
    color: {
      "air-primary": { root: "style-filled" },
      "air-primary-success": { root: "style-filled-success" },
      "air-primary-alert": { root: "style-filled-alert" },
      "air-primary-copilot": { root: "style-filled-copilot" },
      "air-primary-warning": { root: "style-filled-warning" },
      "air-secondary": { root: "style-filled-inverted" },
      // @deprecate ////
      "default": { root: "old-style-default" },
      "danger": { root: "old-style-danger" },
      "success": { root: "old-style-success" },
      "warning": { root: "old-style-warning" },
      "primary": { root: "old-style-primary" },
      "secondary": { root: "old-style-secondary" },
      "collab": { root: "old-style-collab" },
      "ai": { root: "old-style-ai" }
    },
    orientation: {
      horizontal: {
        root: "items-center",
        actions: "items-center"
      },
      vertical: {
        root: "items-center",
        actions: "items-start mt-1"
      }
    },
    title: {
      true: {
        description: "mt-1"
      }
    }
  },
  defaultVariants: {
    color: "air-secondary"
  }
};

const toaster = {
  slots: {
    viewport: "fixed flex flex-col w-[calc(100%-2rem)] sm:w-[384px] z-[100] data-[expanded=true]:h-(--height) focus:outline-none mr-(--scrollbar-width)",
    base: "pointer-events-auto absolute inset-x-0 z-(--index) transform-(--transform) data-[expanded=false]:data-[front=false]:h-(--front-height) data-[expanded=false]:data-[front=false]:*:opacity-0 data-[front=false]:*:transition-opacity data-[front=false]:*:duration-100 data-[state=closed]:animate-[toast-closed_200ms_ease-in-out] data-[state=closed]:data-[expanded=false]:data-[front=false]:animate-[toast-collapsed-closed_200ms_ease-in-out] data-[swipe=move]:transition-none transition-[transform,translate,height] duration-200 ease-out"
  },
  variants: {
    position: {
      "top-left": {
        viewport: "left-4"
      },
      "top-center": {
        viewport: "left-1/2 transform -translate-x-1/2"
      },
      "top-right": {
        viewport: "right-4"
      },
      "bottom-left": {
        viewport: "left-4"
      },
      "bottom-center": {
        viewport: "left-1/2 transform -translate-x-1/2"
      },
      "bottom-right": {
        viewport: "right-4"
      }
    },
    swipeDirection: {
      up: "data-[swipe=end]:animate-[toast-slide-up_200ms_ease-out]",
      right: "data-[swipe=end]:animate-[toast-slide-right_200ms_ease-out]",
      down: "data-[swipe=end]:animate-[toast-slide-down_200ms_ease-out]",
      left: "data-[swipe=end]:animate-[toast-slide-left_200ms_ease-out]"
    }
  },
  compoundVariants: [
    {
      position: [
        "top-left",
        "top-center",
        "top-right"
      ],
      class: {
        viewport: "top-4",
        base: "top-0 data-[state=open]:animate-[slide-in-from-top_200ms_ease-in-out]"
      }
    },
    {
      position: [
        "bottom-left",
        "bottom-center",
        "bottom-right"
      ],
      class: {
        viewport: "bottom-4",
        base: "bottom-0 data-[state=open]:animate-[slide-in-from-bottom_200ms_ease-in-out]"
      }
    },
    {
      swipeDirection: [
        "left",
        "right"
      ],
      class: "data-[swipe=move]:translate-x-(--reka-toast-swipe-move-x) data-[swipe=end]:translate-x-(--reka-toast-swipe-end-x) data-[swipe=cancel]:translate-x-0"
    },
    {
      swipeDirection: [
        "up",
        "down"
      ],
      class: "data-[swipe=move]:translate-y-(--reka-toast-swipe-move-y) data-[swipe=end]:translate-y-(--reka-toast-swipe-end-y) data-[swipe=cancel]:translate-y-0"
    }
  ],
  defaultVariants: {
    position: "top-right"
  }
};

const tooltip = {
  slots: {
    content: [
      "dark",
      "flex flex-row items-center justify-between gap-[5px] will-change-[opacity]",
      "min-h-[37px]",
      "shadow-xl/20",
      "select-none",
      "motion-safe:data-[state=delayed-open]:animate-[scale-in_100ms_ease-out] motion-safe:data-[state=closed]:animate-[scale-out_100ms_ease-in]",
      "p-[10px]",
      "text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset) font-[family-name:var(--ui-font-family-primary)]",
      "bg-(--ui-color-bg-content-primary)/80",
      "text-(--ui-color-design-plain-na-focused-content)",
      "rounded-[calc(var(--popup-window-border-radius)-2px)]",
      // @memo see components/popup.css
      // 'border border-(--popup-window-border)'
      "origin-(--reka-tooltip-content-transform-origin)",
      "pointer-events-auto"
    ].join(" "),
    arrow: "fill-(--ui-color-bg-content-primary)/80",
    text: "text-pretty max-w-[200px] min-w-[100px]",
    // truncate
    kbds: `hidden lg:inline-flex items-center shrink-0 gap-0.5 not-first-of-type:before:content-[''] not-first-of-type:before:me-0.5`,
    kbdsSize: "sm",
    kbdsAccent: "default"
  }
};

const sidebarLayout = {
  slots: {
    root: [
      // @memo use src/runtime/index.css
      "sidebar-layout",
      "text-(--b24ui-typography-label-color)",
      "w-full",
      "flex"
    ].join(" "),
    sidebar: [
      // @memo use src/runtime/air-design-tokens/components/navigation-menu.css
      "air-sidebar",
      "before:absolute before:inset-0 before:z-[-1]",
      "before:bg-(--leftmenu-bg-expanded)",
      "w-[240px]",
      "pe-[3px] rtl:me-[14px]",
      "inset-y-0 left-0",
      "max-lg:hidden"
    ].join(" "),
    sidebarSlideoverContainer: "w-full sm:max-w-80",
    sidebarSlideover: [
      "h-full",
      "overflow-hidden",
      "flex flex-col text-(--b24ui-typography-label-color)",
      "bg-(--ui-color-base-white-fixed) dark:bg-(--ui-color-bg-content-primary) edge-dark:bg-[#21334cf0]",
      "ring-1 ring-(--ui-color-divider-vibrant-less)",
      "shadow-xs",
      "rounded-none"
    ].join(" "),
    sidebarSlideoverBtnClose: [
      "-mb-3",
      "px-4 pt-3"
    ].join(" "),
    contentWrapper: [
      "flex-1 flex flex-col",
      ""
    ].join(" "),
    header: [
      // @memo use src/runtime/air-design-tokens/components/navigation-menu.css
      "air-header",
      "px-(--content-area-shift) min-h-(--topbar-height)",
      "flex items-center gap-x-1"
    ].join(" "),
    headerMenuIcon: "lg:hidden",
    headerWrapper: [
      "min-w-0 flex-1",
      "h-full"
    ].join(" "),
    container: [
      "flex-1 flex flex-col gap-[22px]",
      "lg:min-w-0"
    ].join(" "),
    containerWrapper: [
      "grow group/layout-content"
    ].join(" "),
    pageTopWrapper: [
      "text-(--ui-color-base-1)",
      "flex items-center gap-[12px]"
    ].join(" "),
    pageActionsWrapper: [
      // 'min-h-[28px]', // @memo not work
      "flex flex-col md:flex-row items-start md:items-center justify-start gap-[12px]"
    ].join(" "),
    containerWrapperInner: "size-full",
    pageBottomWrapper: "",
    loadingWrapper: [
      "cursor-wait",
      "isolate absolute z-1000 inset-0",
      "w-full h-dvh",
      "flex flex-row flex-nowrap items-center justify-center"
      // @memo save
      // 'bg-[#00204e]/52', //  //#000000/66 // [#00204e]/52
      // 'motion-safe:backdrop-blur-sm'
    ].join(" "),
    loadingIcon: [
      "text-(--ui-color-design-plain-content-icon-secondary)",
      "size-[110px] animate-spin-slow"
    ].join(" ")
  },
  variants: {
    inner: {
      true: {
        root: [
          "--inner",
          "light",
          "relative isolate",
          "h-full",
          "overflow-hidden"
        ].join(" "),
        sidebar: [
          "relative",
          "z-[0]"
        ].join(" "),
        header: [
          "relative"
        ].join(" "),
        container: [
          // @todo make property
          "mt-0"
        ].join(" "),
        containerWrapper: "",
        pageBottomWrapper: "flex-0"
      },
      false: {
        root: [
          "--app",
          "h-screen min-h-screen",
          "max-lg:flex-col"
        ].join(" "),
        sidebar: [
          "fixed"
        ].join(" "),
        header: [
          "relative"
        ].join(" "),
        container: [
          "relative",
          "mt-[22px]"
        ].join(" ")
      }
    },
    offContentScrollbar: {
      false: "",
      true: ""
    },
    useSidebar: {
      true: "",
      false: ""
    },
    useLightContent: {
      true: {
        containerWrapper: [
          // @memo set context* from component
          "light",
          "text-(--ui-color-text-primary)",
          "bg-(--ui-color-bg-content-primary) "
          // backdrop-blur-md dark:bg-(--ui-color-base-7)/80 edge-dark:bg-(--ui-color-base-7)/95
        ].join(" "),
        loadingIcon: [
          "edge-dark:text-(--ui-color-gray-70)"
        ].join(" ")
      },
      false: {
        container: "px-(--content-area-shift)"
      }
    },
    loading: {
      true: ""
    },
    useNavbar: {
      true: {
        container: ""
      },
      false: {
        loadingWrapper: "h-full",
        container: ""
      }
    }
  },
  compoundVariants: [
    // region inner||main.useLightContent ////
    {
      inner: true,
      useLightContent: true,
      class: {
        container: "",
        pageTopWrapper: "px-0 lg:px-0",
        pageActionsWrapper: "px-0 lg:px-0",
        containerWrapper: [
          "p-[20px]",
          "rounded-(--ui-border-radius-md)"
        ].join(" ")
      }
    },
    {
      inner: false,
      useLightContent: true,
      class: {
        container: "lg:pb-2",
        pageTopWrapper: "px-6 lg:px-0",
        pageActionsWrapper: "px-6 lg:px-0",
        containerWrapper: [
          "p-6 lg:px-[22px] lg:py-[15px]",
          "lg:rounded-(--ui-border-radius-md)"
        ].join(" ")
      }
    },
    // endregion ////
    // region inner.noContentScrollbar ////
    {
      inner: true,
      offContentScrollbar: false,
      class: {
        container: "scrollbar-thin scrollbar-transparent overflow-y-scroll"
      }
    },
    // endregion ////
    // region main.useSidebar ////
    {
      inner: true,
      useSidebar: [true, false],
      class: {
        container: "px-[20px] ps-[20px] pe-[10px] pb-[20px] lg:pt-0 lg:px-[20px] lg:ps-[20px] lg:pe-[10px]"
      }
    },
    {
      inner: false,
      useSidebar: true,
      class: {
        // @memo --content-area-left-shift
        // container: 'lg:px-[calc(var(--content-area-shift)-6px)]',
        container: "lg:px-(--content-area-shift)",
        contentWrapper: "lg:pl-[240px] "
      }
    },
    {
      inner: false,
      useSidebar: false,
      class: {
        container: "px-(--content-area-shift) pb-2 lg:pt-2 lg:px-2",
        contentWrapper: "lg:pl-0"
      }
    },
    // endregion ////
    // region main.useNavbar ////
    {
      inner: true,
      useNavbar: [true, false],
      class: {
        container: "h-full"
      }
    },
    {
      inner: false,
      useNavbar: true,
      class: {
        container: "h-[calc(100dvh-var(--topbar-height))]"
      }
    },
    {
      inner: false,
      useNavbar: false,
      class: {
        container: "h-full"
      }
    }
    // endregion ////
  ],
  defaultVariants: {
    inner: false,
    noContentScrollbar: false,
    useLightContent: true
  }
};

const sidebar = {
  slots: {
    root: [
      "h-full min-h-0",
      "flex flex-col"
    ].join(" ")
  }
};

const sidebarHeader = {
  slots: {
    root: [
      "min-h-(--topbar-height)",
      "flex flex-col",
      "text-(--ui-color-base-1)",
      "border-b border-(--leftmenu-bg-divider)",
      "[&>[data-slot=section]+[data-slot=section]]:mt-2.5"
    ].join(" ")
  }
};

const sidebarBody = {
  slots: {
    root: [
      "py-4",
      "flex flex-1 flex-col",
      "overflow-y-auto",
      "[&>[data-slot=section]+[data-slot=section]]:mt-8"
    ].join(" ")
  },
  variants: {
    scrollbarThin: {
      true: {
        root: "scrollbar-thin scrollbar-transparent"
      }
    }
  },
  defaultVariants: {
    scrollbarThin: true
  }
};

const sidebarFooter = {
  slots: {
    root: [
      "py-4",
      "flex flex-col",
      "max-lg:hidden",
      // 'border-t border-(--leftmenu-bg-divider)',
      "[&>[data-slot=section]+[data-slot=section]]:mt-2.5"
    ].join(" ")
  }
};

const sidebarSection = {
  slots: {
    root: [
      "ps-(--menu-items-block-padding-x) rtl:pe-(--menu-items-block-padding-x)",
      "pe-xs rtl:ps-xs",
      "flex flex-col gap-0.5"
    ].join(" ")
  }
};

const sidebarHeading = {
  slots: {
    root: [
      "mb-1",
      "text-(length:--ui-font-size-xs)/[normal]",
      // text-xs/6
      "font-(--ui-font-weight-medium)",
      "text-(--b24ui-typography-legend-color)"
    ].join(" ")
  }
};

const sidebarSpacer = {
  slots: {
    root: [
      "mt-8",
      "flex-1"
    ].join(" ")
  }
};

const navbar = {
  slots: {
    root: [
      // 'py-2.5', // fix
      "h-full flex-1",
      "flex items-center gap-4"
    ].join(" ")
  }
};

const navbarSection = {
  slots: {
    root: [
      "h-full flex flex-row items-center gap-3"
    ].join(" ")
  }
};

const navbarDivider = {
  slots: {
    root: [
      "h-6 w-px",
      "bg-(--ui-color-divider-vibrant-accent-more)"
    ].join(" ")
  }
};

const navbarSpacer = {
  slots: {
    root: [
      "-ml-4",
      "flex-1"
    ].join(" ")
  }
};

const theme = {
  __proto__: null,
  accordion: accordion,
  advice: advice,
  alert: alert,
  avatar: avatar,
  avatarGroup: avatarGroup,
  badge: badge,
  button: button,
  buttonGroup: buttonGroup,
  calendar: calendar,
  checkbox: checkbox,
  chip: chip,
  collapsible: collapsible,
  container: container,
  countdown: countdown,
  descriptionList: descriptionList,
  dropdownMenu: dropdownMenu,
  form: form,
  formField: formField,
  input: input,
  inputMenu: inputMenu,
  inputNumber: inputNumber,
  kbd: kbd,
  link: link,
  modal: modal,
  navbar: navbar,
  navbarDivider: navbarDivider,
  navbarSection: navbarSection,
  navbarSpacer: navbarSpacer,
  navigationMenu: navigationMenu,
  popover: popover,
  progress: progress,
  radioGroup: radioGroup,
  range: range,
  select: select,
  selectMenu: selectMenu,
  separator: separator,
  sidebar: sidebar,
  sidebarBody: sidebarBody,
  sidebarFooter: sidebarFooter,
  sidebarHeader: sidebarHeader,
  sidebarHeading: sidebarHeading,
  sidebarLayout: sidebarLayout,
  sidebarSection: sidebarSection,
  sidebarSpacer: sidebarSpacer,
  skeleton: skeleton,
  slideover: slideover,
  switch: _switch,
  tabs: tabs,
  textarea: textarea,
  toast: toast,
  toaster: toaster,
  tooltip: tooltip
};

const h1 = {
  slots: {
    base: [
      "relative mb-2",
      "scroll-mt-[calc(45px+var(--b24ui-header-height))] lg:scroll-mt-(--b24ui-header-height)",
      "text-(length:--ui-font-size-5xl)"
    ].join(" ")
  },
  variants: {
    accent: {
      "default": "text-(--b24ui-typography-label-color)",
      "accent": "text-(--ui-color-accent-brand-blue)",
      "accent-more": "text-(--ui-color-accent-soft-element-blue)",
      "less": "text-(--b24ui-typography-description-color)",
      "less-more": "text-(--ui-color-design-plain-na-content-secondary)"
    }
  },
  defaultVariants: {
    accent: "default"
  }
};

const h2 = {
  slots: {
    base: [
      "relative mb-2",
      "scroll-mt-[calc(48px+45px+var(--b24ui-header-height))] lg:scroll-mt-[calc(48px+var(--b24ui-header-height))]",
      "text-(length:--ui-font-size-4xl)"
    ].join(" ")
  },
  variants: {
    accent: {
      "default": "text-(--b24ui-typography-label-color)",
      "accent": "text-(--ui-color-accent-brand-blue)",
      "accent-more": "text-(--ui-color-accent-soft-element-blue)",
      "less": "text-(--b24ui-typography-description-color)",
      "less-more": "text-(--ui-color-design-plain-na-content-secondary)"
    }
  },
  defaultVariants: {
    accent: "default"
  }
};

const h3 = {
  slots: {
    base: [
      "relative mb-2",
      "scroll-mt-[calc(32px+45px+var(--b24ui-header-height))] lg:scroll-mt-[calc(32px+var(--b24ui-header-height))]",
      "text-(length:--ui-font-size-3xl)"
    ].join(" ")
  },
  variants: {
    accent: {
      "default": "text-(--b24ui-typography-label-color)",
      "accent": "text-(--ui-color-accent-brand-blue)",
      "accent-more": "text-(--ui-color-accent-soft-element-blue)",
      "less": "text-(--b24ui-typography-description-color)",
      "less-more": "text-(--ui-color-design-plain-na-content-secondary)"
    }
  },
  defaultVariants: {
    accent: "default"
  }
};

const h4 = {
  slots: {
    base: [
      "relative mb-2",
      "scroll-mt-[calc(32px+45px+var(--b24ui-header-height))] lg:scroll-mt-[calc(32px+var(--b24ui-header-height))]",
      "text-(length:--ui-font-size-2xl)"
    ].join(" ")
  },
  variants: {
    accent: {
      "default": "text-(--b24ui-typography-label-color)",
      "accent": "text-(--ui-color-accent-brand-blue)",
      "accent-more": "text-(--ui-color-accent-soft-element-blue)",
      "less": "text-(--b24ui-typography-description-color)",
      "less-more": "text-(--ui-color-design-plain-na-content-secondary)"
    }
  },
  defaultVariants: {
    accent: "default"
  }
};

const h5 = {
  slots: {
    base: [
      "relative mb-2",
      "scroll-mt-[calc(32px+45px+var(--b24ui-header-height))] lg:scroll-mt-[calc(32px+var(--b24ui-header-height))]",
      "text-(length:--ui-font-size-xl)"
    ].join(" ")
  },
  variants: {
    accent: {
      "default": "text-(--b24ui-typography-label-color)",
      "accent": "text-(--ui-color-accent-brand-blue)",
      "accent-more": "text-(--ui-color-accent-soft-element-blue)",
      "less": "text-(--b24ui-typography-description-color)",
      "less-more": "text-(--ui-color-design-plain-na-content-secondary)"
    }
  },
  defaultVariants: {
    accent: "default"
  }
};

const h6 = {
  slots: {
    base: [
      "relative mb-2",
      "scroll-mt-[calc(32px+45px+var(--b24ui-header-height))] lg:scroll-mt-[calc(32px+var(--b24ui-header-height))]",
      "text-(length:--ui-font-size-md)"
    ].join(" ")
  },
  variants: {
    accent: {
      "default": "text-(--b24ui-typography-label-color)",
      "accent": "text-(--ui-color-accent-brand-blue)",
      "accent-more": "text-(--ui-color-accent-soft-element-blue)",
      "less": "text-(--b24ui-typography-description-color)",
      "less-more": "text-(--ui-color-design-plain-na-content-secondary)"
    }
  },
  defaultVariants: {
    accent: "default"
  }
};

const p = {
  slots: {
    base: [
      "mb-2 last:mb-0",
      "leading-relaxed",
      "text-pretty"
    ].join(" ")
  },
  variants: {
    small: {
      true: "text-sm",
      false: "text-base"
    },
    accent: {
      "default": "text-(--b24ui-typography-label-color)",
      "accent": "text-(--ui-color-accent-brand-blue)",
      "accent-more": "text-(--ui-color-accent-soft-element-blue)",
      "less": "text-(--b24ui-typography-description-color)",
      "less-more": "text-(--ui-color-design-plain-na-content-secondary)"
    }
  },
  defaultVariants: {
    small: false,
    accent: "default"
  }
};

const a = {
  slots: {
    base: [
      "cursor-pointer",
      // fix 'text-(--ui-color-accent-main-primary)',
      "text-(--ui-color-accent-main-link)",
      "hover:underline underline-offset-2",
      "hover:text-(--ui-color-accent-main-primary-alt)",
      "focus-visible:outline-info-text",
      "transition-colors [&>code]:transition-colors",
      "[&>code]:ring-dashed",
      "hover:[&>code]:text-(--ui-color-accent-main-primary-alt) hover:[&>code]:bg-(--ui-color-design-selection-bg) hover:[&>code]:ring-(--ui-color-design-selection-stroke)"
    ].join(" ")
  }
};

const blockquote = {
  slots: {
    base: "mb-2 border-s-4 border-(--ui-color-accent-soft-element-blue) ps-4 italic"
  }
};

const strong = {
  slots: {
    base: ""
  }
};

const em = {
  slots: {
    base: ""
  }
};

const ol = {
  slots: {
    base: [
      "list-decimal",
      "ps-2 mb-2 ms-3",
      "marker:text-(--ui-color-design-plain-content)"
    ].join(" ")
  }
};

const ul = {
  slots: {
    base: [
      "list-disc",
      "ps-2 mb-2 ms-3",
      "marker:text-(--ui-color-design-plain-content)"
    ].join(" ")
  }
};

const li = {
  slots: {
    base: [
      "my-1.5 ps-1.5",
      "leading-relaxed",
      "[&>ul]:my-0 [&>ol]:my-0"
    ].join(" ")
  }
};

const hr = {
  slots: {
    base: [
      "my-4",
      "border-t border-(--ui-color-divider-default)"
    ].join(" ")
  }
};

const table = {
  slots: {
    root: [
      "my-4",
      "overflow-x-auto"
    ].join(" "),
    base: ""
  }
};

const thead = {
  slots: {
    base: [
      "bg-(--ui-color-gray-10)",
      "text-(--ui-color-gray-70)"
    ].join(" ")
  }
};

const tbody = {
  slots: {
    base: ""
  }
};

const tr = {
  slots: {
    base: ""
  }
};

const th = {
  slots: {
    base: ""
  }
};

const td = {
  slots: {
    base: [
      "[&_code]:text-xs/5",
      "[&_p]:my-0 [&_p]:leading-6",
      "[&_ul]:my-0 [&_ol]:my-0 [&_ul]:ps-4.5 [&_ol]:ps-4.5",
      "[&_li]:leading-6 [&_li]:my-0.5"
    ].join(" ")
  }
};

const img = {
  slots: {
    base: ""
  }
};

const code = {
  slots: {
    base: [
      "px-1.5 py-0.5",
      "font-[family-name:var(--ui-font-family-system-mono)] font-(--ui-font-weight-medium)",
      "text-(length:--ui-font-size-sm)/[normal]",
      "rounded-(--ui-border-radius-md)",
      "inline-block",
      "ring-(--b24ui-border-width) ring-inset",
      "text-(--b24ui-color) bg-(--b24ui-background) ring-(--b24ui-border-color)"
    ].join(" ")
  },
  variants: {
    color: {
      "air-primary": { base: "style-filled" },
      "air-primary-success": { base: "style-filled-success" },
      "air-primary-alert": { base: "style-filled-alert" },
      "air-primary-copilot": { base: "style-filled-copilot" },
      "air-primary-warning": { base: "style-filled-warning" },
      // @deprecate ////
      "default": { base: "style-old-default" },
      "danger": { base: "style-old-danger" },
      "success": { base: "style-old-success" },
      "warning": { base: "style-old-warning" },
      "primary": { base: "style-old-primary" },
      "secondary": { base: "style-old-secondary" },
      "collab": { base: "style-old-collab" },
      "ai": { base: "style-old-ai" }
    }
  },
  defaultVariants: {
    color: "default"
  }
};

const pre = {
  slots: {
    root: [
      "my-4 p-4",
      "font-b24-system-mono font-semibold text-sm leading-4",
      "bg-black",
      "overflow-auto rounded w-full",
      "text-pretty"
    ].join(" "),
    base: [
      "text-green-350"
    ].join(" ")
  }
};

const themeProse = {
  __proto__: null,
  a: a,
  blockquote: blockquote,
  code: code,
  em: em,
  h1: h1,
  h2: h2,
  h3: h3,
  h4: h4,
  h5: h5,
  h6: h6,
  hr: hr,
  img: img,
  li: li,
  ol: ol,
  p: p,
  pre: pre,
  strong: strong,
  table: table,
  tbody: tbody,
  td: td,
  th: th,
  thead: thead,
  tr: tr,
  ul: ul
};

const tableWrapper = {
  slots: {
    base: [
      "font-[family-name:var(--ui-font-family-primary)] [&>table]:text-(length:--ui-font-size-md)/(--ui-font-line-height-md) [&>table]:relative [&>table]:w-full [&>table]:text-left [&>table]:rtl:text-right",
      "[&>table]:text-(--b24ui-typography-label-color)",
      //
      "[&>table>thead>tr>td]:align-middle",
      "[&>table>thead>tr>th]:align-middle",
      "[&>table>tbody>tr>td]:align-middle",
      "[&>table>tbody>tr>th]:align-middle",
      "[&>table>tfoot>tr>td]:align-middle",
      "[&>table>tfoot>tr>th]:align-middle",
      //
      "[&>table>thead>tr>td]:whitespace-nowrap [&>table>thead>tr>td]:text-(length:--ui-font-size-md) [&>table>thead>tr>td]:font-(--ui-font-weight-normal)",
      "[&>table>thead>tr>th]:whitespace-nowrap [&>table>thead>tr>th]:text-(length:--ui-font-size-md) [&>table>thead>tr>th]:font-(--ui-font-weight-normal)",
      // '[&>table>tbody>tr>td]:whitespace-nowrap [&>table>tbody>tr>td]:text-(length:--ui-font-size-md) [&>table>tbody>tr>td]:font-(--ui-font-weight-normal)',
      "[&>table>tbody>tr>th]:whitespace-nowrap [&>table>tbody>tr>th]:text-(length:--ui-font-size-md) [&>table>tbody>tr>th]:font-(--ui-font-weight-normal)",
      "[&>table>tfoot>tr>td]:whitespace-nowrap [&>table>tfoot>tr>td]:text-(length:--ui-font-size-md) [&>table>tfoot>tr>td]:font-(--ui-font-weight-normal)",
      "[&>table>tfoot>tr>th]:whitespace-nowrap [&>table>tfoot>tr>th]:text-(length:--ui-font-size-md) [&>table>tfoot>tr>th]:font-(--ui-font-weight-normal)",
      //
      "[&>table>thead>tr]:border-(--ui-color-divider-vibrant-accent-more)",
      "[&>table>tbody>tr]:border-(--ui-color-divider-vibrant-default)",
      "[&>table>tfoot]:border-(--ui-color-divider-vibrant-default)",
      `[&>table>thead>tr]:border-b [&>table>tbody>tr:not(:last-child)]:border-b`,
      "[&>table>tfoot]:border-t"
    ].join(" ")
  },
  variants: {
    size: {
      xs: [
        "[&>table>thead>tr>td]:text-(length:--ui-font-size-xs)/(--ui-font-line-height-2xs) [&>table>thead>tr>td]:px-2 [&>table>thead>tr>td]:py-1",
        "[&>table>thead>tr>th]:text-(length:--ui-font-size-xs)/(--ui-font-line-height-2xs) [&>table>thead>tr>th]:px-2 [&>table>thead>tr>th]:py-1",
        "[&>table>tbody>tr>td]:text-(length:--ui-font-size-xs)/(--ui-font-line-height-2xs) [&>table>tbody>tr>td]:px-2 [&>table>tbody>tr>td]:py-1",
        "[&>table>tbody>tr>th]:text-(length:--ui-font-size-xs)/(--ui-font-line-height-2xs) [&>table>tbody>tr>th]:px-2 [&>table>tbody>tr>th]:py-1",
        "[&>table>tfoot>tr>td]:text-(length:--ui-font-size-xs)/(--ui-font-line-height-2xs) [&>table>tfoot>tr>th]:px-2 [&>table>tfoot>tr>th]:py-1",
        "[&>table>tfoot>tr>th]:text-(length:--ui-font-size-xs)/(--ui-font-line-height-2xs) [&>table>tfoot>tr>th]:px-2 [&>table>tfoot>tr>th]:py-1"
      ].join(" "),
      sm: [
        "[&>table>thead>tr>td]:text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm) [&>table>thead>tr>td]:px-3 [&>table>thead>tr>td]:py-2",
        "[&>table>thead>tr>th]:text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm) [&>table>thead>tr>th]:px-3 [&>table>thead>tr>th]:py-2",
        "[&>table>tbody>tr>td]:text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm) [&>table>tbody>tr>td]:px-3 [&>table>tbody>tr>td]:py-2",
        "[&>table>tbody>tr>th]:text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm) [&>table>tbody>tr>th]:px-3 [&>table>tbody>tr>th]:py-2",
        "[&>table>tfoot>tr>td]:text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm) [&>table>tfoot>tr>th]:px-3 [&>table>tfoot>tr>th]:py-2",
        "[&>table>tfoot>tr>th]:text-(length:--ui-font-size-sm)/(--ui-font-line-height-sm) [&>table>tfoot>tr>th]:px-3 [&>table>tfoot>tr>th]:py-2"
      ].join(" "),
      md: [
        "[&>table>thead>tr>td]:text-(length:--ui-font-size-md)/(--ui-font-line-height-md) [&>table>thead>tr>td]:px-4 [&>table>thead>tr>td]:py-3.5",
        "[&>table>thead>tr>th]:text-(length:--ui-font-size-md)/(--ui-font-line-height-md) [&>table>thead>tr>th]:px-4 [&>table>thead>tr>th]:py-3.5",
        "[&>table>tbody>tr>td]:text-(length:--ui-font-size-md)/(--ui-font-line-height-md) [&>table>tbody>tr>td]:px-4 [&>table>tbody>tr>td]:py-3.5",
        "[&>table>tbody>tr>th]:text-(length:--ui-font-size-md)/(--ui-font-line-height-md) [&>table>tbody>tr>th]:px-4 [&>table>tbody>tr>th]:py-3.5",
        "[&>table>tfoot>tr>td]:text-(length:--ui-font-size-md)/(--ui-font-line-height-md) [&>table>tfoot>tr>th]:px-4 [&>table>tfoot>tr>th]:py-3.5",
        "[&>table>tfoot>tr>th]:text-(length:--ui-font-size-md)/(--ui-font-line-height-md) [&>table>tfoot>tr>th]:px-4 [&>table>tfoot>tr>th]:py-3.5"
      ].join(" "),
      lg: [
        "[&>table>thead>tr>td]:text-(length:--ui-font-size-lg)/(--ui-font-line-height-lg) [&>table>thead>tr>td]:px-5 [&>table>thead>tr>td]:py-4",
        "[&>table>thead>tr>th]:text-(length:--ui-font-size-lg)/(--ui-font-line-height-lg) [&>table>thead>tr>th]:px-5 [&>table>thead>tr>th]:py-4",
        "[&>table>tbody>tr>td]:text-(length:--ui-font-size-lg)/(--ui-font-line-height-lg) [&>table>tbody>tr>td]:px-5 [&>table>tbody>tr>td]:py-4",
        "[&>table>tbody>tr>th]:text-(length:--ui-font-size-lg)/(--ui-font-line-height-lg) [&>table>tbody>tr>th]:px-5 [&>table>tbody>tr>th]:py-4",
        "[&>table>tfoot>tr>td]:text-(length:--ui-font-size-lg)/(--ui-font-line-height-lg) [&>table>tfoot>tr>th]:px-5 [&>table>tfoot>tr>th]:py-4",
        "[&>table>tfoot>tr>th]:text-(length:--ui-font-size-lg)/(--ui-font-line-height-lg) [&>table>tfoot>tr>th]:px-5 [&>table>tfoot>tr>th]:py-4"
      ].join(" ")
    },
    rounded: {
      true: "rounded-(--ui-border-radius-md)",
      false: ""
    },
    zebra: {
      true: [
        "[&>table>tbody>tr]:even:bg-(--ui-color-base-8) [&>table>tbody>tr]:even:[&>td]:bg-(--ui-color-base-8) [&>table>tbody>tr]:even:[&>th]:bg-(--ui-color-base-8)",
        "light:[&>table>tbody>tr]:even:bg-(--ui-color-base-6) light:[&>table>tbody>tr]:even:[&>td]:bg-(--ui-color-base-6) light:[&>table>tbody>tr]:even:[&>th]:bg-(--ui-color-base-6)"
      ].join(" ")
    },
    pinRows: {
      true: [
        "[&>table>thead>tr]:sticky [&>table>thead>tr]:top-0      [&>table>thead>tr]:z-1 dark:[&>table>thead>tr]:bg-(--ui-color-base-5) [&>table>thead>tr]:bg-(--ui-color-base-white-fixed) [&>table>thead>tr]:text-(--ui-color-g-content-grey-1) [&>table>thead>tr]:shadow-bottom-sm",
        // '[&>table>tbody>tr]:sticky [&>table>tbody>tr]:bottom-0 [&>table>tbody>tr]:z-1 dark:[&>table>tbody>tr]:bg-(--ui-color-base-5) [&>table>tbody>tr]:bg-(--ui-color-base-white-fixed) [&>table>tbody>tr]:text-(--ui-color-g-content-grey-1) [&>table>tbody>tr]:shadow-top-sm',
        "[&>table>tfoot>tr]:sticky [&>table>tfoot>tr]:bottom-0   [&>table>tfoot>tr]:z-1 dark:[&>table>tfoot>tr]:bg-(--ui-color-base-5) [&>table>tfoot>tr]:bg-(--ui-color-base-white-fixed) [&>table>tfoot>tr]:text-(--ui-color-g-content-grey-1) [&>table>tfoot>tr]:shadow-top-sm"
      ].join(" ")
    },
    pinCols: {
      true: [
        "[&>table>thead>tr>th]:sticky [&>table>thead>tr>th]:right-0 [&>table>thead>tr>th]:left-0 dark:[&>table>thead>tr>th]:bg-(--ui-color-base-5) [&>table>thead>tr>th]:bg-(--ui-color-base-white-fixed) [&>table>thead>tr>th]:text-(--ui-color-g-content-grey-1)",
        "[&>table>tbody>tr>th]:sticky [&>table>tbody>tr>th]:right-0 [&>table>tbody>tr>th]:left-0 dark:[&>table>tbody>tr>th]:bg-(--ui-color-base-5) [&>table>tbody>tr>th]:bg-(--ui-color-base-white-fixed) [&>table>tbody>tr>th]:text-(--ui-color-g-content-grey-1)",
        "[&>table>tfoot>tr>th]:sticky [&>table>tfoot>tr>th]:right-0 [&>table>tfoot>tr>th]:left-0 dark:[&>table>tfoot>tr>th]:bg-(--ui-color-base-5) [&>table>tfoot>tr>th]:bg-(--ui-color-base-white-fixed) [&>table>tfoot>tr>th]:text-(--ui-color-g-content-grey-1)"
      ].join(" ")
    },
    rowHover: {
      true: [
        "[&>table>tbody>tr]:hover:bg-(--ui-color-base-8) [&>table>tbody>tr]:hover:[&>td]:bg-(--ui-color-base-8) [&>table>tbody>tr]:hover:[&>th]:bg-(--ui-color-base-8)",
        "light:[&>table>tbody>tr]:hover:bg-(--ui-color-base-6) light:[&>table>tbody>tr]:hover:[&>td]:bg-(--ui-color-base-6) light:[&>table>tbody>tr]:hover:[&>th]:bg-(--ui-color-base-6)"
      ].join(" ")
    },
    bordered: {
      true: "border border-(--ui-color-divider-vibrant-default)"
    },
    scrollbarThin: {
      true: "scrollbar-thin"
    }
  },
  defaultVariants: {
    size: "md",
    scrollbarThin: true
  }
};

const themeContent = {
  __proto__: null,
  tableWrapper: tableWrapper
};

function replaceBrackets(value) {
  return value.replace(/\[\[/g, "<").replace(/\]\]/g, ">");
}
function getTemplates(options) {
  const templates = [];
  function generateVariantDeclarations(variants, result, json) {
    return variants.filter((variant) => json.includes(`as typeof ${variant}`)).map((variant) => {
      const keys = Object.keys(result.variants[variant]);
      return `const ${variant} = ${JSON.stringify(keys, null, 2)} as const`;
    });
  }
  for (const component in theme) {
    templates.push({
      filename: `b24ui/${kebabCase(component)}.ts`,
      write: true,
      getContents: async () => {
        const template = theme[component];
        const result = typeof template === "function" ? template(options) : template;
        const variants = Object.entries(result.variants || {}).filter(([_, values]) => {
          const keys = Object.keys(values);
          return keys.some((key) => key !== "true" && key !== "false");
        }).map(([key]) => key);
        let json = JSON.stringify(result, null, 2);
        for (const variant of variants) {
          json = json.replace(new RegExp(`("${variant}": "[^"]+")`, "g"), `$1 as typeof ${variant}[number]`);
          json = json.replace(new RegExp(`("${variant}": \\[\\s*)((?:"[^"]+",?\\s*)+)(\\])`, "g"), (_, before, match, after) => {
            const replaced = match.replace(/("[^"]+")/g, `$1 as typeof ${variant}[number]`);
            return `${before}${replaced}${after}`;
          });
        }
        return [
          ...generateVariantDeclarations(variants, result, json),
          `export default ${json}`
        ].join("\n\n");
      }
    });
  }
  for (const component in themeProse) {
    templates.push({
      filename: `b24ui/prose/${kebabCase(component)}.ts`,
      write: true,
      getContents: async () => {
        const template = themeProse[component];
        const result = typeof template === "function" ? template(options) : template;
        const variants = Object.entries(result.variants || {}).filter(([_, values]) => {
          const keys = Object.keys(values);
          return keys.some((key) => key !== "true" && key !== "false");
        }).map(([key]) => key);
        let json = JSON.stringify(result, null, 2);
        for (const variant of variants) {
          json = json.replace(new RegExp(`("${variant}": "[^"]+")`, "g"), `$1 as typeof ${variant}[number]`);
          json = json.replace(new RegExp(`("${variant}": \\[\\s*)((?:"[^"]+",?\\s*)+)(\\])`, "g"), (_, before, match, after) => {
            const replaced = match.replace(/("[^"]+")/g, `$1 as typeof ${variant}[number]`);
            return `${before}${replaced}${after}`;
          });
        }
        return [
          ...generateVariantDeclarations(variants, result, json),
          `export default ${json}`
        ].join("\n\n");
      }
    });
  }
  for (const component in themeContent) {
    templates.push({
      filename: `b24ui/content/${kebabCase(component)}.ts`,
      write: true,
      getContents: async () => {
        const template = themeContent[component];
        const result = typeof template === "function" ? template(options) : template;
        const variants = Object.entries(result.variants || {}).filter(([_, values]) => {
          const keys = Object.keys(values);
          return keys.some((key) => key !== "true" && key !== "false");
        }).map(([key]) => key);
        let json = JSON.stringify(result, null, 2);
        for (const variant of variants) {
          json = json.replace(new RegExp(`("${variant}": "[^"]+")`, "g"), `$1 as typeof ${variant}[number]`);
          json = json.replace(new RegExp(`("${variant}": \\[\\s*)((?:"[^"]+",?\\s*)+)(\\])`, "g"), (_, before, match, after) => {
            const replaced = match.replace(/("[^"]+")/g, `$1 as typeof ${variant}[number]`);
            return `${before}${replaced}${after}`;
          });
        }
        return [
          ...generateVariantDeclarations(variants, result, json),
          `export default ${json}`
        ].join("\n\n");
      }
    });
  }
  templates.push({
    filename: "b24ui.css",
    write: true,
    getContents: () => `@source "./b24ui";

@theme default inline {}
`
  });
  templates.push({
    filename: "b24ui/index.ts",
    write: true,
    getContents: () => Object.keys(theme).map((component) => `export { default as ${component} } from './${kebabCase(component)}'`).join("\n")
  });
  templates.push({
    filename: "types/b24ui.d.ts",
    getContents: () => replaceBrackets(`import * as b24ui from '#build/b24ui'
import type { TVConfig } from '@bitrix24/b24ui-nuxt'
import type { defaultConfig } from 'tailwind-variants'

type AppConfigUI = {
  tv?: typeof defaultConfig
} & TVConfig<typeof b24ui>

declare module '@nuxt/schema' {
  interface AppConfigInput {
    /**
     * Bitrix24 UI theme configuration
     */
    b24ui?: AppConfigUI
  }
}

export {}
`)
  });
  templates.push({
    filename: "b24ui-image-component.ts",
    write: true,
    getContents: ({ app }) => {
      const image = app?.components?.find((c) => c.pascalName === "NuxtImg" && !/nuxt(?:-nightly)?\/dist\/app/.test(c.filePath));
      return image ? genExport(image.filePath, [{ name: image.export, as: "default" }]) : 'export default "img"';
    }
  });
  return templates;
}
function addTemplates(options, nuxt, resolve) {
  const templates = getTemplates(options);
  for (const template of templates) {
    if (template.filename.endsWith(".d.ts")) {
      addTypeTemplate(template);
    } else {
      addTemplate(template);
    }
  }
  nuxt.hook("prepare:types", ({ references }) => {
    references.push({ path: resolve("./runtime/types/app.config.d.ts") });
  });
}

export { getDefaultUiConfig as a, addTemplates as b, defaultOptions as d, getTemplates as g, name as n, version as v };
